{% extends "uicore/page/base.html" %}
{% load js_tags %}
{% load ui_menu_bars %}
{% load ui_utils %}
{% block title %}Changelog{% endblock %}

{% block submenu %}{% menu_bar_settings %}{% endblock submenu %}
{% block content %}
    <div class="container">
        <h3>Changelog</h3>

    {% if sapath_enabled %}
        <!-- Do all as 1 big change -->

        <div class="card mt-4">
            <div class="card-header">VariantGrid 4.0</div>
            <div class="card-body">
                <h4>Analysis</h4>
                <ul>
                    <li>Performance improvements: node counts, node loading, adding tags</li>
                    <li>Performance improvement creating analysis from template</li>
                    <li>Prevent users from loading nodes multiple times by repeated clicking (this was causing performance issues)</li>
                    <li>#139 - Analysis grid tags / classifications take into account sharing / withdrawn</li>
                    <li>#137 - Tags in analysis shown on analysis listing page</li>
                    <li>New Node - MOI (Mode of Inheritance) - uses zygosity and gene/disease associations (GenCC)</li>
                    <li>EffectNode - has been changed to use the new pathogenicity columns, it automatically switches based on the annotation version used.</li>
                    <li>Tissue Node works again (new Human Protein Atlas annotation)</li>
                    <li>Trio Node Proband HET option removed + replaced with SampleNode (which is faster)</li>
                    <li>Population node, can exclude filtered by gnomAD</li>
                    <li>PhenotypeNode - support MONDO, shows a warning if ontology term doesn't map to genes</li>
                </ul>

                <h4>Annotation</h4>
                <ul>
                    <li>New dbNSFP gene columns (see <a href="{% url 'view_annotation_descriptions' %}">annotation descriptions)</a>
                        GDI, GDI-Phred, P(HI), GHIS, P(rec), HIPred_score, Gene_indispensability_score,
                        Expression(egenetics), Expression(GNF/Atlas),
                        BioCarta Pathway, ConsensusPathDB Pathway, KEGG IDs, KEGG Pathway,
                        GWAS traits, GO process, GO cellular, GO mol function, BioGRID interactions,
                        ConsensusPathDB interactions, gnomAD prob LOF intolerant, gnomAD prob LOF HOM,
                        gnomAD prob LOF tolerant, Essential Gene (CRISPR), Essential Gene (CRISPR2),
                        Essential Gene (Gene Trap)</li>
                    <li>New columns - <a href="https://www.nature.com/articles/s41467-017-00443-5">ALoFT</a>, NMD Escaping Variant</li>
                    <li>Gene annotation column changes:
                        <ul>
                            <li>omim/hpo terms now only use direct links (was too broad due to indirect links)</li>
                            <li>New column "MONDO terms"</li>
                            <li>New MOI columns: eg for GATA2 they would be:
                                <div>
                                    <b>Gene/Disease Supportive or below:</b> <i>monocytopenia with susceptibility to infections=Autosomal dominant (Supportive: Orphanet)</i>
                                </div>
                                <div>
                                    <b>Gene/Disease Moderate or above:</b> <i>deafness-lymphedema-leukemia syndrome=Autosomal dominant (Definitive: TGMI|G2P Strong: Genomics England PanelApp)|acute myeloid leukemia=Autosomal dominant (Strong: Genomics England PanelApp)|myelodysplastic syndrome=Autosomal dominant (Strong: Genomics England PanelApp)|GATA2 deficiency with susceptibility to MDS/AML=Autosomal dominant (Definitive: ClinGen)</i>
                                </div>
                            </li>
                        </ul>
                    </li>
                    <li>Update VEP (v110), updated Mastermind, COSMIC</li>
                    <li>Pathogenicity Predictions have changed:
                        <ul>
                            <li>New columns added: Alphamissense, BayesDel (NoAF), CADD (raw), ClinPred, REVEL, MetaLR, VEST4.

                                <p>These are "rankscores" ie scores scaled 0->1 based on all of the scores in non-synonymous SNVs. There is 1 score per variant (not per transcript)</p>
                                <p>
                                    Previously, we used levels (eg DAMAGING, BENIGN etc), but these were made by taking genome-wide thresholds, and it has been shown that this varies considerably per-gene.
                                    We plan on developing further refinements here, eg being able to automatically use this tool/cutoff for these genes, a different tool/cutoff for other genes etc.
                                </p>
                            </li>
                            <li>Old columns removed: FATHMM, Mutation Assessor, Mutation Taster, Polyphen2, CADD Phred, REVEL Score</li>
                        </ul>
                    </li>
                    <li>gnomAD v4, added some extra fields for GRCh37 gnomAD (v2)</li>
                    <li>Added MAVEdb info/links</li>
                </ul>

                <h4>Classification</h4>
                <ul>
                    <li>Somatic classifications</li>
                    <li>New pathogenicity prediction tool scores now autopopulate classifications</li>
                </ul>

                <h4>Data</h4>
                <ul>
                    <li>Download annotated CSV/VCF from VCF/sample page</li>
                    <li>Samples page (data) - filter for variant type. Fix autocompletes sometimes not searching all builds</li>
                    <li>Cohorts / Pedigree - added user to to make it easier to find</li>
                </ul>

                <h4>Genes</h4>
                <ul>
                    <li>Switch to BioCommons HGVS library (includes adding support for 'inv')</li>
                    <li>Gene Symbol page has been rearranged/updated with new annotation fields</li>
                    <li>Gene/Disease - use GenCC (includes ClinGen) instead of just ClinGen on Gene page, Gene Grid column</li>
                </ul>

                <h4>Ontology</h4>
                <ul>
                    <li>Ontology can have multiple versions (allowing upgrades over time)</li>
                    <li>Patients - phenotype text also matched for MONDO terms</li>
                    <li>Update ontologies to latest versions (and will be able to do this going forward)</li>
                    <li>Ontology term page shows classifications and patients linked to term</li>
                </ul>

                <h4>Search</h4>
                <ul>
                    <li>Search rewritten - gives clearer results, and indication of what was searched in green</li>
                    <li>Can now search for cohort/pedigree, ontology terms</li>
                    <li>dbSNP lookup is faster</li>
                    <li>Improved HGVS resolution (support for gaps, LRG, non-coding transcripts)</li>
                    <li>Manual variant entry - improved user experience creating novel variants via search</li>
                </ul>

                <h4>Tagging</h4>
                <ul>
                    <li>VGP#2439 - Tag Colors used to be 1 per user, can now have multiple which can be shared/set globally (so all lab members can share colors)</li>
                    <li>Tags for Variants grid - grid showing all tags per variant has been put back (was disabled for performance reasons)</li>
                </ul>

                <h4>Variant Details</h4>
                <ul>
                    <li>Show other variants on locus regardless of genotype calls, also link to alleles that were once on the same row in VCF files</li>
                    <li>#215 - show a quick summary of gene info (gene/disease links - click to expand)</li>
                    <li>#788 - show tag counts in grid (showing all tags could get extremely long)</li>
                    <li>#13 - <a href="https://www.ncbi.nlm.nih.gov/refseq/MANE/">MANE</a> transcripts indicated and chosen as default transcript on variant details page/classification</li>
                    <li>#109 - Allow users to see (limited) annotation on intergenic variants</li>
                    <li>#668 - add tooltips to show help about columns on gene / variant pages</li>
                    <li>Nearby variants shows tags and classification summary counts, works across builds</li>
                    <li>Page appears differently for Structural Variants (hiding unavailable annotations, add notes)</li>
                </ul>

               <h4>Variant Grids</h4>
                <ul>
                    <li>#743 - variant grids - fix up column labels. Make some new links (ontology terms,pubmed, dbsnp) - </li>
                    <li>#733 - Variant pages (gene symbol, all variants, tags etc) now styled/links the same as on Anaylsis grid</li>
                </ul>

                <h4>VCF upload</h4>
                <ul>
                    <li>#660 - Give warning if VCF filters are not defined in header (rather than fail to import VCF)</li>
                </ul>
            </div>
        </div>



    {% else %}
        <div class="card mt-4">
            <div class="card-header">24 May 2024</div>
            <div class="card-body">
                <ul>
                    <li>Analysis - export VCF fixes</li>
                    <li>Anaylsis - All Variants Node - can select type (eg symbolic) - useful for testing</li>
                    <li>Analysis - Audit log (record who changes what on nodes)</li>
                    <li>Liftover - added bcftools +liftover to handle anything ClinGen Allele registry can't liftover</li>
                    <li>Search - can add COSMIC variants we don't have (uses API)</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">12 Apr 2024</div>
            <div class="card-body">
                <ul>
                    <li>Annotation - improve upload performance</li>
                    <li>Analysis - bug fix: select text to copy/paste in analysis / grid</li>
                    <li>Analysis - bug fix: sort in analysis grid by sample / genotype columns</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">9 Apr 2024</div>
            <div class="card-body">
                <ul>
                    <li>New columsn hgvs_g, svlen (Structural Variant length), total_db_ref (global variant zygosity count), total_db_unk, predictions_num_benign, predictions_num_pathogenic</li>
                    <li>Variant details - clean up page for SVs</li>
                    <li>Variant details - transcript selections have optional toggle for other annotation consortium transcripts</li>
                    <li>GenomeBuild/Contig pages (eg search for "chrX")</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">28 Mar 2024</div>
            <div class="card-body">
                <ul>
                    <li>Somatic classifications</li>
                    <li>Can download annotated VCF/CSV from VCF page</li>
                    <li>Structural Variant annotation now includes gnomAD, repeat masker</li>
                    <li>Variant details page cleaned up for Structural Variants</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">9 Feb 2024</div>
            <div class="card-body">
                <ul>
                    <li>Upgrade to VEP 110</li>
                    <li>gnomAD 4</li>
                    <li>Added some extra gnomAD fields for GRCh37</li>
                    <li>Alphamissense pathogenicity prediction</li>
                    <li>MAVEdb</li>
                    <li>Support for structural variants (symbolic alts like '&lt;DEL&gt;', '&lt;DUP&gt;' and '&lt;INV&gt;') - these may be missing some annotation like per-base conservation scores</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">25 June 2023</div>
            <div class="card-body">
                <ul>
                    <li>Search rewritten - gives clearer results, and indication of what was searched in green</li>
                    <li>#139 - Analysis grid tags / classifications take into account sharing / withdrawn</li>
                    <li>#783 - Analysis - classifications node can search by lab. New column that shows lab names for internal classifications</li>
                    <li>#790 - Analysis - improve speed of loading variant tags</li>
                    <li>#788 - Variant page - show tag counts in grid (showing all tags could get extremely long)</li>
                    <li>#809 - Variant page - nearby classifications work across builds</li>
                    <li>#819 - classification report now working (could sometimes give permission error)</li>
                    <li>#826 - HGVS - switch library - many issues resolved, can now search for inv</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">26 February 2023</div>
            <div class="card-body">
                <ul>
                    <li>#215 - Variant page - show a quick summary of gene info (gene/disease links - click to expand)</li>
                    <li>#746 - Variant page - 'nearby variants' show gene counts</li>
                    <li>#734 - analysis - tagging variants is now faster/less laggy if you don't have a tag node in the analysis</li>
                    <li>#743 - Variant grids (analysis,tags,all variants etc) Fix up column labels. Make some new links (ontology terms,pubmed, dbsnp) - </li>
                    <li>#763 - pubmed citations shown again on variants page. Fix for some classification autopopulates crashing due to pubmed</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">27 January 2023</div>
            <div class="card-body">
                <ul>
                    <li>VGP#2439 - Tag Colors used to be 1 per user, can now have multiple which can be shared/set globally</li>
                    <li>#137 - Tags shown on analysis listing page, and counted in 'nearby variants'</li>
                    <li>#733 - Variant pages (gene symbol, all variants, tags etc) now styled/links the same as on Anaylsis grid</li>
                    <li>#13 - <a href="https://www.ncbi.nlm.nih.gov/refseq/MANE/">MANE</a> transcripts indicated and chosen as default transcript on variant details page/classification</li>
                    <li>#660 - Give warning if VCF filters are not defined in header (rather than fail to import VCF)</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">3 November 2022</div>
            <div class="card-body">
                <ul>
                    <li>#240 - MergeNode - optimisation</li>
                    <li>#109 - Allow users to see (limited) annotation on intergenic variants</li>
                    <li>#689 - Fix issue where analysis templates not correctly setting</li>
                    <li>#668 - add tooltips to show help about columns on gene / variant pages</li>
                    <li>#692 - Nodes load / cache slightly faster</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">26 August 2022</div>
            <div class="card-body">
                <ul>
                    <li>New dbNSFP gene columns (see <a href="{% url 'view_annotation_descriptions' %}">annotation descriptions)</a>
                        GDI, GDI-Phred, P(HI), GHIS, P(rec), HIPred_score, Gene_indispensability_score,
                        Expression(egenetics), Expression(GNF/Atlas),
                        BioCarta Pathway, ConsensusPathDB Pathway, KEGG IDs, KEGG Pathway,
                        GWAS traits, GO process, GO cellular, GO mol function, BioGRID interactions,
                        ConsensusPathDB interactions, gnomAD prob LOF intolerant, gnomAD prob LOF HOM,
                        gnomAD prob LOF tolerant, Essential Gene (CRISPR), Essential Gene (CRISPR2),
                        Essential Gene (Gene Trap)</li>
                    <li>Gene Symbol page has been rearranged/updated with new fields above</li>
                    <li>Performance improvements - Variant Tags, All variants grid, creating analysis from templates</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">6 August 2022</div>
            <div class="card-body">
                <ul>
                    <li>Analysis - Prevent users from loading nodes multiple times (this was causing performance issues)</li>
                    <li>Annotation - Ontology Version upgrade (latest versions of HPO etc)</li>
                    <li>Nearby variants shows classification summary counts</li>
                    <li>Classification - new pathogenicity prediction tool scores now autopopulate classifications</li>
                    <li>Variant page - show other variants on locus regardless of genotype calls, also link to alleles that were once on the same row in VCF files</li>
                    <li>Cleanup duplicated web page code (internal change only - please report anything that's broken)</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">28 May 2022</div>
            <div class="card-body">
                <ul>
                    <li>Annotation - new columns - <a href="https://www.nature.com/articles/s41467-017-00443-5">ALoFT</a>, NMD Escaping Variant</li>
                    <li>Annotation - VEP (v106), updated Mastermind, COSMIC</li>
                    <li>Annotation - Pathogenicity Predictions have changed:
                        <ul>
                            <li>New columns added: BayesDel (NoAF), CADD (raw), ClinPred, REVEL, MetaLR, VEST4.

                                <p>These are "rankscores" ie scores scaled 0->1 based on all of the scores in non-synonymous SNVs. There is 1 score per variant (not per transcript)</p>
                                <p>
                                    Previously, we used levels (eg DAMAGING, BENIGN etc), but these were made by taking genome-wide thresholds, and it has been shown that this varies considerably per-gene.
                                    We plan on developing further refinements here, eg being able to automatically use this tool/cutoff for these genes, a different tool/cutoff for other genes etc.
                                </p>
                            </li>
                            <li>Old columns removed: FATHMM, Mutation Assessor, Mutation Taster, Polyphen2, CADD Phred, REVEL Score</li>
                        </ul>
                    </li>
                    {% if url_name_visible.analysis %}
                        <li>Analysis - EffectNode - has been changed to use the new pathogenicity columns, it automatically switches based on the annotation version used.</li>
                        <li>Analysis - new columns: variant wiki, gene wiki (added to all columns)</li>
                    {% endif %}

                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">7 May 2022</div>
            <div class="card-body">
                <ul>
                {% if url_name_visible.analysis %}
                    <li>Analysis - Tissue Node works again (new Human Protein Atlas annotation)</li>
                {% endif %}
                    <li>Wiki import/export, better gene/variant wiki grids</li>
                    <li>Genes/Transcripts use cdot internally</li>
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">Mar 3, 2022</div>
            <div class="card-body">
                <ul>
                {% if url_name_visible.analysis %}
                    <li>Analysis - Performance improvements</li>
                    <li>Analysis - Trio Proband HET removed + replaced with SampleNode (which is faster)</li>
                {% endif %}
                {% if url_name_visible.variant_tags %}
                    <li>Variant Tags grid - performance improvements</li>
                {% endif %}
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">Feb 8, 2022</div>
            <div class="card-body">
                <ul>
                    <li>Classifications - added Classification Dashboard</li>
                    <li>Classifications - increased download speed and ability to produce multiple zip entries for a MVL download</li>
                    <li>Ontology - Update outgoing link to MONDO</li>
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">Jan 8, 2022</div>
            <div class="card-body">
                <ul>
                {% if url_name_visible.analysis %}
                    <li>Analysis - pop node, can exclude filtered by gnomAD</li>
                    <li>Analysis - PhenotypeNode can now use MONDO</li>
                    <li>Analysis - Gene annotation column changes:
                        <ul>
                            <li>omim/hpo terms now only use direct links (was too broad due to indirect links)</li>
                            <li>New column "MONDO terms"</li>
                            <li>New MOI columns: eg for GATA2 they would be:
                                <div>
                                    <b>Gene/Disease Supportive or below:</b> <i>monocytopenia with susceptibility to infections=Autosomal dominant (Supportive: Orphanet)</i>
                                </div>
                                <div>
                                    <b>Gene/Disease Moderate or above:</b> <i>deafness-lymphedema-leukemia syndrome=Autosomal dominant (Definitive: TGMI|G2P Strong: Genomics England PanelApp)|acute myeloid leukemia=Autosomal dominant (Strong: Genomics England PanelApp)|myelodysplastic syndrome=Autosomal dominant (Strong: Genomics England PanelApp)|GATA2 deficiency with susceptibility to MDS/AML=Autosomal dominant (Definitive: ClinGen)</i>
                                </div>
                            </li>
                        </ul>
                    </li>
                {% endif %}
                    <li>Patients - phenotype text also matched for MONDO terms</li>
                    <li>Search - dbSNP lookup is faster, can search for cohort/pedigree</li>
                    <li>Data - add user to pedigree/cohort to make it easier to find</li>
                    <li>Manual variant entry - improved user experience creating novel variants via search</li>
                    <li>Gene/Disease - use GenCC (includes ClinGen) instead of just ClinGen on Gene page, Gene Grid column</li>
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">November 30, 2021</div>
            <div class="card-body">
                <ul>
                {% if url_name_visible.analysis %}
                    <li>Analysis - export CSV now matches grid (previously only grid but not CSV had allele frequency as percent and was sorted)</li>
                    <li>Analysis - Phenotype node now shows a warning if ontology term doesn't map to genes</li>
                {% endif %}
                    <li>HGVS - Correctly handle RefSeq gapped alignments, update gene annotation</li>
                    <li>HGVS - support LRG + non-coding transcripts</li>
                    <li>ClinVar - update to latest version</li>
                    <li>Samples page (data) - filter for variant type. Fix autocompletes sometimes not searching all builds</li>
                    <li>VCF import - support DRAGEN VCFs - filters out non-standard variants like CNVs etc - show warnings of # skipped</li>
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">August 24, 2021</div>
            <div class="card-body">
                <ul>
                    <li>Search works for ontology terms</li>
                    <li>Ontology term page shows classifications and patients linked to term</li>
                    {% if site_name == 'Shariant' %}
                        <li>ClinVar Export introduced</li>
                    {% endif %}
                    <li>Allele Frequency - now in range 0-1 (not percent) - and uses values in VCF if provided.</li>
                    <li>Variant Tags can be added/removed from variant details page</li>
                </ul>
            </div>
        </div>

        <br/>
        <a class="toggle-link" data-toggle="collapse" href="#older-changelog" aria-expanded="false">
            Toggle older changes
        </a>

        <div id="older-changelog" class="collapse">
            <div class="card mt-4">
                <div class="card-header">March 10, 2021</div>
                <div class="card-body">
                    <ul>
                        <li>Ontology (OMIM/HPO) updates</li>
                {% if url_name_visible.analysis %}
                        <li>Analysis - GeneListNode - filter to PanelApp panels</li>
                {% endif %}
                        <li>gnomAD 3 (GRCh38 only) - 4 new fields: AC/AN, popmax AC/AN (autopop classifications)</li>
                        <li>Variant tags liftover between builds</li>
                        <li>DamageNode renamed to EffectNode - now has splicing, domain, literature filters</li>
                        <li>Custom Columns - can change name</li>
                    </ul>
                </div>
            </div>

            <div class="card mt-4">
                <div class="card-header">Jan 11, 2021</div>
                <div class="card-body">
                    <ul>
                        <li>Support GRCh38 genome build</li>
                        <li><a class="hover-link" target="_blank" href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00761-2">Somalier</a> - tool to check sample relatedness (see sample page)</li>
                        <li>Added PanelApp Australia to GeneGrid</li>
                        <li>Gene Symbol page shows PanelApp info</li>
                    </ul>
                </div>
            </div>

            <div class="card mt-4">
                <div class="card-header">November 24, 2020</div>
                <div class="card-body">
                    <ul>
                {% if url_name_visible.analysis %}
                        <li>Show tags from other analyses</li>
                        <li>Analyses can be locked (via analysis settings) and set to read-only</li>
                {% endif %}
                        <li>Nearby Variants + Classifications</li>
                        <li>MONDO terms are now recognised in the classification form</li>
                        <li>merge & venn joins should be much faster</li>
                    </ul>
                </div>
            </div>
    {% endif %}
        </div>

    </div>
{% endblock content %}