{% extends "uicore/page/base.html" %}
{% load js_tags %}
{% load ui_menu_bars %}
{% load ui_utils %}
{% block title %}Changelog{% endblock %}

{% block submenu %}{% menu_bar_settings %}{% endblock submenu %}
{% block content %}
    <div class="container">
        <h3>Changelog</h3>

        <div class="card mt-4">
            <div class="card-header">Mar 3, 2022</div>
            <div class="card-body">
                <ul>
                {% if url_name_visible.analysis %}
                    <li>Analysis - Performance improvements</li>
                    <li>Analysis - Trio Proband HET removed + replaced with SampleNode (which is faster)</li>
                {% endif %}
                {% if url_name_visible.variant_tags %}
                    <li>Variant Tags grid - performance improvements</li>
                {% endif %}
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">Feb 8, 2022</div>
            <div class="card-body">
                <ul>
                    <li>Classifications - added Classification Dashboard</li>
                    <li>Classifications - increased download speed and ability to produce multiple zip entries for a MVL download</li>
                    <li>Ontology - Update outgoing link to MONDO</li>
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">Jan 8, 2022</div>
            <div class="card-body">
                <ul>
                {% if url_name_visible.analysis %}
                    <li>Analysis - pop node, can exclude filtered by gnomAD</li>
                    <li>Analysis - PhenotypeNode can now use MONDO</li>
                    <li>Analysis - Gene annotation column changes:
                        <ul>
                            <li>omim/hpo terms now only use direct links (was too broad due to indirect links)</li>
                            <li>New column "MONDO terms"</li>
                            <li>New MOI columns: eg for GATA2 they would be:
                                <div>
                                    <b>Gene/Disease Supportive or below:</b> <i>monocytopenia with susceptibility to infections=Autosomal dominant (Supportive: Orphanet)</i>
                                </div>
                                <div>
                                    <b>Gene/Disease Moderate or above:</b> <i>deafness-lymphedema-leukemia syndrome=Autosomal dominant (Definitive: TGMI|G2P Strong: Genomics England PanelApp)|acute myeloid leukemia=Autosomal dominant (Strong: Genomics England PanelApp)|myelodysplastic syndrome=Autosomal dominant (Strong: Genomics England PanelApp)|GATA2 deficiency with susceptibility to MDS/AML=Autosomal dominant (Definitive: ClinGen)</i>
                                </div>
                            </li>
                        </ul>
                    </li>
                {% endif %}
                    <li>Patients - phenotype text also matched for MONDO terms</li>
                    <li>Search - dbSNP lookup is faster, can search for cohort/pedigree</li>
                    <li>Data - add user to pedigree/cohort to make it easier to find</li>
                    <li>Manual variant entry - improved user experience creating novel variants via search</li>
                    <li>Gene/Disease - use GenCC (includes ClinGen) instead of just ClinGen on Gene page, Gene Grid column</li>
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>


        <div class="card mt-4">
            <div class="card-header">November 30, 2021</div>
            <div class="card-body">
                <ul>
                {% if url_name_visible.analysis %}
                    <li>Analysis - export CSV now matches grid (previously only grid but not CSV had allele frequency as percent and was sorted)</li>
                    <li>Analysis - Phenotype node now shows a warning if ontology term doesn't map to genes</li>
                {% endif %}
                    <li>HGVS - Correctly handle RefSeq gapped alignments, update gene annotation</li>
                    <li>HGVS - support LRG + non-coding transcripts</li>
                    <li>ClinVar - update to latest version</li>
                    <li>Samples page (data) - filter for variant type. Fix autocompletes sometimes not searching all builds</li>
                    <li>VCF import - support DRAGEN VCFs - filters out non-standard variants like CNVs etc - show warnings of # skipped</li>
                    <li>Bug fixes</li>
                </ul>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header">August 24, 2021</div>
            <div class="card-body">
                <ul>
                    <li>Search works for ontology terms</li>
                    <li>Ontology term page shows classifications and patients linked to term</li>
                    {% if site_name == 'Shariant' %}
                        <li>ClinVar Export introduced</li>
                    {% endif %}
                    <li>Allele Frequency - now in range 0-1 (not percent) - and uses values in VCF if provided.</li>
                    <li>Variant Tags can be added/removed from variant details page</li>
                </ul>
            </div>
        </div>

        <br/>
        <a class="toggle-link" data-toggle="collapse" href="#older-changelog" aria-expanded="false">
            Toggle older changes
        </a>

        <div id="older-changelog" class="collapse">
            <div class="card mt-4">
                <div class="card-header">March 10, 2021</div>
                <div class="card-body">
                    <ul>
                        <li>Ontology (OMIM/HPO) updates</li>
                {% if url_name_visible.analysis %}
                        <li>Analysis - GeneListNode - filter to PanelApp panels</li>
                {% endif %}
                        <li>gnomAD 3 (GRCh38 only) - 4 new fields: AC/AN, popmax AC/AN (autopop classifications)</li>
                        <li>Variant tags liftover between builds</li>
                        <li>DamageNode renamed to EffectNode - now has splicing, domain, literature filters</li>
                        <li>Custom Columns - can change name</li>
                    </ul>
                </div>
            </div>

            <div class="card mt-4">
                <div class="card-header">Jan 11, 2021</div>
                <div class="card-body">
                    <ul>
                        <li>Support GRCh38 genome build</li>
                        <li><a class="hover-link" target="_blank" href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00761-2">Somalier</a> - tool to check sample relatedness (see sample page)</li>
                        <li>Added PanelApp Australia to GeneGrid</li>
                        <li>Gene Symbol page shows PanelApp info</li>
                    </ul>
                </div>
            </div>

            <div class="card mt-4">
                <div class="card-header">November 24, 2020</div>
                <div class="card-body">
                    <ul>
                {% if url_name_visible.analysis %}
                        <li>Show tags from other analyses</li>
                        <li>Analyses can be locked (via analysis settings) and set to read-only</li>
                {% endif %}
                        <li>Nearby Variants + Classifications</li>
                        <li>MONDO terms are now recognised in the classification form</li>
                        <li>merge & venn joins should be much faster</li>
                    </ul>
                </div>
            </div>

        </div>

    </div>
{% endblock content %}