# Generated by blat_keys.py on 2020-09-29

from django.db import migrations

def blat_keys(apps, schema_editor):
    EvidenceKey = apps.get_model("variantclassification", "EvidenceKey")
    EvidenceKey.objects.all().delete()
    keys = [
        EvidenceKey(**{'key': '1000_genomes_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': '1000 Genomes', 'sub_label': None, 'description': 'The MAF found for the variant in the 1000 Genomes general population.', 'examples': [], 'options': [], 'see': 'http://www.internationalgenome.org/1000-genomes-browsers', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'af_1kg', 'created': '2019-10-22 22:46:59.648385+00:00', 'modified': '2019-10-22 22:46:59.648425+00:00'}),
        EvidenceKey(**{'key': 'affected_status', 'mandatory': False, 'max_share_level': 'public', 'order': 5, 'label': '', 'sub_label': '', 'description': 'The affected status of the individual tested.', 'examples': [], 'options': [{'key': 'yes', 'index': 1, 'label': 'Affected'}, {'key': 'no', 'index': 2, 'label': 'Not Affected'}, {'key': 'unknown', 'index': 3, 'label': 'Unknown'}, {'key': 'not_applicable', 'index': 4, 'label': 'Not Applicable'}], 'see': '', 'evidence_category': 'HP', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.648555+00:00', 'modified': '2019-10-22 22:46:59.648571+00:00'}),
        EvidenceKey(**{'key': 'age', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 9, 'label': None, 'sub_label': None, 'description': 'Age of the patient at the time of testing.<br>\nIf no time unit is set, it indicates years.<br>\nIf there are two numbers, it indicates a range.<br>If it ends with a +, it indicates that age or higher', 'examples': ['7', '12gestation_weeks', '10-19', '80+'], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'A', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.648641+00:00', 'modified': '2019-11-22 02:40:27.286580+00:00'}),
        EvidenceKey(**{'key': 'align_gvgd', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Align GVGD', 'sub_label': None, 'description': 'Align-GVGD score (GV score/GD score) or prediction. <p> Score range: GV>0 , GD>0 <p> Prediction categories:  "C0", "C15", "C25", "C35", "C45", "C55" or "C65" <p> Align-GVGD is a freely available, web-based program that combines the biophysical characteristics of amino acids and protein multiple sequence alignments to predict where missense substitutions in genes of interest fall in a spectrum from enriched deleterious to enriched neutral. Align-GVGD is an extension of the original Grantham difference to multiple sequence alignments and true simultaneous multiple comparisons.', 'examples': ['140.5/100.3', 'C65'], 'options': [], 'see': 'http://agvgd.hci.utah.edu/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.648793+00:00', 'modified': '2019-10-22 22:46:59.648803+00:00'}),
        EvidenceKey(**{'key': 'allele_depth', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Allele Depth (AD)', 'sub_label': None, 'description': 'Unfiltered allele depth', 'examples': ['80'], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.648868+00:00', 'modified': '2019-10-22 22:46:59.648881+00:00'}),
        EvidenceKey(**{'key': 'allele_frequency', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 8, 'label': 'Allele Frequency (AF)', 'sub_label': None, 'description': 'Allele depth / total reads for locus', 'examples': ['0.5', '0.75', '0.139534884'], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.648947+00:00', 'modified': '2019-10-22 22:46:59.648957+00:00'}),
        EvidenceKey(**{'key': 'allele_origin', 'mandatory': False, 'max_share_level': 'public', 'order': 12, 'label': None, 'sub_label': None, 'description': 'Origin of the variant indicating whether it is heritable (germline) or non-heritable (somatic).', 'examples': [], 'options': [{'index': 1, 'label': 'Not provided'}, {'key': 'germline', 'index': 2}, {'key': 'somatic', 'index': 4}, {'key': 'likely_germline', 'index': 9}, {'key': 'likely_somatic', 'index': 10}, {'key': 'unknown', 'index': 5}, {'key': 'not_tested', 'index': 6}, {'key': 'tested_inconclusive', 'index': 7, 'label': 'Tested inconclusive'}, {'key': 'other', 'index': 8}], 'see': None, 'evidence_category': 'V', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649021+00:00', 'modified': '2020-05-25 01:46:27.755503+00:00'}),
        EvidenceKey(**{'key': 'allelic_in_cis_w_pathogenic', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'This field can be used to specify the second allele using HGVS nomenclature.', 'examples': ['NM_001852.3:c.976C>T'], 'options': [], 'see': None, 'evidence_category': 'A', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649096+00:00', 'modified': '2019-10-22 22:46:59.649107+00:00'}),
        EvidenceKey(**{'key': 'allelic_in_trans_w_dominant', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'This field can be used to specify the second allele using HGVS nomenclature.', 'examples': ['NM_001852.3:c.976C>T'], 'options': [], 'see': None, 'evidence_category': 'A', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649171+00:00', 'modified': '2019-10-22 22:46:59.649182+00:00'}),
        EvidenceKey(**{'key': 'allelic_in_trans_w_pathogenic', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'This field can be used to specify the second allele using HGVS nomenclature.', 'examples': ['NM_001852.3:c.976C>T'], 'options': [], 'see': None, 'evidence_category': 'A', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649246+00:00', 'modified': '2019-10-22 22:46:59.649256+00:00'}),
        EvidenceKey(**{'key': 'ancestry', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'Ancestry', 'sub_label': '', 'description': 'Ancestry of the patient. Click map to view larger image.  <div> <a href="/static/images/populations.png" target="_blank"><img id="population-map" class="right" src="/static/images/populations.png" style="width: 100%"></img></a> </div>', 'examples': [], 'options': [{'key': 'AFR', 'index': 1, 'label': 'African/African American'}, {'key': 'ASJ', 'index': 2, 'label': 'Ashkenazi Jewish'}, {'key': 'AM', 'index': 3, 'label': 'Australo Melanesian'}, {'key': 'CA', 'index': 12, 'label': 'Central Asian'}, {'key': 'EAS', 'index': 4, 'label': 'East Asian'}, {'key': 'FIN', 'index': 5, 'label': 'Finnish'}, {'key': 'AMR', 'index': 6, 'label': 'Latino/Mixed American'}, {'key': 'NFE', 'index': 7, 'label': 'Non-Finnish European'}, {'key': 'PO', 'index': 8, 'label': 'Polynesian'}, {'key': 'SAS', 'index': 9, 'label': 'South Asian'}, {'key': 'SEA', 'index': 10, 'label': 'South East Asian'}, {'key': 'OTH', 'index': 11, 'label': 'Other'}], 'see': '', 'evidence_category': 'HP', 'value_type': 'M', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649323+00:00', 'modified': '2019-10-22 22:46:59.649333+00:00'}),
        EvidenceKey(**{'key': 'ancestry_source', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'Source of ancestry reporting', 'sub_label': '', 'description': 'How was the ancestry of the patient derived?', 'examples': [], 'options': [{'key': 'self', 'index': 1, 'label': 'Self Reported'}, {'key': 'inferred', 'index': 2, 'label': 'Inferred from sequence'}], 'see': '', 'evidence_category': 'HP', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649422+00:00', 'modified': '2019-10-22 22:46:59.649434+00:00'}),
        EvidenceKey(**{'key': 'a_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'A', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649499+00:00', 'modified': '2019-10-22 22:46:59.649509+00:00'}),
        EvidenceKey(**{'key': 'assertion_method', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'Assertion method', 'sub_label': None, 'description': 'The curation method use to interpret the variant. Novel or in-house curation methods can be added as custom values.', 'examples': [], 'options': [{'key': 'acmg', 'index': 1, 'label': 'ACMG (PMID:25741868)', 'aliases': ['ACMG Guidelines']}, {'key': 'bayesian_acmg', 'index': 2, 'label': 'ACMG Bayesian (PMID:29300386)'}, {'key': 'modified_ACMG', 'index': 3, 'label': 'Modified ACMG'}], 'see': None, 'evidence_category': 'HI', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649570+00:00', 'modified': '2020-05-25 01:46:28.064270+00:00'}),
        EvidenceKey(**{'key': 'a_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'A', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649645+00:00', 'modified': '2019-10-22 22:46:59.649655+00:00'}),
        EvidenceKey(**{'key': 'autopopulate', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 100, 'label': '', 'sub_label': '', 'description': 'Details about the auto-population of this classification based on variant/transcript/sample details', 'examples': [], 'options': [], 'see': '', 'evidence_category': 'SO', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649715+00:00', 'modified': '2019-10-22 22:46:59.649725+00:00'}),
        EvidenceKey(**{'key': 'ba1', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BA1', 'sub_label': 'MAF >5%', 'description': 'Allele frequency is >5% in Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation Consortium', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'P', 'value_type': 'C', 'default_crit_evaluation': 'BA', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649789+00:00', 'modified': '2019-10-22 22:46:59.649799+00:00'}),
        EvidenceKey(**{'key': 'bdgp', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BDGP', 'sub_label': 'Splicing', 'description': 'Berkeley Drosophila Genome Project splice predictor score. <p> Score range: 0 to 1 <p>', 'examples': [], 'options': [], 'see': 'http://www.fruitfly.org/seq_tools/splice.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.649863+00:00', 'modified': '2019-10-22 22:46:59.649873+00:00'}),
        EvidenceKey(**{'key': 'bic_accession', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'BIC Accession Number', 'sub_label': None, 'description': 'Breast Cancer Information Core (BIC) accession number. BIC is an International Collaborative Effort hosted by NHGRI of BRCA1 and BRCA2 variants. Multiple accession numbers can be separated using a comma. <p> NOTE: BIC is no longer being updated.', 'examples': [], 'options': [], 'see': 'https://research.nhgri.nih.gov/bic/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650025+00:00', 'modified': '2019-10-22 22:46:59.650083+00:00'}),
        EvidenceKey(**{'key': 'bic_clinical_classification', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'BIC Clinical Classification', 'sub_label': None, 'description': 'Breast Cancer Information Core (BIC) clinical classification. BIC is an International Collaborative Effort hosted by NHGRI of BRCA1 and BRCA2 variants. <p> NOTE: BIC is no longer being updated.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650247+00:00', 'modified': '2019-10-22 22:46:59.650259+00:00'}),
        EvidenceKey(**{'key': 'bp1', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 35, 'label': 'BP1', 'sub_label': 'Missense when truncating primary cause of disease', 'description': 'Missense variant in a gene for which primarily truncating variants are known to cause disease', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'BP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650324+00:00', 'modified': '2019-10-22 22:46:59.650334+00:00'}),
        EvidenceKey(**{'key': 'bp2', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BP2', 'sub_label': 'Cis (any) or trans (dominant)', 'description': 'Observed in trans with a pathogenic variant for a fully penetrant dominant gene/disorder or observed in cis with a pathogenic variant in any inheritance pattern', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'A', 'value_type': 'C', 'default_crit_evaluation': 'BP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650398+00:00', 'modified': '2019-10-22 22:46:59.650409+00:00'}),
        EvidenceKey(**{'key': 'bp3', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 34, 'label': 'BP3', 'sub_label': 'In-frame indel in repetitive region', 'description': 'In-frame deletions/insertions in a repetitive region without a known function', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'BP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650473+00:00', 'modified': '2019-10-22 22:46:59.650483+00:00'}),
        EvidenceKey(**{'key': 'bp4', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 31, 'label': 'BP4', 'sub_label': 'No impact predicted', 'description': 'Multiple lines of computational evidence suggest no impact on gene or gene product (conservation, evolutionary, splicing impact, etc.)', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'BP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650547+00:00', 'modified': '2019-10-22 22:46:59.650557+00:00'}),
        EvidenceKey(**{'key': 'bp5', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BP5', 'sub_label': 'Alt mechanism of disease for case', 'description': 'Variant found in a case with an alternate molecular basis for disease', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'O', 'value_type': 'C', 'default_crit_evaluation': 'BP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650621+00:00', 'modified': '2019-10-22 22:46:59.650631+00:00'}),
        EvidenceKey(**{'key': 'bp6', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BP6', 'sub_label': 'Benign, reputable source', 'description': 'Reputable source recently reports variant as benign, but the evidence is not available to the laboratory to perform an independent evaluation', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'DB', 'value_type': 'C', 'default_crit_evaluation': 'BP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650695+00:00', 'modified': '2019-10-22 22:46:59.650705+00:00'}),
        EvidenceKey(**{'key': 'bp7', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 33, 'label': 'BP7', 'sub_label': 'Synonymous, no impact splicing, not conserved', 'description': 'A synonymous (silent) variant for which splicing prediction algorithms predict no impact to the splice consensus sequence nor the creation of a new splice site AND the nucleotide is not highly conserved', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'BP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650769+00:00', 'modified': '2019-10-22 22:46:59.650780+00:00'}),
        EvidenceKey(**{'key': 'bs1', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BS1', 'sub_label': 'MAF too high for disorder', 'description': 'Allele frequency is greater than expected for disorder', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'P', 'value_type': 'C', 'default_crit_evaluation': 'BS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650844+00:00', 'modified': '2019-10-22 22:46:59.650854+00:00'}),
        EvidenceKey(**{'key': 'bs2', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BS2', 'sub_label': 'Observation in controls is inconsistent with disease penetrance', 'description': 'Observed in a healthy adult individual for a recessive (homozygous), dominant (heterozygous), or X-linked (hemizygous) disorder, with full penetrance expected at an early age', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'P', 'value_type': 'C', 'default_crit_evaluation': 'BS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650918+00:00', 'modified': '2019-10-22 22:46:59.650928+00:00'}),
        EvidenceKey(**{'key': 'bs3', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BS3', 'sub_label': 'No effect in functional assay', 'description': 'Well-established in vitro or in vivo functional studies show no damaging effect on protein function or splicing', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'F', 'value_type': 'C', 'default_crit_evaluation': 'BS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.650991+00:00', 'modified': '2019-10-22 22:46:59.651001+00:00'}),
        EvidenceKey(**{'key': 'bs4', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'BS4', 'sub_label': 'Lack of segregation', 'description': 'Lack of segregation in affected members of a family', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'S', 'value_type': 'C', 'default_crit_evaluation': 'BS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651068+00:00', 'modified': '2019-10-22 22:46:59.651079+00:00'}),
        EvidenceKey(**{'key': 'cadd', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'CADD', 'sub_label': 'Meta-predictor', 'description': 'Combined Annotation Dependent Depletion (CADD) score. <p> Score range: -7.535 to 35.789 <p>CADD is a tool for scoring the deleteriousness of single nucleotide variants as well as insertion/deletions variants in the human genome.', 'examples': ['23.5'], 'options': [], 'see': 'https://cadd.gs.washington.edu/', 'evidence_category': 'CP', 'value_type': 'L', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'cadd_phred', 'created': '2019-10-22 22:46:59.651143+00:00', 'modified': '2019-10-22 22:46:59.651153+00:00'}),
        EvidenceKey(**{'key': 'capture_method', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 3, 'label': None, 'sub_label': None, 'description': None, 'examples': [], 'options': [{'key': 'agilent_sureselect_qxt_crev1', 'index': 1, 'label': 'Aglient SureSelect QXT CREv1'}, {'key': 'nextera_nrcce', 'index': 2, 'label': 'Nextera NRCCE'}, {'key': 'roche_seqcap', 'index': 3, 'label': 'Roche SeqCap'}], 'see': None, 'evidence_category': 'HT', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651224+00:00', 'modified': '2019-10-22 22:46:59.651235+00:00'}),
        EvidenceKey(**{'key': 'casecontrol_or', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'Case Control Odds Ratio', 'sub_label': None, 'description': 'Odds ratio from case control study, including confidence intervals and confidence levels. Odds ratio is a measure of association between a genotype and a phenotype. <p> Write as: [Odds ratio] (Confidence Interval) w/ [confidence level] <p> NOTE: To meet PS4 requirement, odds ratio is >5.0, and the confidence interval around the estimate of odds ratio does not include 1.0.', 'examples': ['15.4 (9.6-18.1) w/ 95% confidence'], 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651300+00:00', 'modified': '2019-10-22 22:46:59.651310+00:00'}),
        EvidenceKey(**{'key': 'casecontrol_rr', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'Case Control Relative Risk', 'sub_label': None, 'description': 'Relative risk from case control study, including confidence intervals and confidence levels. Relative risk is a measure of association between a genotype and a phenotype. <p> Write as: [Relative Risk] (Confidence Interval) w/ [confidence level] <p> NOTE: To meet PS4 requirement, relative risk is >5.0, and the confidence interval around the estimate of relative risk does not include 1.0.', 'examples': ['15.4 (9.6-18.1) w/ 95% confidence'], 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651374+00:00', 'modified': '2019-10-22 22:46:59.651387+00:00'}),
        EvidenceKey(**{'key': 'case_controls', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'Case control', 'sub_label': None, 'description': 'Prevalence of cases compared to controls.', 'examples': None, 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651453+00:00', 'modified': '2019-10-22 22:46:59.651463+00:00'}),
        EvidenceKey(**{'key': 'cd_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 40, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'CP', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651528+00:00', 'modified': '2019-10-22 22:46:59.651538+00:00'}),
        EvidenceKey(**{'key': 'c_hgvs', 'mandatory': False, 'max_share_level': 'public', 'order': 6, 'label': 'c.HGVS', 'sub_label': None, 'description': 'HGVS coding change, using (versioned) RefSeq Transcript IDs. If no version is supplied the most recent version in our database will be used, which may not be what you want! <p>If multiple HGVS expressions are provided, they must all match to the same genomic coordinate.   <p>The <a href="http://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/landing" target="_blank">ClinGen Allele Registry converter</a> can be used to confirm HGVS nomenclature. See <a href="http://varnomen.hgvs.org/recommendations/DNA/variant/substitution/" target="_blank">here</a> for HGVS-nomenclature guidelines.', 'examples': ['NM_001080463.1:c.5972T>A', 'NM_001377.3(DYNC2H1):c.5972T>A'], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': 'hgvs_c', 'created': '2019-10-22 22:46:59.651601+00:00', 'modified': '2019-10-22 22:46:59.651611+00:00'}),
        EvidenceKey(**{'key': 'clingen_allele_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'ClinGen Canonical Allele Identifier', 'sub_label': '', 'description': 'A globally unique identifier for each sequence change mapped across genome builds.', 'examples': ['CA321211'], 'options': [], 'see': 'http://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/landing', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651681+00:00', 'modified': '2019-10-22 22:46:59.651692+00:00'}),
        EvidenceKey(**{'key': 'clinical_diagnosis', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 1, 'label': None, 'sub_label': None, 'description': 'Diagnosis for which the patient meets specified diagnostic criteria (with or without biochemical or molecular testing). Suspected diagnoses can be entered by prefixing the term with a question mark.   Recommended disease ontologies include <a href="https://www.omim.org/" target="_blank">OMIM</a>, <a href="https://www.ebi.ac.uk/ols/ontologies/mondo" target="_blank">MONDO</a>, <a href="http://www.orphadata.org/cgi-bin/index.php#ontologies" target="_blank">ORDO</a> and <a href="https://www.ebi.ac.uk/ols/ontologies/doid" target="_blank">DOID</a> .', 'examples': ['?OMIM:219700', 'MONDO:0016248'], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651757+00:00', 'modified': '2019-10-22 22:46:59.651767+00:00'}),
        EvidenceKey(**{'key': 'clinical_significance', 'mandatory': True, 'max_share_level': 'public', 'order': 1, 'label': None, 'sub_label': None, 'description': '[Clinical significance levels additional to those specified in ACMG/AMP can be added on a lab by lab basis, e.g. as shown here with 3A,3B,3C, however, these will be hidden unless specified and all additional levels will be mapped back to standard levels e.g. VUS for all other users]<br><br>  The clinical significance of the specified variant in the context of the condition analysed.', 'examples': [], 'options': [{'vg': '1', 'key': 'B', 'index': 1, 'label': 'Benign (1)'}, {'vg': '2', 'key': 'LB', 'index': 2, 'label': 'Likely Benign (2)'}, {'vg': '3', 'key': 'VUS', 'index': 3, 'label': 'VUS (3)', 'aliases': ['US', 'VOUS']}, {'vg': '3', 'key': 'VUS_A', 'index': 4, 'label': 'VUS A (3A)', 'aliases': ['US_A']}, {'vg': '3', 'key': 'VUS_B', 'index': 5, 'label': 'VUS B (3B)', 'aliases': ['US_B']}, {'vg': '3', 'key': 'VUS_C', 'index': 6, 'label': 'VUS C (3C)', 'aliases': ['US_C']}, {'vg': '4', 'key': 'LP', 'index': 7, 'label': 'Likely Pathogenic (4)'}, {'vg': '5', 'key': 'P', 'index': 8, 'label': 'Pathogenic (5)'}, {'vg': '0', 'key': 'D', 'index': 9, 'label': 'Drug Response'}, {'vg': '0', 'key': 'R', 'index': 10, 'label': 'Risk Factor'}, {'vg': '0', 'key': 'A', 'index': 11, 'label': 'Artefact'}], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'HI', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.651831+00:00', 'modified': '2019-11-20 22:25:32.726254+00:00'}),
        EvidenceKey(**{'key': 'clinvar_allele_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Allele ID', 'sub_label': '', 'description': "This identifier is for each individual nucleotide change in the set of variants that were interpreted. In other words, if the variation report is for interpretation of a haplotype defined by five SNPs, each SNP has its own Allele ID. The set of SNPs that make up the haplotype are represented by the Variation ID. Use 'not reported' if the variant is not present in ClinVar. <p> ClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence.", 'examples': ['18835', 'not reported'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinvar_allele_id', 'created': '2019-10-22 22:46:59.651905+00:00', 'modified': '2019-10-22 22:46:59.651915+00:00'}),
        EvidenceKey(**{'key': 'clinvar_clinical_significance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Clinical Significance', 'sub_label': '', 'description': 'Clinical significance/s from ClinVar.', 'examples': [], 'options': [{'key': 'not_provided', 'index': 0, 'label': ''}, {'key': 'Benign', 'index': 1, 'label': ''}, {'key': 'Benign/Likely_benign', 'index': 2, 'label': ''}, {'key': 'Likely_benign', 'index': 3, 'label': ''}, {'key': 'Uncertain_significance', 'index': 4, 'label': ''}, {'key': 'Likely_pathogenic', 'index': 5, 'label': ''}, {'key': 'Pathogenic/Likely_pathogenic', 'index': 6, 'label': ''}, {'key': 'Pathogenic', 'index': 7, 'label': ''}, {'key': 'Affects', 'index': 8, 'label': ''}, {'key': 'Conflicting_interpretations_of_pathogenicity', 'index': 9, 'label': ''}, {'key': 'association', 'index': 10, 'label': ''}, {'key': 'drug_response', 'index': 11, 'label': ''}, {'key': 'protective', 'index': 12, 'label': ''}, {'key': 'risk_factor', 'index': 13, 'label': ''}, {'key': 'other', 'index': 14, 'label': ''}], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/', 'evidence_category': 'DB', 'value_type': 'M', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinical_significance', 'created': '2019-10-22 22:46:59.651983+00:00', 'modified': '2019-10-22 22:46:59.651993+00:00'}),
        EvidenceKey(**{'key': 'clinvar_clinical_sources', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Clinical Sources', 'sub_label': '', 'description': "The variant's clinical sources reported as tag-value pairs of database and variant identifier", 'examples': ['VariationID:1234'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinvar_clinical_sources', 'created': '2019-10-22 22:46:59.652062+00:00', 'modified': '2019-10-22 22:46:59.652072+00:00'}),
        EvidenceKey(**{'key': 'clinvar_disease_database_name', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Conditions', 'sub_label': '', 'description': 'Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN', 'examples': ['MedGen:C2676676'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinvar_disease_database_name', 'created': '2019-10-22 22:46:59.652140+00:00', 'modified': '2019-10-22 22:46:59.652150+00:00'}),
        EvidenceKey(**{'key': 'clinvar_origin', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Origin', 'sub_label': '', 'description': 'Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other', 'examples': ['VariationID:1234'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinvar_origin', 'created': '2019-10-22 22:46:59.652218+00:00', 'modified': '2019-10-22 22:46:59.652229+00:00'}),
        EvidenceKey(**{'key': 'clinvar_preferred_disease_name', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Preferred Disease Name', 'sub_label': '', 'description': "ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB", 'examples': ['VariationID:1234'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinvar_preferred_disease_name', 'created': '2019-10-22 22:46:59.652297+00:00', 'modified': '2019-10-22 22:46:59.652307+00:00'}),
        EvidenceKey(**{'key': 'clinvar_review_status', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Review Status', 'sub_label': '', 'description': 'Star ranking from 0-4:<br> 0 - no assertion provided, no assertion criteria provided or no assertion for individual variant<br> 1 - criteria provided single submitter or criteria provided conflicting interpretations<br> 2 - criteria provided multiple submitters no conflicts<br> 3 - reviewed by expert panel<br> 4 - practice guideline', 'examples': [], 'options': [{'key': 'no_assertion_provided', 'index': 0, 'label': 'No assertion provided (0*)'}, {'key': 'no_assertion_criteria_provided', 'index': 1, 'label': 'No assertion criteria provided (0*)'}, {'key': 'no_interpretation_for_the_single_variant', 'index': 2, 'label': 'No interpretation for the single variant (0*)'}, {'key': 'criteria_provided|_conflicting_interpretations', 'index': 3, 'label': 'Criteria provided - conflicting interpretations (1*)'}, {'key': 'criteria_provided|_single_submitter', 'index': 4, 'label': 'Criteria provided - single submitter (1*)'}, {'key': 'criteria_provided|_multiple_submitters|_no_conflicts', 'index': 5, 'label': 'Criteria provided - multiple submitters w no conflicts (2*)'}, {'key': 'reviewed_by_expert_panel', 'index': 6, 'label': 'Reviewed by expert panel (3*)'}, {'key': 'practice_guideline', 'index': 7, 'label': 'Practise guideline (4*)'}], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/', 'evidence_category': 'DB', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinvar_review_status', 'created': '2019-10-22 22:46:59.652375+00:00', 'modified': '2019-10-22 22:46:59.652385+00:00'}),
        EvidenceKey(**{'key': 'clinvar_variation_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ClinVar Variation ID', 'sub_label': '', 'description': 'The Variation ID is a unique identifier for the set of sequence changes that were interpreted. In other words, if the variation report is for a haplotype defined by five SNPs, there is a single Variation ID that represents that combination of SNPs in cis.', 'examples': ['3796'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/clinvar/docs/variation_report/#allele', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'clinvar_variation_id', 'created': '2019-10-22 22:46:59.652454+00:00', 'modified': '2019-10-22 22:46:59.652464+00:00'}),
        EvidenceKey(**{'key': 'cohort_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Cohort ID', 'sub_label': None, 'description': 'Cohort or project name used for the analysis.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.652536+00:00', 'modified': '2019-10-22 22:46:59.652548+00:00'}),
        EvidenceKey(**{'key': 'condel', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'Condel', 'sub_label': 'Meta-predictor', 'description': 'Condel 2.0 is a method to assess the outcome of non-synonymous SNVs using a CONsensus DELeteriousness score that combines various tools (MutationAssessor, FATHMM).', 'examples': [], 'options': [], 'see': 'http://bg.upf.edu/condel/home', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.652618+00:00', 'modified': '2019-10-22 22:46:59.652629+00:00'}),
        EvidenceKey(**{'key': 'condition', 'mandatory': True, 'max_share_level': 'public', 'order': 0, 'label': 'Condition under curation', 'sub_label': None, 'description': '\nA condition with an association to the gene for which the variant was interpreted against.<br/>\nInput OMIM, MeSH, MedGen, UMLS, Orphanet, HPO reference numbers and/or free text.<br/><br/>\nFor the recording of phenotypes observed in the patient, see the "Patient phenotype" field.<br/>\n<br/>\nIn the case of uncertainty, list several comma separated terms followed by the word "uncertain".<br/>\n<br/>\nSee the Clinvar recommendations below for more detai\n', 'examples': ['OMIM:219700', 'HP:0007544', 'OMIM:615916, OMIM:611554 uncertain'], 'options': [], 'see': 'ftp://ftp.ncbi.nlm.nih.gov/pub/medgen/presentations/Conditions_Phenotypes.pptx', 'evidence_category': 'H', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.652693+00:00', 'modified': '2020-02-13 07:19:37.528666+00:00'}),
        EvidenceKey(**{'key': 'condition_incidence', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Incidence of condition under curation', 'sub_label': '', 'description': 'The reported incidence of the condition for which the variant was interpreted.', 'examples': [], 'options': [], 'see': '', 'evidence_category': 'H', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.652774+00:00', 'modified': '2019-10-22 22:46:59.652788+00:00'}),
        EvidenceKey(**{'key': 'consent', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 3, 'label': 'Consent codes', 'sub_label': None, 'description': "Can be used to indicate any restrictions placed on the data entered. Enter GA4GH consent code abbreviations separated by a colon. <p> Consent codes defaults are usually set on laboratory sign-up based on the laboratory's standard clinical consent.", 'examples': ['GS-[Australia];NPU'], 'options': [], 'see': 'https://www.ga4gh.org/wp-content/uploads/DataUseBeacon_160209_tab_0.pdf', 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.652854+00:00', 'modified': '2019-10-22 22:46:59.652865+00:00'}),
        EvidenceKey(**{'key': 'consurf', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ConSurf', 'sub_label': None, 'description': 'The ConSurf server is a bioinformatics tool for estimating the evolutionary conservation of amino/nucleic acid positions in a protein/DNA/RNA molecule based on the phylogenetic relations between homologous sequences.', 'examples': [], 'options': [], 'see': 'http://consurf.tau.ac.il/2016/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.652929+00:00', 'modified': '2019-10-22 22:46:59.652939+00:00'}),
        EvidenceKey(**{'key': 'contribution', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 1, 'label': None, 'sub_label': None, 'description': 'Asserted contribution of the variant to the patient\'s phenotype.<p> Full:\tVariant fully explains the patient\'s whole phenotype (for autosomal recessive disorders, use full for each variant in a compound heterozygote)<br> Partial: Variant fully explains part of the patient\'s phenotype <br> Uncertain: Contribution to patient\'s phenotype is unknown <br> None: Variant has no discernible contribution to patient\'s phenotype<br>  Schema based on the UK <a href="https://decipher.sanger.ac.uk/" target="_blank">DECIPHER</a> project.', 'examples': [], 'options': [{'key': 'full', 'index': 1, 'label': ''}, {'key': 'partial', 'index': 2, 'label': ''}, {'key': 'uncertain', 'index': 3, 'label': ''}, {'key': 'none', 'index': 0, 'label': ''}], 'see': None, 'evidence_category': 'HI', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653003+00:00', 'modified': '2019-10-22 22:46:59.653013+00:00'}),
        EvidenceKey(**{'key': 'cosmic_cnt', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'COSMIC Sample Count', 'sub_label': None, 'description': 'The number of samples from COSMIC carrying the variant. In the case of a variant assigned with multiple COSMIC IDs, there are multiple values with comma as separator, listed in the order as for the IDs in the COSMIC_ID column.', 'examples': ['COSM1234'], 'options': [], 'see': 'https://cancer.sanger.ac.uk', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'cosmic_count', 'created': '2019-10-22 22:46:59.653077+00:00', 'modified': '2019-10-22 22:46:59.653087+00:00'}),
        EvidenceKey(**{'key': 'cosmic_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'COSMIC ID', 'sub_label': None, 'description': 'COSMIC (Catalogue of Somatic Mutations In Cancer) Mutation ID. A variant can have more than one IDs, as COSMIC assigns different IDs to the same variation affecting different transcripts. Multiple COSMIC-IDs are separated by a comma.', 'examples': ['COSM1234'], 'options': [], 'see': 'https://cancer.sanger.ac.uk', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'cosmic_id', 'created': '2019-10-22 22:46:59.653151+00:00', 'modified': '2019-10-22 22:46:59.653161+00:00'}),
        EvidenceKey(**{'key': 'cosmic_phe', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'COSMIC Phenotype', 'sub_label': None, 'description': 'Phenotype annotation from COSMIC (Catalogue of Somatic Mutations In Cancer)', 'examples': ['COSM1234'], 'options': [], 'see': 'https://cancer.sanger.ac.uk', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653230+00:00', 'modified': '2019-10-22 22:46:59.653241+00:00'}),
        EvidenceKey(**{'key': 'cp_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 40, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'CP', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653305+00:00', 'modified': '2019-10-22 22:46:59.653315+00:00'}),
        EvidenceKey(**{'key': 'curated_by', 'mandatory': False, 'max_share_level': 'institution', 'order': 1, 'label': 'Curated/reviewed by', 'sub_label': None, 'description': 'Name of individual that most recently curated or reviewed the variant.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'SO', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653390+00:00', 'modified': '2019-10-22 22:46:59.653401+00:00'}),
        EvidenceKey(**{'key': 'curation_context', 'mandatory': False, 'max_share_level': 'public', 'order': 5, 'label': '', 'sub_label': '', 'description': 'Was the variant curated as part of a NATA accredited test? <p> This field is usually set as a default value at laboratory sign-up.', 'examples': [], 'options': [{'key': 'accredited', 'index': 1, 'label': 'Accredited diagnostic testing'}, {'key': 'research', 'index': 2, 'label': 'Research'}, {'key': 'expert', 'index': 3, 'label': 'Expert group'}, {'key': 'other', 'index': 4, 'label': 'Other'}], 'see': '', 'evidence_category': 'HI', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653465+00:00', 'modified': '2019-10-22 22:46:59.653475+00:00'}),
        EvidenceKey(**{'key': 'curation_date', 'mandatory': False, 'max_share_level': 'public', 'order': 2, 'label': 'Curation/review date', 'sub_label': '', 'description': 'Date that most recent curation or review was completed.', 'examples': [], 'options': [], 'see': '', 'evidence_category': 'SO', 'value_type': 'D', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653540+00:00', 'modified': '2019-10-22 22:46:59.653554+00:00'}),
        EvidenceKey(**{'key': 'curation_system', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 6, 'label': None, 'sub_label': None, 'description': 'Software used to perform curation. Set automatically by the system.', 'examples': ['Agilent Alissa Interpret', 'VariantGrid', 'Golden Helix VarSeq VS Clinical'], 'options': None, 'see': None, 'evidence_category': 'HI', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': True, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653620+00:00', 'modified': '2019-10-22 22:46:59.653627+00:00'}),
        EvidenceKey(**{'key': 'curation_verified_by', 'mandatory': False, 'max_share_level': 'institution', 'order': 3, 'label': None, 'sub_label': None, 'description': 'Individual who verified the curation of this variant.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'SO', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653674+00:00', 'modified': '2019-10-22 22:46:59.653682+00:00'}),
        EvidenceKey(**{'key': 'curation_verified_date', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 4, 'label': None, 'sub_label': None, 'description': 'Date that the curation verification was completed.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'SO', 'value_type': 'D', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653730+00:00', 'modified': '2019-10-22 22:46:59.653737+00:00'}),
        EvidenceKey(**{'key': 'db_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'DB', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.653785+00:00', 'modified': '2019-10-22 22:46:59.653793+00:00'}),
        EvidenceKey(**{'key': 'db_rs_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 1, 'label': 'dbSNP ID', 'sub_label': '', 'description': 'Reference SNP (rs) Report', 'examples': ['rs12345'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/SNP/', 'evidence_category': 'P', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'dbsnp_rs_id', 'created': '2019-10-22 22:46:59.653838+00:00', 'modified': '2019-10-22 22:46:59.653845+00:00'}),
        EvidenceKey(**{'key': 'dbscsnv_ada_score', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 21, 'label': 'dbscSNV (Ada) score', 'sub_label': 'Splicing', 'description': 'Splicing consensus score (Ada) a 0-1 probability of altering splicing. From dbscSNV', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267638/', 'evidence_category': 'CP', 'value_type': 'L', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'dbscsnv_ada_score', 'created': '2019-10-22 22:46:59.653898+00:00', 'modified': '2019-10-22 22:46:59.653906+00:00'}),
        EvidenceKey(**{'key': 'dbscsnv_rf_score', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 22, 'label': 'dbscSNV (RF) score', 'sub_label': 'Splicing', 'description': 'Splicing consensus score (RF) a 0-1 probability of altering splicing. From dbscSNV', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267638/', 'evidence_category': 'CP', 'value_type': 'L', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'dbscsnv_rf_score', 'created': '2019-10-22 22:46:59.653957+00:00', 'modified': '2019-10-22 22:46:59.653965+00:00'}),
        EvidenceKey(**{'key': 'db_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'DB', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654020+00:00', 'modified': '2019-10-22 22:46:59.654028+00:00'}),
        EvidenceKey(**{'key': 'decipher_clinical_significance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'DECIPHER Pathogenicity', 'sub_label': None, 'description': 'Pathogenicity of the variant in DECIPHER (Pathogenic, Likely Pathogenic, Uncertain, Likely Benign, Benign).', 'examples': [], 'options': [], 'see': 'https://decipher.sanger.ac.uk/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654074+00:00', 'modified': '2019-10-22 22:46:59.654081+00:00'}),
        EvidenceKey(**{'key': 'decipher_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'DECIPHER ID', 'sub_label': None, 'description': 'DECIPHER (DatabasE of genomiC varIation and Phenotype in Humans using Ensembl Resources) ID. <p> DECIPHER is an interactive web-based database which incorporates a suite of tools designed to aid the interpretation of genomic variants.', 'examples': ['251248'], 'options': [], 'see': 'https://decipher.sanger.ac.uk/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654131+00:00', 'modified': '2019-10-22 22:46:59.654139+00:00'}),
        EvidenceKey(**{'key': 'decipher_phenotype', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'DECIPHER Phenotype', 'sub_label': None, 'description': 'Phenotype/s of patient in DECIPHER.', 'examples': ['HP:0000490'], 'options': [], 'see': 'https://decipher.sanger.ac.uk/', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654187+00:00', 'modified': '2019-10-22 22:46:59.654195+00:00'}),
        EvidenceKey(**{'key': 'denovo_points', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'De Novo Points', 'sub_label': None, 'description': 'Number of points assigned to de novo evidence. The Sequence Variant Interpretation (SVI) Working Group proposes a point-based system to determine the strength of de novo evidence (ACMG/AMP criteria codes PS2 and PM6) based upon three parameters: <br> ● confirmed versus assumed status <br> ● phenotypic consistency <br> ● number of de novo observations <br> Refer to <a href="https://www.clinicalgenome.org/site/assets/files/8490/recommendation_ps2_and_pm6_acmgamp_critiera_version_1_0.pdf" target="_blank">SVI document</a>', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'D', 'value_type': 'L', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654243+00:00', 'modified': '2020-09-28 00:05:44.337588+00:00'}),
        EvidenceKey(**{'key': 'disease_onset', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 4, 'label': 'Disease onset', 'sub_label': None, 'description': 'When was the disease start to affect the patient?', 'examples': ['Unknown', 'Early', 'Late'], 'options': [], 'see': None, 'evidence_category': 'H', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-30 23:06:09.940712+00:00', 'modified': '2019-10-30 23:06:09.940749+00:00'}),
        EvidenceKey(**{'key': 'dob', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'Date of Birth', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'D', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654298+00:00', 'modified': '2019-10-22 22:46:59.654306+00:00'}),
        EvidenceKey(**{'key': 'd_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'D', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654354+00:00', 'modified': '2019-10-22 22:46:59.654361+00:00'}),
        EvidenceKey(**{'key': 'd_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'D', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654406+00:00', 'modified': '2019-10-22 22:46:59.654413+00:00'}),
        EvidenceKey(**{'key': 'ensembl_gene_id', 'mandatory': False, 'max_share_level': 'public', 'order': 0, 'label': 'Ensembl Gene ID', 'sub_label': None, 'description': 'Ensembl gene identifier', 'examples': ['ENSG00000179348'], 'options': [], 'see': 'https://asia.ensembl.org/info/genome/stable_ids/index.html', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654458+00:00', 'modified': '2019-10-22 22:46:59.654466+00:00'}),
        EvidenceKey(**{'key': 'ensembl_protein_id', 'mandatory': False, 'max_share_level': 'public', 'order': 5, 'label': 'Ensembl Protein ID', 'sub_label': '', 'description': 'Ensembl identifier for a protein.', 'examples': ['ENSP00000473460'], 'options': [], 'see': 'https://asia.ensembl.org/info/genome/stable_ids/index.html', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'ensembl_protein', 'created': '2019-10-22 22:46:59.654514+00:00', 'modified': '2019-10-22 22:46:59.654522+00:00'}),
        EvidenceKey(**{'key': 'ensembl_transcript_id', 'mandatory': False, 'max_share_level': 'public', 'order': 4, 'label': 'Ensembl Transcript ID', 'sub_label': None, 'description': 'Ensembl transcript identifier.', 'examples': ['ENST00000367796'], 'options': [], 'see': 'https://asia.ensembl.org/info/genome/stable_ids/index.html', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654573+00:00', 'modified': '2019-10-22 22:46:59.654581+00:00'}),
        EvidenceKey(**{'key': 'entrez_gene_id', 'mandatory': False, 'max_share_level': 'public', 'order': 0, 'label': 'Entrez GeneID', 'sub_label': None, 'description': 'Entrez gene identifier', 'examples': ['5728'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3013746', 'evidence_category': 'V', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-30 23:06:09.940840+00:00', 'modified': '2019-10-30 23:06:09.940853+00:00'}),
        EvidenceKey(**{'key': 'ese_finder_ese', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 21, 'label': 'ESEfinder - ESE', 'sub_label': 'Splicing', 'description': 'Exonic splicing enhancer (ESE) finder v3 SRProteins score. <p> Score range: 0 to 100 <p> ESEfinder s a web-based resource that facilitates rapid analysis of exon sequences to identify putative ESEs responsive to the human SR proteins SF2/ASF, SC35, SRp40 and SRp55, and to predict whether exonic mutations disrupt such elements.', 'examples': [], 'options': [], 'see': 'http://krainer01.cshl.edu/cgi-bin/tools/ESE3/esefinder.cgi', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654629+00:00', 'modified': '2019-10-22 22:46:59.654637+00:00'}),
        EvidenceKey(**{'key': 'ese_finder_splice', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 21, 'label': 'ESEfinder - SpliceSites', 'sub_label': 'Splicing', 'description': 'Exonic splicing enhancer (ESE) finder v3 SpliceSite score. <p> Score range: 0 to 100 <p> ESEfinder is a web-based resource that facilitates rapid analysis of exon sequences to identify putative ESEs responsive to the human SR proteins SF2/ASF, SC35, SRp40 and SRp55, and to predict whether exonic mutations disrupt such elements.', 'examples': [], 'options': [], 'see': 'http://krainer01.cshl.edu/cgi-bin/tools/ESE3/esefinder.cgi', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654685+00:00', 'modified': '2019-10-22 22:46:59.654692+00:00'}),
        EvidenceKey(**{'key': 'esp_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ESP', 'sub_label': None, 'description': 'The MAF found for the variant in ESP general population.', 'examples': [], 'options': [], 'see': 'http://evs.gs.washington.edu/EVS/', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654740+00:00', 'modified': '2019-10-22 22:46:59.654747+00:00'}),
        EvidenceKey(**{'key': 'exac_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'ExAC', 'sub_label': None, 'description': 'The MAF found for the variant in the ExAC general population.', 'examples': [], 'options': [], 'see': 'http://exac.broadinstitute.org/', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654795+00:00', 'modified': '2019-10-22 22:46:59.654802+00:00'}),
        EvidenceKey(**{'key': 'exon', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 14, 'label': None, 'sub_label': None, 'description': 'Used to indicate which exon the variant is located in.', 'examples': ['3 of 25'], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'exon', 'created': '2019-10-22 22:46:59.654850+00:00', 'modified': '2019-10-22 22:46:59.654858+00:00'}),
        EvidenceKey(**{'key': 'exskip', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 21, 'label': 'EX-SKIP', 'sub_label': 'Splicing', 'description': 'EX-SKIP score. <p> EX-SKIP is simple utility that compares the ESE/ESS profile of a wild-type and a mutated allele to quickly determine which exonic variant has the highest chance to skip this exon.', 'examples': [], 'options': [], 'see': 'http://ex-skip.img.cas.cz/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654910+00:00', 'modified': '2019-10-22 22:46:59.654918+00:00'}),
        EvidenceKey(**{'key': 'family_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': -5, 'label': 'Family ID', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.654970+00:00', 'modified': '2019-10-22 22:46:59.654978+00:00'}),
        EvidenceKey(**{'key': 'fathmm', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'fathmm', 'sub_label': None, 'description': 'Functional Analysis through Hidden Markov Models (v2.3) score or prediction. <p> Score range: -16.13 to 10.64 <p> Prediction categories: "Damaging" or "Tolerated" <p/> Fathmm is a high-throughput web-server capable of predicting the functional consequences of both coding variants, i.e. non-synonymous single nucleotide variants (nsSNVs), and non-coding variants.', 'examples': ['Damaging'], 'options': [], 'see': 'http://fathmm.biocompute.org.uk/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'fathmm_pred_most_damaging', 'created': '2019-10-22 22:46:59.655027+00:00', 'modified': '2019-10-22 22:46:59.655035+00:00'}),
        EvidenceKey(**{'key': 'fitcons', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'fitCons', 'sub_label': None, 'description': 'Fitness consequences of functional annotation (fitCons), integrates functional assays (such as ChIP-Seq) with selective pressure inferred using the INSIGHT method.', 'examples': [], 'options': [], 'see': 'http://compgen.cshl.edu/fitCons/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655086+00:00', 'modified': '2019-10-22 22:46:59.655094+00:00'}),
        EvidenceKey(**{'key': 'flossies_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'FLOSSIES', 'sub_label': None, 'description': 'The MAF found for the variant in the Fabulous Ladies Over Seventy (FLOSSIES) population. FLOSSIES contains data from germline DNA of approximately 10,000 women (7,000 European American and 3,000 African American), older than age 70 years who have never had cancer.', 'examples': [], 'options': [], 'see': 'https://whi.color.com/', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655145+00:00', 'modified': '2019-10-22 22:46:59.655153+00:00'}),
        EvidenceKey(**{'key': 'f_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'F', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655202+00:00', 'modified': '2019-10-22 22:46:59.655210+00:00'}),
        EvidenceKey(**{'key': 'fsplice', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'FSPLICE', 'sub_label': 'Splicing', 'description': 'FSPLICE allows users to search for both donor and acceptor sites and to define thresholds for them independently.', 'examples': [], 'options': [], 'see': 'http://www.softberry.com/berry.phtml?topic=fsplice&group=programs&subgroup=gfind', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655258+00:00', 'modified': '2019-10-22 22:46:59.655266+00:00'}),
        EvidenceKey(**{'key': 'f_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'F', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655313+00:00', 'modified': '2019-10-22 22:46:59.655321+00:00'}),
        EvidenceKey(**{'key': 'functional_domain_literature', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 13, 'label': 'Functional domain from literature', 'sub_label': None, 'description': 'Enter functional domain name and literature including PubMed IDs', 'examples': ["zinc finger domain - 'Long et al. Pc2-mediated sumoylation of Smad-interacting protein 1 attenuates transcriptional repression of E-cadherin. J Biol Chem. (2005) (PMID: 16061479)"], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pubmed/', 'evidence_category': 'F', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655369+00:00', 'modified': '2019-10-22 22:46:59.655377+00:00'}),
        EvidenceKey(**{'key': 'g_conservation_assertion', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 16, 'label': 'Nucleotide conservation assertion', 'sub_label': 'Conservation', 'description': 'Assertion of conservation of the reference nucleotide base across species. Can be determined using <a href="https://genome.ucsc.edu/" target="_blank">UCSC Genome Browser</a> and in silico tools. Conservation is of low relevance for truncating variants.  See equivalent field Protein conservation assertion for protein conservation.', 'examples': [], 'options': [{'key': 'very_high', 'index': 1, 'label': ''}, {'key': 'high', 'index': 2, 'label': ''}, {'key': 'moderate', 'index': 3, 'label': ''}, {'key': 'low', 'index': 4, 'label': ''}], 'see': None, 'evidence_category': 'CP', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655428+00:00', 'modified': '2019-10-22 22:46:59.655435+00:00'}),
        EvidenceKey(**{'key': 'gene_constraint', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': None, 'sub_label': None, 'description': 'Degree to which a gene is tolerant to loss of function. Based on z scores & pLI values calculated in gnomAD. In general, a z>2 (for missense) or pLi> 0.9 indicates that LOF of the gene is not tolerated.', 'examples': [], 'options': [{'key': 'tolerant', 'index': 1, 'label': ''}, {'key': 'intolerant', 'index': 2, 'label': ''}], 'see': 'https://gnomad.broadinstitute.org/', 'evidence_category': 'H', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655484+00:00', 'modified': '2019-10-22 22:46:59.655491+00:00'}),
        EvidenceKey(**{'key': 'gene_disease_validity', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Gene-disease validity', 'sub_label': None, 'description': 'The strength of the evidence supporting  a claim that variation in the gene under curation causes the condition under curation. <p> The ClinGen gene curation working group have developed a scoring system (see this <a href="https://search.clinicalgenome.org/kb/gene-validity/9972" target ="_blank">BRCA1 report </a> as an example) for determining gene-disease validity and their gene specific reports can  be found on the <a href = "https://search.clinicalgenome.org/kb/gene-validity" target ="_blank">gene-disease validity results</a> page. OMIM, GeneCards and MedGen are also excellent resources for determining gene-disease validity. Note that current ACMG guidelines are only applicable to variants in genes with an established gene-disease relationship. <p> Phenotype expansion: for genes where the phenotype or condition under curation is similar or overlapping with an established gene-disease relationship (intended for research use only).', 'examples': [], 'options': [{'key': 'no_reported_evidence', 'index': 0, 'label': ''}, {'key': 'disputed', 'index': 1, 'label': ''}, {'key': 'limited', 'index': 2, 'label': ''}, {'key': 'moderate', 'index': 3, 'label': ''}, {'key': 'strong', 'index': 4, 'label': ''}, {'key': 'definitive', 'index': 5, 'label': ''}, {'key': 'refuted', 'index': 6, 'label': ''}, {'key': 'phenotype_expansion', 'index': 7, 'label': ''}], 'see': None, 'evidence_category': 'H', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655539+00:00', 'modified': '2019-10-22 22:46:59.655547+00:00'}),
        EvidenceKey(**{'key': 'gene_omim_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 3, 'label': 'Gene OMIM ID', 'sub_label': None, 'description': 'The OMIM ID of the gene.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'omim_id', 'created': '2019-10-22 22:46:59.655595+00:00', 'modified': '2019-10-22 22:46:59.655603+00:00'}),
        EvidenceKey(**{'key': 'gene_penetrance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 6, 'label': None, 'sub_label': None, 'description': None, 'examples': [], 'options': [{'key': 'complete', 'index': 0, 'label': ''}, {'key': 'reduced', 'index': 1, 'label': ''}, {'key': 'low', 'index': 0, 'label': ''}], 'see': None, 'evidence_category': 'H', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655654+00:00', 'modified': '2019-10-22 22:46:59.655662+00:00'}),
        EvidenceKey(**{'key': 'gene_splicer', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'GeneSplicer', 'sub_label': 'Splicing', 'description': 'GeneSplicer score. <p> GeneSplicer is a system for detecting splice sites in the genomic DNA of various eukaryotes.', 'examples': [], 'options': [], 'see': 'https://ccb.jhu.edu/software/genesplicer/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655710+00:00', 'modified': '2019-10-22 22:46:59.655718+00:00'}),
        EvidenceKey(**{'key': 'gene_symbol', 'mandatory': False, 'max_share_level': 'public', 'order': 1, 'label': 'Gene symbol', 'sub_label': None, 'description': 'This is automatically set in the form via HGVS.', 'examples': [], 'options': [], 'see': 'https://www.genenames.org/', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655766+00:00', 'modified': '2019-10-22 22:46:59.655774+00:00'}),
        EvidenceKey(**{'key': 'genome_build', 'mandatory': True, 'max_share_level': 'public', 'order': 2, 'label': 'Genome Build', 'sub_label': None, 'description': 'Genome build for the variant, including patch number.', 'examples': ['hg19', 'GRCh37.p13', 'GRCh38'], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655822+00:00', 'modified': '2019-10-22 22:46:59.655830+00:00'}),
        EvidenceKey(**{'key': 'genotype_quality', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'Genotype Quality (GQ)', 'sub_label': None, 'description': 'GQ - Phred-scaled confidence that the genotype assignment (GT) is correct', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.655878+00:00', 'modified': '2019-10-22 22:46:59.655885+00:00'}),
        EvidenceKey(**{'key': 'gerp', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 15, 'label': 'GERP', 'sub_label': 'Conservation', 'description': 'Genomic Evolutionary Rate Profiling (GERP) score. <p> Score range: -12.3 to 6.17 <p> GERP identifies constrained elements in multiple alignments by quantifying substitution deficits. These deficits represent substitutions that would have occurred if the element were neutral DNA, but did not occur because the element has been under functional constraint. We refer to these deficits as "Rejected Substitutions". Rejected substitutions are a natural measure of constraint that reflects the strength of past purifying selection on the element.', 'examples': [], 'options': [], 'see': 'http://mendel.stanford.edu/SidowLab/downloads/gerp/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'gerp_pp_rs', 'created': '2019-10-22 22:46:59.655933+00:00', 'modified': '2019-10-22 22:46:59.655941+00:00'}),
        EvidenceKey(**{'key': 'gerp_pp', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 15, 'label': 'GERP++', 'sub_label': 'Conservation', 'description': "GERP++ (previously referred to as GERP2) score. <p> GERP++ consists of two programs, gerpcol and gerpelem. Gerpcol esimates constraint for each column of the alignment; gerpelem then identifies constrained elements from gerpcol's output.", 'examples': [], 'options': [], 'see': 'http://mendel.stanford.edu/SidowLab/downloads/gerp/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656064+00:00', 'modified': '2019-10-22 22:46:59.656075+00:00'}),
        EvidenceKey(**{'key': 'g_hgvs', 'mandatory': False, 'max_share_level': 'public', 'order': 5, 'label': 'g.HGVS', 'sub_label': None, 'description': '<p>HGVS Genomic coordinate, using (versioned) RefSeq contigs. <p>If multiple HGVS expressions are provided, they must all map to the same genomic coordinate.  <p>The <a href="http://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/landing" target="_blank">ClinGen Allele Registry converter</a> can be used to confirm HGVS nomenclature. See <a href="http://varnomen.hgvs.org/recommendations/DNA/variant/substitution/" target="_blank">here</a> for HGVS-nomenclature guidelines.', 'examples': ['NC_000011.9:g.103048382T>A'], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656127+00:00', 'modified': '2019-10-22 22:46:59.656135+00:00'}),
        EvidenceKey(**{'key': 'gnomad_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD', 'sub_label': None, 'description': 'The MAF found for the variant in gnomAD general population.', 'examples': [], 'options': [], 'see': 'http://gnomad.broadinstitute.org/', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'gnomad_af', 'created': '2019-10-22 22:46:59.656184+00:00', 'modified': '2019-10-22 22:46:59.656196+00:00'}),
        EvidenceKey(**{'key': 'gnomad_allele_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD allele count', 'sub_label': None, 'description': 'Total allele count (including both exomes and genomes) for the variant under curation in gnomAD.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656245+00:00', 'modified': '2019-10-22 22:46:59.656253+00:00'}),
        EvidenceKey(**{'key': 'gnomad_allele_number', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD allele number', 'sub_label': None, 'description': 'Total allele number (including both exomes and genomes) for the variant under curation in gnomAD.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-30 23:06:09.940905+00:00', 'modified': '2019-10-30 23:06:09.940914+00:00'}),
        EvidenceKey(**{'key': 'gnomad_hemi_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD hemizygotes count', 'sub_label': None, 'description': 'Total number of individuals hemizygous for alternate allele in gnomAD.', 'examples': None, 'options': None, 'see': None, 'evidence_category': 'P', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2020-06-19 06:51:26.298902+00:00', 'modified': '2020-06-19 06:51:26.298939+00:00'}),
        EvidenceKey(**{'key': 'gnomad_hom_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD homozygotes count', 'sub_label': None, 'description': 'Total number of homozygotes (including both exomes and genomes) for the variant under curation in gnomAD.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': 'gnomad_hom_alt', 'created': '2019-10-22 22:46:59.656301+00:00', 'modified': '2019-10-22 22:46:59.656309+00:00'}),
        EvidenceKey(**{'key': 'gnomad_oe_lof', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD LoF observed-expected score', 'sub_label': None, 'description': "Observed/expected (oe) score for the gene being 'loss-of-function' intolerant. Low oe values are indicative of strong intolerance.  <b>It is important to consider the confidence interval when using oe.</b> Please include the CI in brackets when entering data as per the example above. <p> From gnomad: A gene with an LoF oe of 0.4 can be interpreted as a gene where only 40% of the expected loss-of-function variants were observed and therefore is likely under selection against LoF variants. In addition, the 90% CI allows us to clearly distinguish cases where there is a lot of uncertainty about the constraint for that gene due to sample size... for interpretation of Mendelian cases, we suggest using the upper bound of the oe confidence interval < 0.35 if a hard threshold is needed.", 'examples': ['0.73 (0.59 - 0.92)'], 'options': [], 'see': None, 'evidence_category': 'H', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656366+00:00', 'modified': '2020-06-19 06:51:26.303200+00:00'}),
        EvidenceKey(**{'key': 'gnomad_popmax', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD highest pop', 'sub_label': None, 'description': 'Population with the highest gnomAD MAF.', 'examples': [], 'options': [{'key': 'AFR', 'index': 1, 'label': 'African/African American'}, {'key': 'AMR', 'index': 2, 'label': 'Latino'}, {'key': 'ASJ', 'index': 3, 'label': 'Ashkenazi Jewish'}, {'key': 'EAS', 'index': 4, 'label': 'East Asian'}, {'key': 'FIN', 'index': 5, 'label': 'Finnish'}, {'key': 'NFE', 'index': 6, 'label': 'Non-Finnish European'}, {'key': 'SAS', 'index': 7, 'label': 'South Asian'}, {'key': 'OTH', 'index': 8, 'label': 'Other'}], 'see': 'https://gnomad.broadinstitute.org/', 'evidence_category': 'P', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'gnomad_popmax', 'created': '2019-10-22 22:46:59.656423+00:00', 'modified': '2019-10-22 22:46:59.656432+00:00'}),
        EvidenceKey(**{'key': 'gnomad_popmax_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD highest MAF', 'sub_label': None, 'description': 'The highest MAF found for the variant in a sub-population in gnomAD.', 'examples': [], 'options': [], 'see': 'https://gnomad.broadinstitute.org/', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'gnomad_popmax_af', 'created': '2019-10-22 22:46:59.656483+00:00', 'modified': '2019-10-22 22:46:59.656491+00:00'}),
        EvidenceKey(**{'key': 'gnomad_popmax_allele_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD highest pop allele count', 'sub_label': None, 'description': 'Allele count for gnomAD sub-population with highest frequency (stored in gnomad_popmax)', 'examples': [42, 12345], 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-30 23:06:09.940964+00:00', 'modified': '2019-10-30 23:06:09.940972+00:00'}),
        EvidenceKey(**{'key': 'gnomad_popmax_allele_number', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'gnomAD highest pop allele number', 'sub_label': None, 'description': 'Allele number for gnomAD sub-population with highest frequency (stored in gnomad_popmax)', 'examples': [27250, 57309], 'options': [], 'see': None, 'evidence_category': 'P', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-30 23:06:09.941021+00:00', 'modified': '2019-10-30 23:06:09.941030+00:00'}),
        EvidenceKey(**{'key': 'grantham', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Grantham Score', 'sub_label': None, 'description': 'Grantham score. <p> Score range: 0 to 215 <p> Grantham score is a prediction of the effect of substitutions between amino acids based on chemical properties, including polarity and molecular volume.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'CP', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'grantham', 'created': '2019-10-22 22:46:59.656542+00:00', 'modified': '2019-10-22 22:46:59.656550+00:00'}),
        EvidenceKey(**{'key': 'gtr_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 1, 'label': 'GTR ID', 'sub_label': None, 'description': 'The Genetic Test Registry ID.', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/gtr/', 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656602+00:00', 'modified': '2019-10-22 22:46:59.656611+00:00'}),
        EvidenceKey(**{'key': 'heteroplasmy', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 13, 'label': None, 'sub_label': None, 'description': 'Percent heteroplasmy for mitochondrial variants.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656663+00:00', 'modified': '2019-10-22 22:46:59.656672+00:00'}),
        EvidenceKey(**{'key': 'hgmd', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'HGMD Accession Number', 'sub_label': None, 'description': 'HGMD (The Human Gene Mutation Database) Accession Number for the variant. <p> HGMD represents an attempt to collate all known (published) gene lesions responsible for human inherited disease.', 'examples': ['CM041678'], 'options': [], 'see': 'http://www.hgmd.cf.ac.uk/ac/index.php', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656721+00:00', 'modified': '2019-10-22 22:46:59.656729+00:00'}),
        EvidenceKey(**{'key': 'hgmd_evidence', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'HGMD Evidence', 'sub_label': None, 'description': 'Evidence described in HGMD. It is suggested that the PubMed ID is also included.', 'examples': [], 'options': [], 'see': 'http://www.hgmd.cf.ac.uk/ac/index.php', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656778+00:00', 'modified': '2019-10-22 22:46:59.656786+00:00'}),
        EvidenceKey(**{'key': 'hgmd_phenotype', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'HGMD Phenotype', 'sub_label': None, 'description': 'Phenotype specified in HGMD.', 'examples': [], 'options': [], 'see': 'http://www.hgmd.cf.ac.uk/ac/index.php', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656834+00:00', 'modified': '2019-10-22 22:46:59.656841+00:00'}),
        EvidenceKey(**{'key': 'hgnc_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 4, 'label': 'HGNC ID', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'hgnc_id', 'created': '2019-10-22 22:46:59.656892+00:00', 'modified': '2019-10-22 22:46:59.656901+00:00'}),
        EvidenceKey(**{'key': 'hmtdb', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'HmtDB', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': 'https://www.hmtdb.uniba.it/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.656953+00:00', 'modified': '2019-10-22 22:46:59.656961+00:00'}),
        EvidenceKey(**{'key': 'hrec', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 4, 'label': 'HREC number', 'sub_label': None, 'description': 'Human Resources Ethics Committee (HREC) approval number.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657009+00:00', 'modified': '2019-10-22 22:46:59.657017+00:00'}),
        EvidenceKey(**{'key': 'h_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'Summary', 'sub_label': None, 'description': "Gene summary. This field can be used to describe the evidence demonstrating the gene's role in the condition under curation or to further document the known mechanism(s) of disease, particularly when the gene has not been thoroughly curated.", 'examples': [], 'options': [], 'see': None, 'evidence_category': 'H', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657065+00:00', 'modified': '2019-10-22 22:46:59.657073+00:00'}),
        EvidenceKey(**{'key': 'human_splicing_finder', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'Human Splicing Finder', 'sub_label': 'Splicing', 'description': 'Human Splicing Finder (HSF) score. <p> Score range: 0 to 100 <p> HSF predicts the effects of mutations on splicing signals or to identify splicing motifs in any human sequence.', 'examples': [], 'options': [], 'see': 'http://www.umd.be/HSF3/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657124+00:00', 'modified': '2019-10-22 22:46:59.657132+00:00'}),
        EvidenceKey(**{'key': 'insert_size', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': None, 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657180+00:00', 'modified': '2019-10-22 22:46:59.657188+00:00'}),
        EvidenceKey(**{'key': 'internal_db_het_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'HET samples in lab database', 'sub_label': None, 'description': "Number of heterozygous samples in the lab's internal database", 'examples': [], 'options': [], 'see': None, 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'total_db_het', 'created': '2019-10-22 22:46:59.657236+00:00', 'modified': '2019-10-22 22:46:59.657244+00:00'}),
        EvidenceKey(**{'key': 'internal_db_hom_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'HOM samples in lab database', 'sub_label': None, 'description': "Number of homozygous samples in the lab's internal database", 'examples': [], 'options': [], 'see': None, 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'total_db_hom', 'created': '2019-10-22 22:46:59.657296+00:00', 'modified': '2019-10-22 22:46:59.657304+00:00'}),
        EvidenceKey(**{'key': 'internal_samples_100_percent_20x_gene_coverage', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Internal Samples with 100% of gene with 20x coverage', 'sub_label': None, 'description': "The number of internal samples from lab database with 100% coverage of the gene (greater than 20x). This can be used with internal_db_hom_count/internal_db_het_count to work out a rough estimate of a lab's population frequency without having to do joint-VCF calling on every sample.", 'examples': [], 'options': [], 'see': None, 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657363+00:00', 'modified': '2019-10-22 22:46:59.657372+00:00'}),
        EvidenceKey(**{'key': 'interpretation_summary', 'mandatory': False, 'max_share_level': 'public', 'order': 6, 'label': '', 'sub_label': '', 'description': 'A summary of the asserted interpretation of this variant.', 'examples': [], 'options': [], 'see': '', 'evidence_category': 'HI', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657421+00:00', 'modified': '2019-10-22 22:46:59.657429+00:00'}),
        EvidenceKey(**{'key': 'interpro_protein_domain', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'InterPro', 'sub_label': None, 'description': 'InterPro name or ID. <p> InterPro is a database of protein families, providing domain prediction. It aggregates predictive models, known as signatures, from several different databases.', 'examples': ['Zinc finger, PHD-finger', 'IPR019787'], 'options': [], 'see': 'https://www.ebi.ac.uk/interpro', 'evidence_category': 'F', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': 'interpro_domain', 'created': '2019-10-22 22:46:59.657477+00:00', 'modified': '2019-10-22 22:46:59.657485+00:00'}),
        EvidenceKey(**{'key': 'intervention', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 7, 'label': None, 'sub_label': None, 'description': 'Clinical intervention', 'examples': ['bilateral masectomy 42yr'], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657540+00:00', 'modified': '2019-10-22 22:46:59.657548+00:00'}),
        EvidenceKey(**{'key': 'intron', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 15, 'label': None, 'sub_label': None, 'description': 'Used to indicate which intron the variant is located in.', 'examples': ['3 of 25'], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'intron', 'created': '2019-10-22 22:46:59.657596+00:00', 'modified': '2019-10-22 22:46:59.657604+00:00'}),
        EvidenceKey(**{'key': 'in_vitro_assay', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'Free text summary of in vitro functional assay results. Include PubMed IDs if the evidence is derived from the literature.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'F', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657656+00:00', 'modified': '2019-10-22 22:46:59.657663+00:00'}),
        EvidenceKey(**{'key': 'in_vivo_assay', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'Free text summary of in vivo functional assay results. Include PubMed IDs if the evidence is derived from the literature.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'F', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657712+00:00', 'modified': '2019-10-22 22:46:59.657720+00:00'}),
        EvidenceKey(**{'key': 'literature', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 2, 'label': 'Literature', 'sub_label': '', 'description': 'Publications related to this variant and the interpretation of these findings. To store a link to the publication enter the Pubmed ID in the literature summary.', 'examples': ['This variant has been investigated in a mutational screen <a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=12846734" target="_blank">PMID: 12846734</a>.'], 'options': [], 'see': '', 'evidence_category': 'L', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': 'pubmed', 'created': '2019-10-22 22:46:59.657768+00:00', 'modified': '2019-10-22 22:46:59.657776+00:00'}),
        EvidenceKey(**{'key': 'loftool', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'LoFtool', 'sub_label': None, 'description': 'LoFtool provides a rank of genic intolerance and consequent susceptibility to disease based on the ratio of Loss-of-function (LoF) to synonymous mutations for each gene in 60,706 individuals from ExAC, adjusting for the gene de novo mutation rate and evolutionary protein conservation. The lower the LoFtool gene score percentile the most intolerant is the gene to functional variation. For more details please see (Fadista J et al. 2017), PMID:27563026.', 'examples': ['0.295'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pubmed/27563026', 'evidence_category': 'CP', 'value_type': 'L', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'loftool', 'created': '2019-10-22 22:46:59.657827+00:00', 'modified': '2019-10-22 22:46:59.657834+00:00'}),
        EvidenceKey(**{'key': 'lovd_database', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'LOVD Database ID', 'sub_label': None, 'description': 'Leiden Open Variation Database (LOVD) Database ID. LOVD is a flexible, freely available tool for gene-centred collection and display of DNA variations. LOVD 3.0 extends this idea to also provide patient-centred data storage and storage of NGS data, even of variants outside of genes.', 'examples': ['VD_000010'], 'options': [], 'see': 'http://www.lovd.nl/3.0/search', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657886+00:00', 'modified': '2019-10-22 22:46:59.657894+00:00'}),
        EvidenceKey(**{'key': 'lovd_effect', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'LOVD Effect', 'sub_label': None, 'description': 'LOVD Effect. The variant’s effect on a protein’s function. <p> Effect categories: "-" (Does not affect function), "-?" (Probably does not affect function), "?" (Effect unknown), "+?" (Probably affects function), "+" (Affects function).', 'examples': [], 'options': [], 'see': 'http://www.lovd.nl/3.0/search', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657942+00:00', 'modified': '2019-10-22 22:46:59.657950+00:00'}),
        EvidenceKey(**{'key': 'lrt', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'LRT', 'sub_label': None, 'description': 'Likelihood Ration Test (LRT) is a likelihood ratio test for significantly conserved amino acid positions within the human proteome.', 'examples': [], 'options': [], 'see': 'http://www.genetics.wustl.edu/jflab/lrt_query.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.657998+00:00', 'modified': '2019-10-22 22:46:59.658006+00:00'}),
        EvidenceKey(**{'key': 'mapp', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'MAPP', 'sub_label': None, 'description': 'Multivariate Analysis of Protein Polymorphism: MAPP MAPP considers the physicochemical variation present in a column of a protein sequence alignment and, on the basis of this variation, predicts the impact of all possible amino acid substitutions on the function of the protein.', 'examples': [], 'options': [], 'see': 'http://mendel.stanford.edu/SidowLab/downloads/MAPP/index.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658054+00:00', 'modified': '2019-10-22 22:46:59.658062+00:00'}),
        EvidenceKey(**{'key': 'match_maker_exchange', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 29, 'label': 'Matchmaker Exchange', 'sub_label': None, 'description': 'Has the variant been shared with the <a href="https://www.matchmakerexchange.org/#Supporters" target="_blank"> Matchmaker Exchange</a> network?  <p>The Matchmaker Exchange network is designed to facilitate the matching of cases with similar phenotypic and genotypic profiles (matchmaking). Variants are shared by entering them in software  available from one of the participating organisations (Matchmaker nodes) e.g. <a href="http://patientarchive.org/" traget="_blank">Patient Archive (Australian Genomics) </a>, DECIPHER, GeneMatcher, MyGene2, MatchBox (for Broad Institute Collaborators) etc.<p> Check your lab\'s policy on sharing before entering variants in any of these systems and make sure to avoid double entry where possible. <p> It is anticipated that Shariant users will be able to flag variants for sharing with the Matchmaker Exchange at variant submission in Phase 2 of Shariant development (2020).', 'examples': [], 'options': [{'key': 'y', 'index': 1, 'label': 'Yes'}, {'key': 'n', 'index': 2, 'label': 'No'}], 'see': None, 'evidence_category': 'SO', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658110+00:00', 'modified': '2019-10-22 22:46:59.658118+00:00'}),
        EvidenceKey(**{'key': 'maternal_inheritance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'Maternal inheritance', 'examples': [], 'options': [{'key': 'hom_ref', 'index': 1, 'label': ''}, {'key': 'hom_alt', 'index': 2, 'label': ''}, {'key': 'het', 'index': 3, 'label': ''}, {'key': 'hemi_alt', 'index': 4, 'label': ''}], 'see': None, 'evidence_category': 'D', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658166+00:00', 'modified': '2019-10-22 22:46:59.658173+00:00'}),
        EvidenceKey(**{'key': 'max_ent_scan', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'MaxEntScan', 'sub_label': 'Splicing', 'description': "MaxEntScan score. <p> MaxEntScan applies algorithms for derivation and scoring of constrained-marginal maximum entropy distributions, with applications to 5' and 3' splice sites.", 'examples': [], 'options': [], 'see': 'http://genes.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658221+00:00', 'modified': '2019-10-22 22:46:59.658229+00:00'}),
        EvidenceKey(**{'key': 'mechanism_of_disease', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 4, 'label': None, 'sub_label': None, 'description': 'What are the known mechanisms of disease associated with this condition? Assess literature, <a href="https://gnomad.broadinstitute.org/" target="_blank">gnomAD</a>, <a href="https://www.ncbi.nlm.nih.gov/books/NBK1116/" target="_blank">NCBI Gene Reviews</a>, <a href="https://www.omim.org/" target="_blank">OMIM</a>, etc. <br><br>  <b>LOF</b>: Loss of function variant. Can be missense, truncating etc. <b>GOF</b>: Gain of function variant. An overactive gene product or novel function for gene product. <b>DN</b>: Dominant negative. A variant where the mutated gene product adversely affects the other (wild type) gene product.', 'examples': [], 'options': [{'key': 'lof', 'index': 1, 'label': 'Loss of function'}, {'key': 'gof', 'index': 2, 'label': 'Gain of function'}, {'key': 'dn', 'index': 3, 'label': 'Dominant negative'}, {'key': 'gene_fusion', 'index': 5}, {'key': 'unknown', 'index': 4}], 'see': None, 'evidence_category': 'H', 'value_type': 'M', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658276+00:00', 'modified': '2020-06-30 07:38:42.839406+00:00'}),
        EvidenceKey(**{'key': 'median_read_length', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': None, 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658332+00:00', 'modified': '2019-10-22 22:46:59.658340+00:00'}),
        EvidenceKey(**{'key': 'metalr', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'Meta-LR', 'sub_label': 'Meta-predictor', 'description': 'MetaSVM is based on 10 component scores (SIFT, PolyPhen-2 HDIV, PolyPhen-2 HVAR, GERP++, MutationTaster, Mutation Assessor, FATHMM, LRT, SiPhy, PhyloP) and the maximum frequency observed in the 1000 genomes populations.', 'examples': [], 'options': [], 'see': 'https://sites.google.com/site/jpopgen/dbNSFP', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658388+00:00', 'modified': '2019-10-22 22:46:59.658396+00:00'}),
        EvidenceKey(**{'key': 'metasvm', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'Meta-SVM', 'sub_label': 'Meta-predictor', 'description': 'MetaSVM is based on 10 component scores (SIFT, PolyPhen-2 HDIV, PolyPhen-2 HVAR, GERP++, MutationTaster, Mutation Assessor, FATHMM, LRT, SiPhy, PhyloP) and the maximum frequency observed in the 1000 genomes populations.', 'examples': [], 'options': [], 'see': 'https://sites.google.com/site/jpopgen/dbNSFP', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658444+00:00', 'modified': '2019-10-22 22:46:59.658451+00:00'}),
        EvidenceKey(**{'key': 'mitomap', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Mitomap Frequency', 'sub_label': None, 'description': 'Mitomap Overall Frequency (%). <p> MitoMap is a human mitochondrial genome database.', 'examples': [''], 'options': [], 'see': 'https://www.mitomap.org', 'evidence_category': 'P', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658500+00:00', 'modified': '2019-10-22 22:46:59.658507+00:00'}),
        EvidenceKey(**{'key': 'mode_of_inheritance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': '', 'sub_label': '', 'description': 'The known mode of inheritance of the condition tested as listed in <a href="https://www.omim.org/" target="_blank">OMIM</a> or other reputable source.', 'examples': [], 'options': [{'key': 'autosomal_dominant', 'index': 1, 'label': ''}, {'key': 'autosomal_recessive', 'index': 2, 'label': ''}, {'key': 'autosomal_unknown', 'index': 3, 'label': ''}, {'key': 'sex_limited_autosomal_dominant', 'index': 4, 'label': ''}, {'key': 'x_linked', 'index': 5, 'label': ''}, {'key': 'x_linked_recessive', 'index': 6, 'label': ''}, {'key': 'x_linked_dominant', 'index': 7, 'label': ''}, {'key': 'y_linked', 'index': 8, 'label': ''}, {'key': 'mitochondrial', 'index': 9, 'label': ''}, {'key': 'oligogenic', 'index': 10, 'label': ''}, {'key': 'somatic', 'index': 11, 'label': ''}, {'key': 'unknown', 'index': 12, 'label': ''}, {'key': 'other', 'index': 13, 'label': ''}, {'key': 'digenic_recessive', 'index': 14, 'label': ''}, {'key': 'digenic_dominant', 'index': 15, 'label': ''}, {'key': 'isolated_cases', 'index': 16, 'label': ''}, {'key': 'multifactorial', 'index': 17, 'label': ''}, {'key': 'imprinting', 'index': 18, 'label': ''}], 'see': '', 'evidence_category': 'H', 'value_type': 'M', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658555+00:00', 'modified': '2019-10-22 22:46:59.658563+00:00'}),
        EvidenceKey(**{'key': 'molecular_consequence', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': '', 'sub_label': '', 'description': 'The predicted effect of a variant on a biological sequence as described by Sequence Ontology.', 'examples': [], 'options': [{'so': 'SO:0001624', 'key': '3_prime_utr_variant', 'index': 1, 'label': '3 prime UTR variant'}, {'so': 'SO:0001623', 'key': '5_prime_utr_variant', 'index': 2, 'label': '5 prime UTR variant'}, {'so': 'SO:0001580', 'key': 'coding_sequence_variant', 'index': 3, 'label': 'Coding sequence variant'}, {'so': 'SO:0001632', 'key': 'downstream_gene_variant', 'index': 4, 'label': 'Downstream gene variant'}, {'so': 'SO:0001907', 'key': 'feature_elongation', 'index': 5, 'label': 'Feature elongation'}, {'so': 'SO:0001906', 'key': 'feature_truncation', 'index': 6, 'label': 'Feature truncation'}, {'so': 'SO:0001589', 'key': 'frameshift_variant', 'index': 7, 'label': 'Frameshift variant'}, {'so': 'SO:0001626', 'key': 'incomplete_terminal_codon_variant', 'index': 8, 'label': 'Incomplete terminal codon variant'}, {'so': 'SO:0001822', 'key': 'inframe_deletion', 'index': 9, 'label': 'Inframe deletion'}, {'so': 'SO:0001821', 'key': 'inframe_insertion', 'index': 10, 'label': 'Inframe insertion'}, {'so': 'SO:0001628', 'key': 'intergenic_variant', 'index': 11, 'label': 'Intergenic variant'}, {'so': 'SO:0001627', 'key': 'intron_variant', 'index': 12, 'label': 'Intron variant'}, {'so': 'SO:0001620', 'key': 'mature_mirna_variant', 'index': 13, 'label': 'Mature miRNA variant'}, {'so': 'SO:0001583', 'key': 'missense_variant', 'index': 14, 'label': 'Missense variant'}, {'so': 'SO:0001621', 'key': 'nmd_transcript_variant', 'index': 15, 'label': 'NMD transcript variant'}, {'so': 'SO:0001792', 'key': 'non_coding_transcript_exon_variant', 'index': 16, 'label': 'Non coding transcript exon variant'}, {'so': 'SO:0001619', 'key': 'non_coding_transcript_variant', 'index': 17, 'label': 'Non coding transcript variant'}, {'so': 'SO:0001818', 'key': 'protein_altering_variant', 'index': 18, 'label': 'Protein altering variant'}, {'so': 'SO:0001894', 'key': 'regulatory_region_ablation', 'index': 19, 'label': 'Regulatory region ablation'}, {'so': 'SO:0001891', 'key': 'regulatory_region_amplification', 'index': 20, 'label': 'Regulatory region amplification'}, {'so': 'SO:0001566', 'key': 'regulatory_region_variant', 'index': 21, 'label': 'Regulatory region variant'}, {'so': 'SO:0001574', 'key': 'splice_acceptor_variant', 'index': 22, 'label': 'Splice acceptor variant'}, {'so': 'SO:0001575', 'key': 'splice_donor_variant', 'index': 23, 'label': 'Splice donor variant'}, {'so': 'SO:0001630', 'key': 'splice_region_variant', 'index': 24, 'label': 'Splice region variant'}, {'so': 'SO:0002012', 'key': 'start_lost', 'index': 25, 'label': 'Start lost'}, {'so': 'SO:0002019', 'key': 'start_retained_variant', 'index': 26, 'label': 'Start retained variant'}, {'so': 'SO:0001587', 'key': 'stop_gained', 'index': 27, 'label': 'Stop gained'}, {'so': 'SO:0001578', 'key': 'stop_lost', 'index': 28, 'label': 'Stop lost'}, {'so': 'SO:0001567', 'key': 'stop_retained_variant', 'index': 29, 'label': 'Stop retained variant'}, {'so': 'SO:0001819', 'key': 'synonymous_variant', 'index': 30, 'label': 'Synonymous variant'}, {'so': 'SO:0001782', 'key': 'tf_binding_site_variant', 'index': 31, 'label': 'TF binding site variant'}, {'so': 'SO:0001895', 'key': 'tfbs_ablation', 'index': 32, 'label': 'TFBS ablation'}, {'so': 'SO:0001892', 'key': 'tfbs_amplification', 'index': 33, 'label': 'TFBS amplification'}, {'so': 'SO:0001893', 'key': 'transcript_ablation', 'index': 34, 'label': 'Transcript ablation'}, {'so': 'SO:0001889', 'key': 'transcript_amplification', 'index': 35, 'label': 'Transcript amplification'}, {'so': 'SO:0001631', 'key': 'upstream_gene_variant', 'index': 36, 'label': 'Upstream gene variant'}], 'see': 'http://www.sequenceontology.org/', 'evidence_category': 'V', 'value_type': 'M', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'consequence', 'created': '2019-10-22 22:46:59.658611+00:00', 'modified': '2019-10-22 22:46:59.658619+00:00'}),
        EvidenceKey(**{'key': 'mutational_hotspot', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 16, 'label': None, 'sub_label': None, 'description': 'A region or protein domain with a high deleterious mutation rate. Detail the number of pathogenic vs total classifications and the source of the data.', 'examples': ['Hot-spot region (+/- 50bp) 19 pathogenic, 0 benign, 26 total classifications = 73.1% pathogenic (ClinVar)', "Interpro protein domain 'Zinc finger - B-box' 90 pathogenic, 0 benign, 100 total classifications = 90% pathogenic (Uniprot)"], 'options': [], 'see': None, 'evidence_category': 'F', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658673+00:00', 'modified': '2019-10-22 22:46:59.658681+00:00'}),
        EvidenceKey(**{'key': 'mutation_assessor', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Mutation Assessor', 'sub_label': None, 'description': 'This server predicts the functional impact of amino-acid substitutions in proteins, such as mutations discovered in cancer or missense polymorphisms. The functional impact is assessed based on evolutionary conservation of the affected amino acid in protein homologs. The method has been validated on a large set (60k) of disease associated (OMIM) and polymorphic variants. To explore the functional impact of missense mutations found in The Cancer Genome Atlas please use cBioPortal for Cancer Genomics.', 'examples': [], 'options': [], 'see': 'http://mutationassessor.org', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'mutation_assessor_pred_most_damaging', 'created': '2019-10-22 22:46:59.658735+00:00', 'modified': '2019-10-22 22:46:59.658743+00:00'}),
        EvidenceKey(**{'key': 'mutation_taster', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'MutationTaster', 'sub_label': None, 'description': 'MutationTaster score or prediction. <p> Score range: 0 to 1 <p> Prediction categories: "Disease causing", "Disease causing automatic", "Polymorphism", "Polymorphism automatic". <p> MutationTaster is a free web-based application to evaluate DNA sequence variants for their disease-causing potential. The software performs a battery of in silico tests to estimate the impact of the variant on the gene product / protein. Tests are made on both, protein and DNA level, MutationTaster is hence not limited to substitutions of single amino acids but can also handle synonymous or intronic variants.', 'examples': [], 'options': [], 'see': 'http://www.mutationtaster.org/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'mutation_taster_pred_most_damaging', 'created': '2019-10-22 22:46:59.658795+00:00', 'modified': '2019-10-22 22:46:59.658803+00:00'}),
        EvidenceKey(**{'key': 'mutpred', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'MutPred', 'sub_label': None, 'description': 'MutPred2 score. <p> Score range: 0 to 1 <p> MutPred is a web application tool developed to classify an amino acid substitution as disease-associated or neutral in human.', 'examples': [], 'options': [], 'see': 'http://mutpred.mutdb.org', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658854+00:00', 'modified': '2019-10-22 22:46:59.658861+00:00'}),
        EvidenceKey(**{'key': 'mutpred_lof', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'MutPred-LOF', 'sub_label': None, 'description': 'MutPred-Splice score. <p> Score range: 0 to 1 <p>MutPred-LOF is a machine learning-based method and software package that integrates genetic and molecular data to reason probabilistically about the pathogenicity of frameshifting indels and stop gain variants.', 'examples': [], 'options': [], 'see': 'http://mutpredlof.cs.indiana.edu/index.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658909+00:00', 'modified': '2019-10-22 22:46:59.658917+00:00'}),
        EvidenceKey(**{'key': 'mutpred_splice', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'MutPred-Splice', 'sub_label': 'Splicing', 'description': 'MutPred-Splice score. <p> Score range: 0 to 1 <p> MutPred-Splice is a random forest-based approach to prioritize exonic variants (missense or samesense) which are likely to disrupt pre-mRNA splicing from whole-genome sequencing data sets.', 'examples': [], 'options': [], 'see': 'http://www.mutdb.org/mutpredsplice/submit.htm', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.658965+00:00', 'modified': '2019-10-22 22:46:59.658972+00:00'}),
        EvidenceKey(**{'key': 'netgene2', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'NetGene2', 'sub_label': 'Splicing', 'description': 'NetGene2 score. <p> NetGene2 is a service producing neural network predictions of splice sites in human, C. elegans and A. thaliana DNA.', 'examples': [], 'options': [], 'see': 'http://www.cbs.dtu.dk/services/NetGene2/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659020+00:00', 'modified': '2019-10-22 22:46:59.659028+00:00'}),
        EvidenceKey(**{'key': 'nn_splice', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'NNSplice', 'sub_label': 'Splicing', 'description': 'NNSplice score. <p> Score range: 0 to 1 <p> NNSplice analyses the structure of donor and acceptor sites using a separate neural network recogniser for each site.', 'examples': [], 'options': [], 'see': 'http://www.fruitfly.org/seq_tools/splice.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659076+00:00', 'modified': '2019-10-22 22:46:59.659084+00:00'}),
        EvidenceKey(**{'key': 'nssnpanalyzer', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'nsSNPAnalyzer', 'sub_label': None, 'description': 'nsSNPAnalyzer is a tool to predict whether a nonsynonymous single nucleotide polymorphism (nsSNP) has a phenotypic effect.', 'examples': [], 'options': [], 'see': 'http://snpanalyzer.uthsc.edu/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659134+00:00', 'modified': '2019-10-22 22:46:59.659142+00:00'}),
        EvidenceKey(**{'key': 'nucleic_acid_source', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Nucleic Acid Source', 'sub_label': None, 'description': 'Type of nucleic acid sequenced.', 'examples': [], 'options': [{'key': 'DNA', 'index': 0, 'label': 'DNA'}, {'key': 'RNA', 'index': 1, 'label': 'RNA'}], 'see': None, 'evidence_category': 'HT', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659190+00:00', 'modified': '2019-10-22 22:46:59.659198+00:00'}),
        EvidenceKey(**{'key': 'o_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'O', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659246+00:00', 'modified': '2019-10-22 22:46:59.659254+00:00'}),
        EvidenceKey(**{'key': 'o_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'O', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659300+00:00', 'modified': '2019-10-22 22:46:59.659307+00:00'}),
        EvidenceKey(**{'key': 'owner', 'mandatory': False, 'max_share_level': 'institution', 'order': 0, 'label': None, 'sub_label': None, 'description': "The individual who is primarily responsible for this record. Enter another user's username or email address to assign them responsibility of this record.", 'examples': [], 'options': [], 'see': None, 'evidence_category': 'SO', 'value_type': 'U', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659352+00:00', 'modified': '2019-10-22 22:46:59.659360+00:00'}),
        EvidenceKey(**{'key': 'panther', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'PANTHER', 'sub_label': None, 'description': 'PANTHER estimates the likelihood of a particular nonsynonymous (amino-acid changing) coding SNP to cause a functional impact on the protein. It calculates the length of time (in millions of years) a given amino acid has been preserved in the lineage leading to the protein of interest.', 'examples': [], 'options': [], 'see': 'http://www.pantherdb.org/tools/csnpScoreForm.jsp', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659408+00:00', 'modified': '2019-10-22 22:46:59.659415+00:00'}),
        EvidenceKey(**{'key': 'paternal_inheritance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'Paternal inheritance', 'examples': [], 'options': [{'key': 'hom_alt', 'index': 2, 'label': ''}, {'key': 'hom_ref', 'index': 1, 'label': ''}, {'key': 'het', 'index': 3, 'label': ''}, {'key': 'hemi_alt', 'index': 4, 'label': ''}, {'key': 'hemi_ref', 'index': 5, 'label': ''}], 'see': None, 'evidence_category': 'D', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659463+00:00', 'modified': '2019-10-22 22:46:59.659471+00:00'}),
        EvidenceKey(**{'key': 'patient_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': -10, 'label': 'Patient ID', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659519+00:00', 'modified': '2019-10-22 22:46:59.659526+00:00'}),
        EvidenceKey(**{'key': 'patient_phenotype', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': None, 'sub_label': None, 'description': 'The observable traits of the patient. <br><br> Input HPO reference numbers and/or free text.', 'examples': ['HP:0004322'], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659574+00:00', 'modified': '2020-02-13 07:19:37.531366+00:00'}),
        EvidenceKey(**{'key': 'patient_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 100, 'label': None, 'sub_label': None, 'description': 'This field should not be used for Shariant.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659630+00:00', 'modified': '2019-10-22 22:46:59.659638+00:00'}),
        EvidenceKey(**{'key': 'p_conservation_assertion', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 16, 'label': 'Protein conservation assertion', 'sub_label': 'Conservation', 'description': 'Assertion of conservation of the amino acid across species. Can be determined using <a href="https://genome.ucsc.edu/" target="_blank">UCSC Genome Browser</a> and in silico tools. Conservation is of low relevance for truncating variants.  See equivalent field Nucleotide conservation assertion for nucleotide conservation.', 'examples': [], 'options': [{'key': 'very_high', 'index': 1, 'label': ''}, {'key': 'high', 'index': 2, 'label': ''}, {'key': 'moderate', 'index': 3, 'label': ''}, {'key': 'low', 'index': 4, 'label': ''}], 'see': None, 'evidence_category': 'CP', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659686+00:00', 'modified': '2019-10-22 22:46:59.659693+00:00'}),
        EvidenceKey(**{'key': 'pfam_protein_domain', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'Pfam', 'sub_label': None, 'description': 'Pfam accession, ID or description. <p> Pfam is a database of protein families, each represented by multiple sequence alignments and hidden Markov models.', 'examples': ['Tubulin-tyrosine ligase family', 'TTL', 'PF03133'], 'options': [], 'see': 'https://pfam.xfam.org', 'evidence_category': 'F', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': 'domains', 'created': '2019-10-22 22:46:59.659741+00:00', 'modified': '2019-10-22 22:46:59.659749+00:00'}),
        EvidenceKey(**{'key': 'pharmgkb', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 25, 'label': 'PharmGKB', 'sub_label': None, 'description': 'Pharmacogenomics knowledgebase including clinical guidelines, drug labels, gene-drug associations and genotype-phenotype relationships.', 'examples': [], 'options': [], 'see': 'https://www.pharmgkb.org', 'evidence_category': 'DB', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659803+00:00', 'modified': '2019-10-22 22:46:59.659811+00:00'}),
        EvidenceKey(**{'key': 'phastcons', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 15, 'label': 'phastCons', 'sub_label': 'Conservation', 'description': 'phastCons score. <p> Score range: 0 to 1 <p> phastCons is a program for identifying evolutionarily conserved elements in a multiple sequence alignment based on a phylogenetic hidden Markov model (phylo-HMM) when given a phylogenetic tree.', 'examples': [], 'options': [], 'see': 'http://compgen.bscb.cornell.edu/phast/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659860+00:00', 'modified': '2019-10-22 22:46:59.659867+00:00'}),
        EvidenceKey(**{'key': 'phd_snp', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'PhD-SNP', 'sub_label': None, 'description': 'Predictor of human Deleterious Single Nucleotide Polymorphisms (PhD-SNP)', 'examples': [], 'options': [], 'see': 'http://snps.biofold.org/phd-snp/phd-snp.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.659914+00:00', 'modified': '2019-10-22 22:46:59.659922+00:00'}),
        EvidenceKey(**{'key': 'p_hgvs', 'mandatory': False, 'max_share_level': 'public', 'order': 7, 'label': 'p.HGVS', 'sub_label': None, 'description': 'HGVS protein change.  <p> If multiple HGVS expressions are provided, they must all match to the same genomic coordinate.  <p>  <p>The <a href="http://reg.clinicalgenome.org/redmine/projects/registry/gene_to_hgvs" target="_blank">ClinGen Allele Registry converter</a> can be used to confirm HGVS nomenclature.<br> See <a href="http://varnomen.hgvs.org/recommendations/protein/variant/substitution/" target="_blank">here</a> for HGVS-nomenclature guidelines.', 'examples': ['p.Met268Thr'], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'hgvs_p', 'created': '2019-10-22 22:46:59.659970+00:00', 'modified': '2019-10-22 22:46:59.659978+00:00'}),
        EvidenceKey(**{'key': 'phred_likelihood', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'Phred Likelihood', 'sub_label': None, 'description': 'Normalized Phred-scaled likelihoods of the possible genotypes.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660029+00:00', 'modified': '2019-10-22 22:46:59.660037+00:00'}),
        EvidenceKey(**{'key': 'phylop', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 15, 'label': 'PhyloP', 'sub_label': 'Conservation', 'description': 'PhyloP score. <p> PhyloP measures evolutionary conservation at individual alignment sites. The scores are interpreted as follows compared to the evolution expected under neutral drift.', 'examples': [], 'options': [], 'see': 'http://compgen.cshl.edu/phast/background.php', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660085+00:00', 'modified': '2019-10-22 22:46:59.660093+00:00'}),
        EvidenceKey(**{'key': 'pm1', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PM1', 'sub_label': 'In hot spot or functional domain', 'description': 'Located in a mutational hot spot and/or critical and well-established functional domain (e.g., active site of an enzyme) without benign variation', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'F', 'value_type': 'C', 'default_crit_evaluation': 'PM', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660140+00:00', 'modified': '2019-10-22 22:46:59.660148+00:00'}),
        EvidenceKey(**{'key': 'pm2', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PM2', 'sub_label': 'Absent in population databases', 'description': 'Absent from controls (or at extremely low frequency if recessive) in Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation Consortium', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'P', 'value_type': 'C', 'default_crit_evaluation': 'PM', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660196+00:00', 'modified': '2019-10-22 22:46:59.660204+00:00'}),
        EvidenceKey(**{'key': 'pm3', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PM3', 'sub_label': 'Trans (recessive)', 'description': 'For recessive disorders, detected in trans with a pathogenic variant', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'A', 'value_type': 'C', 'default_crit_evaluation': 'PM', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660255+00:00', 'modified': '2019-10-22 22:46:59.660263+00:00'}),
        EvidenceKey(**{'key': 'pm4', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 34, 'label': 'PM4', 'sub_label': 'Indel w/ protein length change', 'description': 'Protein length changes as a result of in-frame deletions/insertions in a nonrepeat region or stop-loss variants', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'PM', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660312+00:00', 'modified': '2019-10-22 22:46:59.660320+00:00'}),
        EvidenceKey(**{'key': 'pm5', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 36, 'label': 'PM5', 'sub_label': 'Novel missense in same codon as known missense', 'description': 'Novel missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'PM', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660368+00:00', 'modified': '2019-10-22 22:46:59.660376+00:00'}),
        EvidenceKey(**{'key': 'pm6', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PM6', 'sub_label': 'Assumed de novo', 'description': 'Assumed de novo, but without confirmation of paternity and maternity', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'D', 'value_type': 'C', 'default_crit_evaluation': 'PM', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660426+00:00', 'modified': '2019-10-22 22:46:59.660434+00:00'}),
        EvidenceKey(**{'key': 'pmut_prediction', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'PMut', 'sub_label': 'Meta-predictor', 'description': 'PMUT is a Neural network based pathogenicity predictor for protein-coding variants.  <p>See <a href="https://www.ncbi.nlm.nih.gov/pubmed/28453649" target="_blank">"PMut: a web-based tool for the annotation of pathological variants on proteins, 2017 update"</a> for details.', 'examples': ['0.72 (86%) Disease', '0.10 (96%) Neutral'], 'options': [], 'see': 'http://mmb.pcb.ub.es/PMut/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660482+00:00', 'modified': '2019-10-22 22:46:59.660489+00:00'}),
        EvidenceKey(**{'key': 'polyphen', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Poly-Phen', 'sub_label': None, 'description': 'PolyPhen is a tool that predicts possible impact of an amino acid substitution on the structure and function of a human protein using physical and comparative considerations.  Structural features such as amino acid atomic contacts and solvent accessibility are also assessed and empirically determined cut-offs used to predict if the substitution is ‘probably damaging’, ‘possibly damaging’ or ‘benign’.  ????Think this as been superseded by PolyPhen-2???? Webpage auto re-directs to PolyPhen-2.', 'examples': [], 'options': [], 'see': 'http://genetics.bwh.harvard.edu/pph/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660537+00:00', 'modified': '2019-10-22 22:46:59.660545+00:00'}),
        EvidenceKey(**{'key': 'polyphen2', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Poly-Phen2', 'sub_label': None, 'description': 'PolyPhen-2 (Polymorphism Phenotyping v2) score or prediction. <p> Score range: 0 to 1 <p> Prediction categories: "Probably damaging", "Possibly damaging", "Benign" <p> PolyPhen is a tool that predicts possible impact of an amino acid substitution on the structure and function of a human protein using physical and comparative considerations.', 'examples': [], 'options': [], 'see': 'http://genetics.bwh.harvard.edu/pph2', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'polyphen2_hvar_pred_most_damaging', 'created': '2019-10-22 22:46:59.660593+00:00', 'modified': '2019-10-22 22:46:59.660601+00:00'}),
        EvidenceKey(**{'key': 'p_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'P', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660653+00:00', 'modified': '2019-10-22 22:46:59.660660+00:00'}),
        EvidenceKey(**{'key': 'pp1', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PP1', 'sub_label': 'Co-segregation with disease', 'description': 'Cosegregation with disease in multiple affected family members in a gene definitively known to cause the disease', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'S', 'value_type': 'C', 'default_crit_evaluation': 'PP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660706+00:00', 'modified': '2019-10-22 22:46:59.660713+00:00'}),
        EvidenceKey(**{'key': 'pp2', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PP2', 'sub_label': 'Missense, where missense common cause of disease', 'description': 'Missense variant in a gene that has a low rate of benign missense variation and in which missense variants are a common mechanism of disease', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'F', 'value_type': 'C', 'default_crit_evaluation': 'PP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660761+00:00', 'modified': '2019-10-22 22:46:59.660770+00:00'}),
        EvidenceKey(**{'key': 'pp3', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 31, 'label': 'PP3', 'sub_label': 'Deleterious effect predicted', 'description': 'Multiple lines of computational evidence support a deleterious effect on the gene or gene product (conservation, evolutionary, splicing impact, etc.)', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'PP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660821+00:00', 'modified': '2019-10-22 22:46:59.660829+00:00'}),
        EvidenceKey(**{'key': 'pp4', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PP4', 'sub_label': 'Phenotype or FHx highly specific', 'description': 'Patient’s phenotype or family history is highly specific for a disease with a single genetic etiology', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'O', 'value_type': 'C', 'default_crit_evaluation': 'PP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660877+00:00', 'modified': '2019-10-22 22:46:59.660885+00:00'}),
        EvidenceKey(**{'key': 'pp5', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PP5', 'sub_label': 'Pathogenic, reputable source', 'description': 'Reputable source recently reports variant as pathogenic, but the evidence is not available to the laboratory to perform an independent evaluation', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'DB', 'value_type': 'C', 'default_crit_evaluation': 'PP', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660935+00:00', 'modified': '2019-10-22 22:46:59.660943+00:00'}),
        EvidenceKey(**{'key': 'proband_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'Proband Count', 'sub_label': None, 'description': 'The aggregate number of all reported probands (including ClinVar, case reports, internal database counts etc.) <b> meeting the phenotypic criteria </b> for the gene and disorder being curated. For variants that are absent or rare in large population studies, the number of probands can be used to determine how to apply the PS4 criterion. Probands should be from unrelated families.', 'examples': None, 'options': None, 'see': None, 'evidence_category': 'P', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2020-06-19 06:51:26.309069+00:00', 'modified': '2020-06-19 06:51:26.309091+00:00'}),
        EvidenceKey(**{'key': 'provean', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'PROVEAN', 'sub_label': None, 'description': 'PROVEAN (Protein Variation Effect Analyzer) score or prediction. <p> Score range: -14 to 14 <p> Prediction categories: "Deleterious", "Neutral" <p> PROVEAN is a software tool which predicts whether an amino acid substitution or indel has an impact on the biological function of a protein.', 'examples': [], 'options': [], 'see': 'http://provean.jcvi.org/index.php', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.660992+00:00', 'modified': '2019-10-22 22:46:59.660999+00:00'}),
        EvidenceKey(**{'key': 'ps1', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 38, 'label': 'PS1', 'sub_label': 'Same aa change as established pathogenic variant', 'description': 'Same amino acid change as a previously established pathogenic variant regardless of nucleotide change', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'PS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661048+00:00', 'modified': '2019-10-22 22:46:59.661055+00:00'}),
        EvidenceKey(**{'key': 'ps2', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PS2', 'sub_label': 'De novo, maternity & paternity confirmed', 'description': 'De novo (both maternity and paternity confirmed) in a patient with the disease and no family history', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'D', 'value_type': 'C', 'default_crit_evaluation': 'PS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661103+00:00', 'modified': '2019-10-22 22:46:59.661111+00:00'}),
        EvidenceKey(**{'key': 'ps3', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PS3', 'sub_label': 'Effect in functional assay', 'description': 'Well-established in vitro or in vivo functional studies supportive of a damaging effect on the gene or gene product', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'F', 'value_type': 'C', 'default_crit_evaluation': 'PS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661159+00:00', 'modified': '2019-10-22 22:46:59.661167+00:00'}),
        EvidenceKey(**{'key': 'ps4', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'PS4', 'sub_label': 'Prevalence in affecteds increased over controls', 'description': 'The prevalence of the variant in affected individuals is significantly increased compared with the prevalence in controls', 'examples': [], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/', 'evidence_category': 'P', 'value_type': 'C', 'default_crit_evaluation': 'PS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661214+00:00', 'modified': '2019-10-22 22:46:59.661222+00:00'}),
        EvidenceKey(**{'key': 'p_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'P', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661270+00:00', 'modified': '2019-10-22 22:46:59.661277+00:00'}),
        EvidenceKey(**{'key': 'pubmed_gene_search_count', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 2, 'label': 'PubMed Gene count', 'sub_label': '', 'description': 'Number of PubMed results for gene symbol.', 'examples': [42], 'options': [], 'see': '', 'evidence_category': 'L', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-11-06 22:38:13.756896+00:00', 'modified': '2019-11-06 22:38:13.756929+00:00'}),
        EvidenceKey(**{'key': 'pvs1', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'PVS1', 'sub_label': 'Null variant, where LOF known mechanism', 'description': 'Null variant (nonsense, frameshift, canonical ±1 or 2 splice sites, initiation codon, single or multiexon deletion) in a gene where LOF is a known mechanism of disease', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'CP', 'value_type': 'C', 'default_crit_evaluation': 'PVS', 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661322+00:00', 'modified': '2019-10-22 22:46:59.661330+00:00'}),
        EvidenceKey(**{'key': 'read_depth', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 8, 'label': 'Read depth (DP)', 'sub_label': None, 'description': 'Depth of coverage. Total number of high quality filtered reads for all alleles.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661384+00:00', 'modified': '2019-10-22 22:46:59.661392+00:00'}),
        EvidenceKey(**{'key': 'read_qual_score', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 9, 'label': 'Read Quality Score', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661441+00:00', 'modified': '2019-10-22 22:46:59.661448+00:00'}),
        EvidenceKey(**{'key': 'redcap_record_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'REDcap record ID', 'sub_label': None, 'description': 'When using the REDcap data export, all records with the same redcap_record_id will be combined into a single row. This works across organisations and labs.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661496+00:00', 'modified': '2019-10-22 22:46:59.661504+00:00'}),
        EvidenceKey(**{'key': 'reference_depth', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Reference Read Depth', 'sub_label': None, 'description': 'Number of reads for the reference allele', 'examples': ['42'], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661553+00:00', 'modified': '2019-10-22 22:46:59.661561+00:00'}),
        EvidenceKey(**{'key': 'refseq_protein_id', 'mandatory': False, 'max_share_level': 'public', 'order': 4, 'label': 'RefSeq Protein ID', 'sub_label': '', 'description': 'RefSeq identifier for a protein.', 'examples': ['NP_005036.2'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/protein', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661609+00:00', 'modified': '2019-10-22 22:46:59.661617+00:00'}),
        EvidenceKey(**{'key': 'refseq_transcript_id', 'mandatory': False, 'max_share_level': 'public', 'order': 3, 'label': 'RefSeq Transcript ID', 'sub_label': None, 'description': 'RefSeq transcript identifier. <p> Be careful in the selection of the transcript for curation. For assistance the MANE <a href="https://www.slideshare.net/GenomeRef/mane-v2-final" target="_blank"> (Matched Annotation by NCBI and EMBL-EBI) </a>project is in the process of documenting a recommended representative transcript for each gene. Where available these transcripts are annotated with an orange "MANE preferred" tag in the ClinGen Allele registry e.g. <a href="http://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/by_caid?caid=CA386612638" target="_blank">CA386612638 </a><p>', 'examples': ['NM_001267550.1'], 'options': [], 'see': 'https://www.ncbi.nlm.nih.gov/nuccore', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661665+00:00', 'modified': '2019-10-22 22:46:59.661672+00:00'}),
        EvidenceKey(**{'key': 'regional_missense_constraint', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': None, 'sub_label': None, 'description': 'Is there evidence that the variant occurs in a functionally constrained region? (see Gnomad and Decipher)', 'examples': [], 'options': [{'key': 'constrained', 'index': 1, 'label': ''}, {'key': 'tolerant', 'index': 2, 'label': ''}], 'see': None, 'evidence_category': 'F', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661720+00:00', 'modified': '2019-10-22 22:46:59.661728+00:00'}),
        EvidenceKey(**{'key': 'repeat_masker', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 16, 'label': None, 'sub_label': None, 'description': 'RepeatMasker is a program that screens DNA sequences for interspersed repeats and low complexity DNA sequences', 'examples': ['(CGGGG)n', 'AluSp'], 'options': [], 'see': 'http://www.repeatmasker.org/', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'repeat_masker', 'created': '2019-10-22 22:46:59.661776+00:00', 'modified': '2019-10-22 22:46:59.661784+00:00'}),
        EvidenceKey(**{'key': 'report_date', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': None, 'sub_label': None, 'description': "Date that the most recent report was completed.<p> Click on Compare 'shared versions' in the Share status box to see prior reporting history (if available).", 'examples': [], 'options': [], 'see': None, 'evidence_category': 'SO', 'value_type': 'D', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661836+00:00', 'modified': '2019-10-22 22:46:59.661844+00:00'}),
        EvidenceKey(**{'key': 'report_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 16, 'label': 'Report ID', 'sub_label': None, 'description': "ID of most recent report generated for this variant. <p> Click on Compare 'shared versions' in the Share status box to see prior reporting history (if available).", 'examples': ['16284253', 'not reported'], 'options': [], 'see': None, 'evidence_category': 'SO', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661892+00:00', 'modified': '2019-10-22 22:46:59.661899+00:00'}),
        EvidenceKey(**{'key': 'report_type', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': None, 'sub_label': None, 'description': "Type of report generated for the most recent report. <p> Click on Compare 'shared versions' in the Share status box to see prior reporting history (if available).", 'examples': [], 'options': [{'key': 'interim', 'index': 1, 'label': ''}, {'key': 'final', 'index': 2, 'label': ''}, {'key': 'supplemental', 'index': 3, 'label': ''}, {'key': 'amended', 'index': 4, 'label': ''}, {'key': 'other', 'index': 5, 'label': ''}], 'see': None, 'evidence_category': 'SO', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.661947+00:00', 'modified': '2019-10-22 22:46:59.661955+00:00'}),
        EvidenceKey(**{'key': 'reputable_source_quality', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 19, 'label': None, 'sub_label': None, 'description': "A field for keeping track the quality of the quality of data as it is found from reputable sources.<br> Specifically if there's multiple entries provided, are they in agreement with each other (concordant).<br> Also are the entries provided with summary evidence.<br> If both these things are true BP6 or PP5 can be met. Otherwise those fields should be set to Not Met.", 'examples': [], 'options': [{'key': 'concordant', 'index': 1, 'label': ''}, {'key': 'discordant', 'index': 2, 'label': ''}, {'key': 'summary_evidence', 'index': 3, 'label': ''}, {'key': 'no_summary_evidence', 'index': 4, 'label': ''}], 'see': None, 'evidence_category': 'DB', 'value_type': 'M', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662003+00:00', 'modified': '2019-10-22 22:46:59.662011+00:00'}),
        EvidenceKey(**{'key': 'rescue_ese', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 21, 'label': 'RESCUE-ESE', 'sub_label': 'Splicing', 'description': 'RESCUE-ESE hexanucleotide sequence prediction. <p> RESCUE-ESE is an online tool for identifying candidate ESEs in vertebrate exons.', 'examples': [], 'options': [], 'see': 'http://genes.mit.edu/burgelab/rescue-ese/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662063+00:00', 'modified': '2019-10-22 22:46:59.662071+00:00'}),
        EvidenceKey(**{'key': 'revel', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': 'REVEL', 'sub_label': 'Meta-predictor', 'description': 'Rare Exome Variant Ensemble Learner (REVEL) score. <p> Score range: 0 to 1 <p> REVEL is an ensemble method for predicting the pathogenicity of missense variants on the basis of individual tools: MutPred, FATHMM, VEST, PolyPhen, SIFT, PROVEAN, MutationAssessor, MutationTaster, LRT, GERP, SiPhy, phyloP, and phastCons.', 'examples': ['0.894'], 'options': [], 'see': 'https://sites.google.com/site/revelgenomics/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'revel_score', 'created': '2019-10-22 22:46:59.662122+00:00', 'modified': '2019-10-22 22:46:59.662130+00:00'}),
        EvidenceKey(**{'key': 'review_comment', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Comments on curation and/or review process', 'sub_label': None, 'description': "Use this box to store comments related to the review or curation process such as details about corrections to fields or thoughts about the classification that might be useful for other curators. Note that variant interpretation text should be stored in the 'Interpretation Summary' field above and not here.", 'examples': ['Probably LP, but missing second supporting publication. ', 'Partial review only. Updated splicing interpretation based on new tool.'], 'options': [], 'see': None, 'evidence_category': 'SO', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662182+00:00', 'modified': '2019-10-22 22:46:59.662193+00:00'}),
        EvidenceKey(**{'key': 'sample_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Sample ID', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662242+00:00', 'modified': '2019-10-22 22:46:59.662250+00:00'}),
        EvidenceKey(**{'key': 'sample_type', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Sample Type', 'sub_label': None, 'description': 'Sample type used for sequencing.', 'examples': [], 'options': [{'key': 'tumor', 'index': 1, 'label': ''}, {'key': 'cvs', 'index': 2, 'label': 'CVS'}, {'key': 'amniocentesis', 'index': 3, 'label': ''}, {'key': 'blood', 'index': 4, 'label': ''}, {'key': 'buccal', 'index': 5, 'label': ''}, {'key': 'saliva', 'index': 6, 'label': ''}, {'key': 'hair', 'index': 7, 'label': ''}, {'key': 'blood_spot', 'index': 8, 'label': ''}], 'see': None, 'evidence_category': 'HT', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662298+00:00', 'modified': '2019-10-22 22:46:59.662306+00:00'}),
        EvidenceKey(**{'key': 'search_terms', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 1, 'label': None, 'sub_label': None, 'description': 'Search terms used to find literature for this variant', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'L', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662354+00:00', 'modified': '2019-10-22 22:46:59.662362+00:00'}),
        EvidenceKey(**{'key': 'segregation', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 14, 'label': 'Segregation # Families', 'sub_label': None, 'description': 'Enter the number of unrelated families (not counting the family analysed) showing segregation of >1 case.  Reference details can be stored in the comments to this field (references entered as PMID:XXX will be automatically recognised and added to the bibliography). Full comments and details can be stored in the segregation data summary section below.', 'examples': ['21'], 'options': [], 'see': None, 'evidence_category': 'S', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662409+00:00', 'modified': '2019-10-22 22:46:59.662417+00:00'}),
        EvidenceKey(**{'key': 'segregation_affectedcarriers', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'Segregation # Affected Carriers', 'sub_label': None, 'description': 'Enter the number of affected carriers of the variant.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'S', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662465+00:00', 'modified': '2019-10-22 22:46:59.662473+00:00'}),
        EvidenceKey(**{'key': 'segregation_affectednoncarriers', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'Segregation # Affected Non-carriers', 'sub_label': None, 'description': 'Enter the number of affected non-carriers of the variant.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'S', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662521+00:00', 'modified': '2019-10-22 22:46:59.662528+00:00'}),
        EvidenceKey(**{'key': 'segregation_bayes', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': 'Segregation Bayes Factor', 'sub_label': None, 'description': 'Bayes factor for segregation. Refer to <a href="https://www.sciencedirect.com/science/article/pii/S000292970762028X?via%3Dihub" target="_blank">Thompson et al. 2003</a>.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'S', 'value_type': 'L', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662576+00:00', 'modified': '2019-10-22 22:46:59.662584+00:00'}),
        EvidenceKey(**{'key': 'segregation_lod', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 13, 'label': 'Segregation LOD Score', 'sub_label': None, 'description': 'Enter the segregation LOD score.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'S', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662632+00:00', 'modified': '2019-10-22 22:46:59.662640+00:00'}),
        EvidenceKey(**{'key': 'segregation_meioses', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Segregation # Meioses', 'sub_label': None, 'description': 'Enter the number of informative meioses (segregations) in the family. "How many times does the variant segregate with disease in a family?"', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'S', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662688+00:00', 'modified': '2019-10-22 22:46:59.662695+00:00'}),
        EvidenceKey(**{'key': 'segregation_unaffectedcarriers', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 11, 'label': 'Segregation # Unaffected Carriers', 'sub_label': None, 'description': 'Enter the number of unaffected carriers of the variant.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'S', 'value_type': 'I', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662743+00:00', 'modified': '2019-10-22 22:46:59.662751+00:00'}),
        EvidenceKey(**{'key': 'segregation_within_family', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': None, 'sub_label': None, 'description': "Does this variant segregate within the proband's family? <p> Full comments and details can be stored in the segregation data summary section below. See de novo section for maternal and paternal inheritance. Be aware of the potential for identification when describing familial inheritance. <p>   If needed, reference details can be stored in the comments to this field (references entered as PMID:(NNNx) will be automatically recognised and added to the bibliography).", 'examples': [], 'options': [{'key': 'y', 'index': 1, 'label': 'Yes'}, {'key': 'n', 'index': 2, 'label': 'No'}], 'see': None, 'evidence_category': 'S', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662799+00:00', 'modified': '2019-10-22 22:46:59.662807+00:00'}),
        EvidenceKey(**{'key': 'sequencing_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 2, 'label': 'Sequencing ID', 'sub_label': None, 'description': 'SampleID from the SampleSheet.csv used for basecalling', 'examples': ['PATIENT_0042_SAMPLE_0002_S1'], 'options': [], 'see': 'http://support.illumina.com/content/dam/illumina-support/help/BaseSpaceHelp_v2/Content/Vault/Informatics/Sequencing_Analysis/BS/swSEQ_mBS_SampleSheet.htm', 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662859+00:00', 'modified': '2019-10-22 22:46:59.662867+00:00'}),
        EvidenceKey(**{'key': 'sequencing_method', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 5, 'label': None, 'sub_label': None, 'description': 'Method used to detect the sequence variation.', 'examples': [], 'options': [{'key': 'wgs', 'index': 1, 'label': 'WGS'}, {'key': 'wes', 'index': 2, 'label': 'WES'}, {'key': 'targeted_panel', 'index': 3, 'label': ''}, {'key': 'rna_seq', 'index': 4, 'label': 'RNA-seq'}, {'key': 'sanger', 'index': 5, 'label': ''}], 'see': None, 'evidence_category': 'HT', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662915+00:00', 'modified': '2019-10-22 22:46:59.662923+00:00'}),
        EvidenceKey(**{'key': 'sequencing_platform', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 4, 'label': None, 'sub_label': None, 'description': None, 'examples': [], 'options': [{'key': 'Illumina_HiSeq', 'index': 1, 'label': 'Illumina HiSeq'}, {'key': 'Illumina_MiSeq', 'index': 2, 'label': 'Illumina MiSeq'}, {'key': 'Illumina_NextSeq', 'index': 3, 'label': 'Illumina NextSeq'}, {'key': 'PacBio_Sequel', 'index': 4, 'label': 'PacBio Sequel'}], 'see': None, 'evidence_category': 'HT', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.662971+00:00', 'modified': '2019-10-22 22:46:59.662979+00:00'}),
        EvidenceKey(**{'key': 'sex', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 8, 'label': 'Sex', 'sub_label': None, 'description': 'Sex of the patient.', 'examples': [], 'options': [{'key': 'male', 'index': 1, 'label': 'Male'}, {'key': 'female', 'index': 2, 'label': 'Female'}, {'key': 'mixture', 'index': 3, 'label': 'Sex other'}], 'see': None, 'evidence_category': 'HP', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663026+00:00', 'modified': '2019-10-22 22:46:59.663034+00:00'}),
        EvidenceKey(**{'key': 'sift', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'SIFT', 'sub_label': None, 'description': 'Sorting Intolerant From Tolerant (SIFT) score or prediction. <p> Score range: 0 to 1 <p> Prediction categories: "Damaging", "Tolerated" <p> SIFT predicts whether an amino acid substitution affects protein function based on sequence homology and the physical properties of amino acids. SIFT can be applied to naturally occurring nonsynonymous polymorphisms and laboratory-induced missense mutations.', 'examples': [], 'options': [], 'see': 'http://sift.bii.a-star.edu.sg/', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'sift', 'created': '2019-10-22 22:46:59.663082+00:00', 'modified': '2019-10-22 22:46:59.663090+00:00'}),
        EvidenceKey(**{'key': 'siphy', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 15, 'label': 'SiPhy', 'sub_label': 'Conservation', 'description': 'SiPhy implements rigorous statistical tests to detect bases under selection from a multiple alignment data. It takes full advantage of deeply sequenced phylogenies to estimate both unlikely substitution patterns as well as slowdowns or accelerations in mutation rates.', 'examples': [], 'options': [], 'see': 'http://portals.broadinstitute.org/genome_bio/siphy/index.html', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663141+00:00', 'modified': '2019-10-22 22:46:59.663149+00:00'}),
        EvidenceKey(**{'key': 'snp_go', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'SNPs&GO', 'sub_label': None, 'description': None, 'examples': [], 'options': [], 'see': 'http://snps-and-go.biocomp.unibo.it/snps-and-go', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663199+00:00', 'modified': '2019-10-22 22:46:59.663207+00:00'}),
        EvidenceKey(**{'key': 's_other', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Other', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'S', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': True, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663255+00:00', 'modified': '2019-10-22 22:46:59.663263+00:00'}),
        EvidenceKey(**{'key': 'specimen_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 2, 'label': 'Specimen ID', 'sub_label': None, 'description': 'Specimen code', 'examples': ['DNA#24601'], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2020-02-14 04:25:41.025340+00:00', 'modified': '2020-02-14 04:25:41.025384+00:00'}),
        EvidenceKey(**{'key': 'spliceai', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 21, 'label': 'SpliceAI', 'sub_label': 'Splicing', 'description': 'SpliceAI score. <p> Score range: 0 to 1 <p> SpliceAI is a splice prediction tool developed by Illumina. It uses a deep neural network that accurately predicts splice junctions from an arbitrary pre-mRNA transcript sequence, enabling precise prediction of noncoding genetic variants that cause cryptic splicing. . <p> See <a href="https://www.ncbi.nlm.nih.gov/pubmed/30661751" target="_blank"> "Predicting Splicing from Primary Sequence with Deep Learning"</a> for more details. <p> Programmatic access available from <a href="https://github.com/Illumina/SpliceAI" target="_blank">Github. <p>', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663308+00:00', 'modified': '2019-10-22 22:46:59.663315+00:00'}),
        EvidenceKey(**{'key': 'splice_region', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 23, 'label': 'VEP SpliceRegion plugin', 'sub_label': 'Splicing', 'description': 'Plugin for the Ensembl Variant Effect Predictor (VEP) provides more granular predictions of splicing effects.', 'examples': ['splice_polypyrimidine_tract_variant', 'splice_donor_5th_base_variant'], 'options': [], 'see': 'https://asia.ensembl.org/info/docs/tools/vep/script/vep_plugins.html#spliceregion', 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'splice_region', 'created': '2019-10-22 22:46:59.663364+00:00', 'modified': '2019-10-22 22:46:59.663372+00:00'}),
        EvidenceKey(**{'key': 'splicesitefinder_like', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 20, 'label': 'SSF-like', 'sub_label': 'Splicing', 'description': 'SpliceSiteFinder-like (SSF-like) score. <p> SSF-like is based on position weight matrices computed from a set of human constitutive exon/intron junctions for donor and acceptor sites.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'CP', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663423+00:00', 'modified': '2019-10-22 22:46:59.663431+00:00'}),
        EvidenceKey(**{'key': 'splicing_assay', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 0, 'label': 'Splicing assay', 'sub_label': None, 'description': 'Free text summary of splicing assay results. Include PubMed IDs if the evidence is derived from the literature.', 'examples': [], 'options': [], 'see': None, 'evidence_category': 'F', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663479+00:00', 'modified': '2019-10-22 22:46:59.663487+00:00'}),
        EvidenceKey(**{'key': 'splicing_assertion', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 22, 'label': None, 'sub_label': 'Splicing', 'description': 'Interpretation of predicted impact of variant on splicing.', 'examples': [], 'options': [{'key': 'donor_site_disruption', 'index': 1, 'label': 'Donor site disruption'}, {'key': 'acceptor_site_disruption', 'index': 2, 'label': 'Acceptor site disruption'}, {'key': 'donor_site_creation', 'index': 3, 'label': 'Donor site creation'}, {'key': 'acceptor_site_creation', 'index': 4, 'label': 'Acceptor site creation'}, {'key': 'branchpoint_site_disruption', 'index': 5, 'label': 'Branchpoint site disruption'}, {'key': 'ese_disruption', 'index': 6, 'label': 'ESE disruption'}, {'key': 'ise_disruption', 'index': 7, 'label': 'ISE disruption'}, {'key': 'predicted_to_alter_splicing', 'index': 9, 'label': ''}, {'key': 'no_effect', 'index': 8, 'label': ''}], 'see': None, 'evidence_category': 'CP', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': True, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663535+00:00', 'modified': '2019-10-22 22:46:59.663543+00:00'}),
        EvidenceKey(**{'key': 's_summary', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 30, 'label': 'Summary', 'sub_label': None, 'description': None, 'examples': None, 'options': None, 'see': None, 'evidence_category': 'S', 'value_type': 'T', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663591+00:00', 'modified': '2019-10-22 22:46:59.663598+00:00'}),
        EvidenceKey(**{'key': 'topmed_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'TopMed', 'sub_label': None, 'description': 'The MAF found for the variant in TOPMed. TOPMed consists of ~155k participants from >80 different studies.', 'examples': ['0.01'], 'options': [], 'see': 'https://www.nhlbiwgs.org', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'topmed_af', 'created': '2019-10-22 22:46:59.663643+00:00', 'modified': '2019-10-22 22:46:59.663651+00:00'}),
        EvidenceKey(**{'key': 'uk10k_af', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'UK10K', 'sub_label': None, 'description': 'The MAF found for the variant in UK10K. UK10K consists of 10,000 whole genomes/exomes.', 'examples': [], 'options': [], 'see': 'https://www.uk10k.org', 'evidence_category': 'P', 'value_type': 'N', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'af_uk10k', 'created': '2019-10-22 22:46:59.663702+00:00', 'modified': '2019-10-22 22:46:59.663710+00:00'}),
        EvidenceKey(**{'key': 'uniprot_id', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 4, 'label': 'UniProt ID', 'sub_label': None, 'description': 'Universal Protein Knowledgebase Identifier.', 'examples': ['Q924Y4'], 'options': [], 'see': 'https://www.uniprot.org', 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'uniprot_id', 'created': '2019-10-22 22:46:59.663761+00:00', 'modified': '2019-10-22 22:46:59.663768+00:00'}),
        EvidenceKey(**{'key': 'vaast_gene_rank', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 7, 'label': 'VAAST Gene Rank', 'sub_label': None, 'description': 'VAAST (Variant Annotation, Analysis, and Search Tool) is a probabilistic search tool for identifying damaged genes and their disease-causing variants.', 'examples': ['2'], 'options': [], 'see': 'http://www.hufflab.org/software/vaast', 'evidence_category': 'H', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663824+00:00', 'modified': '2019-10-22 22:46:59.663835+00:00'}),
        EvidenceKey(**{'key': 'variant_class', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 9, 'label': None, 'sub_label': None, 'description': 'The  class of a variant according to its component alleles and its mapping to the reference genome ', 'examples': ['SNV', 'indel', 'copy_number_variation'], 'options': [{'so': 'SO:0001483', 'key': 'SNV', 'index': 1}, {'so': 'SO:0001645', 'key': 'genetic_marker', 'index': 2}, {'so': 'SO:1000002', 'key': 'substitution', 'index': 3}, {'so': 'SO:0000705', 'key': 'tandem_repeat', 'index': 4}, {'so': 'SO:0002063', 'key': 'Alu_insertion', 'index': 5}, {'so': 'SO:0001784', 'key': 'complex_structural_alteration', 'index': 6}, {'so': 'SO:1000005', 'key': 'complex_substitution', 'index': 7}, {'so': 'SO:0001742', 'key': 'copy_number_gain', 'index': 8}, {'so': 'SO:0001743', 'key': 'copy_number_loss', 'index': 9}, {'so': 'SO:0001019', 'key': 'copy_number_variation', 'index': 10}, {'so': 'SO:1000035', 'key': 'duplication', 'index': 11}, {'so': 'SO:0001873', 'key': 'interchromosomal_breakpoint', 'index': 12}, {'so': 'SO:0002060', 'key': 'interchromosomal_translocation', 'index': 13}, {'so': 'SO:0001874', 'key': 'intrachromosomal_breakpoint', 'index': 14}, {'so': 'SO:0002061', 'key': 'intrachromosomal_translocation', 'index': 15}, {'so': 'SO:1000036', 'key': 'inversion', 'index': 16}, {'so': 'SO:0001786', 'key': 'loss_of_heterozygosity', 'index': 17}, {'so': 'SO:0002066', 'key': 'mobile_element_deletion', 'index': 18}, {'so': 'SO:0001837', 'key': 'mobile_element_insertion', 'index': 19}, {'so': 'SO:0001838', 'key': 'novel_sequence_insertion', 'index': 20}, {'so': 'SO:0002096', 'key': 'short_tandem_repeat_variation', 'index': 21}, {'so': 'SO:1000173', 'key': 'tandem_duplication', 'index': 22}, {'so': 'SO:0000199', 'key': 'translocation', 'index': 23}, {'so': 'SO:0000159', 'key': 'deletion', 'index': 24}, {'so': 'SO:1000032', 'key': 'indel', 'index': 25}, {'so': 'SO:0000667', 'key': 'insertion', 'index': 26}, {'so': 'SO:0001059', 'key': 'sequence_alteration', 'index': 27}, {'so': 'SO:0000051', 'key': 'probe', 'index': 28}], 'see': None, 'evidence_category': 'V', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': 'variant_class', 'created': '2019-10-22 22:46:59.663885+00:00', 'modified': '2019-10-22 22:46:59.663893+00:00'}),
        EvidenceKey(**{'key': 'variant_confirmed', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 10, 'label': 'Variant confirmed by sanger or orthogonal method', 'sub_label': None, 'description': 'Was the variant confirmed by Sanger sequencing or another orthogonal method?', 'examples': [], 'options': [{'key': 'y', 'index': 1, 'label': 'yes'}, {'key': 'n', 'index': 2, 'label': 'no'}, {'key': 'pending', 'index': 3, 'label': ''}, {'key': 'recommended', 'index': 4, 'label': ''}], 'see': None, 'evidence_category': 'SO', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.663944+00:00', 'modified': '2019-10-22 22:46:59.663952+00:00'}),
        EvidenceKey(**{'key': 'variant_coordinate', 'mandatory': False, 'max_share_level': 'public', 'order': 4, 'label': None, 'sub_label': None, 'description': 'Chromosome : Position Ref > Alt', 'examples': [' 11:103048382 T>A'], 'options': [], 'see': None, 'evidence_category': 'V', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': True, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.664000+00:00', 'modified': '2019-10-22 22:46:59.664008+00:00'}),
        EvidenceKey(**{'key': 'variant_inheritance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 12, 'label': '', 'sub_label': '', 'description': 'Mode in which the variant under curation has been inherited in this patient. <p> Note that this concept is distinct from disease inheritance (see gene section), which is used to describe the mode of disease inheritance e.g. autosomal dominant etc.', 'examples': [], 'options': [{'key': 'de_novo', 'index': 1, 'label': ''}, {'key': 'uniparental', 'index': 2, 'label': ''}, {'key': 'uniparental_disomy', 'index': 3, 'label': ''}, {'key': 'biparental', 'index': 4, 'label': ''}, {'key': 'maternal', 'index': 5, 'label': ''}, {'key': 'paternal', 'index': 6, 'label': ''}, {'key': 'unknown', 'index': 7, 'label': ''}], 'see': '', 'evidence_category': 'V', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.665095+00:00', 'modified': '2019-10-22 22:46:59.665112+00:00'}),
        EvidenceKey(**{'key': 'variant_penetrance', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 6, 'label': 'Variant penetrance', 'sub_label': None, 'description': 'The overall probability that carriers of a type of variant or specific variant in a given gene will develop specified conditions until a specified age or during lifetime.  Note that ACMG/AMP guidelines should not be applied for low-penetrance variants.', 'examples': [], 'options': [{'key': 'high', 'index': 1, 'label': 'High'}, {'key': 'moderate', 'index': 2, 'label': 'Moderate'}, {'key': 'low', 'index': 3, 'label': 'Low'}, {'key': 'unknown', 'index': 4, 'label': 'Unknown'}], 'see': None, 'evidence_category': 'H', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': True, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.665167+00:00', 'modified': '2019-10-22 22:46:59.665176+00:00'}),
        EvidenceKey(**{'key': 'variant_qc', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 7, 'label': 'Variant QC', 'sub_label': None, 'description': 'Does the variant pass visual inspection (e.g. using IGV)? <br><br>  Check variant in cis that are in close proximity to each other.', 'examples': [], 'options': [{'key': 'pass', 'index': 1, 'label': ''}, {'key': 'fail', 'index': 0, 'label': ''}, {'key': 'artefact', 'index': 3, 'label': ''}, {'key': 'possible_artefact', 'index': 4, 'label': ''}, {'key': 'strand_bias', 'index': 5, 'label': ''}], 'see': None, 'evidence_category': 'HT', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.665225+00:00', 'modified': '2019-10-22 22:46:59.665233+00:00'}),
        EvidenceKey(**{'key': 'variant_reported', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 15, 'label': 'Variant included in report', 'sub_label': None, 'description': 'Was the variant included on a report?', 'examples': [], 'options': [{'key': 'primary_finding', 'index': 1, 'label': ''}, {'key': 'secondary_finding', 'index': 2, 'label': ''}, {'key': 'incidental_finding', 'index': 3, 'label': ''}, {'key': 'not_included', 'index': 4, 'label': ''}], 'see': None, 'evidence_category': 'SO', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.665282+00:00', 'modified': '2019-10-22 22:46:59.665290+00:00'}),
        EvidenceKey(**{'key': 'vcf_filter', 'mandatory': False, 'max_share_level': 'logged_in_users', 'order': 26, 'label': 'VCF Filter', 'sub_label': None, 'description': 'VCF Filters for locus', 'examples': ['PASS', 'LowQual', 'QDFilter'], 'options': [], 'see': None, 'evidence_category': 'HT', 'value_type': 'F', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.665338+00:00', 'modified': '2019-10-22 22:46:59.665346+00:00'}),
        EvidenceKey(**{'key': 'zygosity', 'mandatory': True, 'max_share_level': 'public', 'order': 11, 'label': None, 'sub_label': None, 'description': 'Zygosity in the tested individual.', 'examples': [], 'options': [{'key': 'heterozygous', 'index': 1, 'label': ''}, {'key': 'homozygous', 'index': 2, 'label': ''}, {'key': 'hemizygous', 'index': 3, 'label': ''}, {'key': 'mosaic', 'index': 4, 'label': ''}, {'key': 'compound_heterozygous', 'index': 5, 'label': ''}, {'key': 'heteroplasmic', 'index': 6, 'label': ''}, {'key': 'not_applicable', 'index': 7, 'label': ''}], 'see': None, 'evidence_category': 'V', 'value_type': 'S', 'default_crit_evaluation': None, 'allow_custom_values': False, 'hide': False, 'immutable': False, 'copy_consensus': False, 'variantgrid_column_id': None, 'created': '2019-10-22 22:46:59.665692+00:00', 'modified': '2019-10-22 22:46:59.665700+00:00'}),
    ]
    EvidenceKey.objects.bulk_create(keys)


class Migration(migrations.Migration):
    dependencies = [
        ('variantclassification', '0001_initial'),

    ]

    operations = [
        migrations.RunPython(blat_keys)
    ]
