{% if variant_data %}
    {% load static %}
    {% load js_tags %}
    {% comment %}
    Define a JS function "hotspot_graph_click_func(text, transcript_accession, protein_position)" to get callbacks from plotly clicks
    {% endcomment %}
    <div class="hotspot-graph" id="hotspot-graph-{{ uuid }}"></div>
    <style>
        .gnomad-af-slider {
            margin-left: 15px;
            width: 400px;
        }

        .gnomad-af-slider label {
          position: absolute;
          width: 20px;
          margin-top: 15px;
          margin-left: -10px;
          text-align: center;
        }

        .active-row {
            background-color: #f6f6ff !important;
            font-weight: bold;
        }
    </style>
    <div class="slider-container">
    {% if has_graph_filter_toolbar %}
        <label class="left" for="hotspot-graph-af-slider-{{ uuid }}">Graph gnomAD max %: </label>
        <div class="left gnomad-af-slider" id="hotspot-graph-af-slider-{{ uuid }}"></div>
    {% endif %}
        <a class="toggle-link" data-toggle="collapse" href="#hotspot-transcripts-{{ uuid }}">HotSpot graph transcripts...</a>
        <div class="clear"></div>
        <div id="hotspot-transcripts-{{ uuid }}" class="collapse">
            <div>Note: The protein position (x coordinate) is determined by VEP using their transcript versions.
                The protein domains come from <a href="https://www.ebi.ac.uk/interpro/">PFam (now InterPro)</a>
                If the versions don't match, the different version will be included.
            </div>
            {{ lookup_method }}
            <table>
            {% for accession, active, canonical_tag, details in transcript_options %}
                <tr class="{% if active %}active-row{% endif %}">
                    <td>
                        <a class="hotspot-load-transcript-link" accession="{{ accession }}">{{ accession }}</a>
                    </td>
                    <td>
                        {{ details }}
                    </td>
                    <td>
                        {% if canonical_tag %}
                            <span class="pills">
                                <span class="pill">{{ canonical_tag }}</span>
                            </span>
                        {% endif %}
                    </td>
                </tr>
            {% endfor %}
            </table>
        </div>
    </div>
    <script>
    (function() {
        const GNOMAD_PERCENT = [0.1, 1, 5, 100];
        let molecularConsequenceColors = {{ molecular_consequence_colors | jsonify }};
        let numCodons = {{ transcript_version.num_codons }};
        let domains = {{ domains | jsonify }};
        let variantData = {{ variant_data | jsonify }};
        let transcriptUrls = {{ transcript_urls | jsonify }};
        let hotSpotId = "hotspot-graph-{{ uuid }}";
        let hotSpotDiv = $("#" + hotSpotId);

        function drawTranscriptModel(selector, gnomADMaxPercent) {
            const DEFAULT_COLORS = [
                '#d62728',  // brick red
                '#2ca02c',  // cooked asparagus green
                '#1f77b4',  // muted blue
                '#ff7f0e',  // safety orange
                '#e377c2',  // raspberry yogurt pink
                '#bcbd22',  // curry yellow-green
                '#17becf',  // blue-teal
                '#9467bd',  // muted purple
                '#8c564b'   // chestnut brown
            ];

            let titleText = "{{ title }}";
            if (gnomADMaxPercent != 100) {
                titleText += "\ngnomAD max " + gnomADMaxPercent + "%";
            }
            let gnomadAFMax = gnomADMaxPercent / 100.0;
            let consequences = Object.keys(molecularConsequenceColors).sort();
            let yMax = 5; // show minimum so gene diagram doesn't get too fat
            let variantsByConsequence = {};
            for (let i = 0; i < consequences.length; ++i) {
                variantsByConsequence[consequences[i]] = {x: [], y: [], text: []};
            }

            for (let i = 0; i < variantData.length; ++i) {
                let vd = variantData[i];
                let text = vd[0];
                let x = vd[1];
                let consequence = vd[2];
                let gnomadAf = vd[3];
                let numSamples = vd[4];
                let vbc = variantsByConsequence[consequence];

                if (gnomadAf && gnomadAf > gnomadAFMax) {
                    continue;
                }

                vbc["x"].push(x);
                vbc["y"].push(numSamples);
                vbc["text"].push(text);
            }

            const BAR_WIDTH = 1.5;
            let data = [];
            // variantData is unique per allele, but there may be different alleles that have the same protein position
            // and consequence on the graph. The bar is by default stacked so this will push it up but we need to put
            // the lollypop on top after working out how high it is.
            let mergedXData = {};

            for (let i = 0; i < consequences.length; ++i) {
                let consequence = consequences[i];
                let v = variantsByConsequence[consequence];
                let color = molecularConsequenceColors[consequence];

                let barTrace = {
                    x: v.x,
                    y: v.y,
                    width: Array.from({length: v.y.length}).map(x => BAR_WIDTH),
                    hoverinfo: 'skip',
                    name: consequence,
                    type: 'bar',
                    marker: {
                        color: color
                    }
                };
                data.push(barTrace);

                for (let j = 0; j < v.x.length; ++j) {
                    let x = v.x[j];
                    let mergedData = mergedXData[x] || {count: 0, text: {}, last_consequence: null};
                    mergedData.count += v.y[j];
                    mergedData.text[v.text[j]] = 1;
                    mergedData.last_consequence = consequence;
                    mergedXData[x] = mergedData;
                }
            }

            // Lollypop head
            let consequenceLollypops = {};
            for (const [ x, mergedData ] of Object.entries(mergedXData)) {
                let lpData = consequenceLollypops[mergedData.last_consequence] || {x: [], y: [], text: []};
                lpData.x.push(x);
                lpData.y.push(mergedData.count);
                lpData.text.push(Object.keys(mergedData.text).join(" "));
                consequenceLollypops[mergedData.last_consequence] = lpData;
            }

            for ([consequence, lpData] of Object.entries(consequenceLollypops)) {
                let color = molecularConsequenceColors[consequence];
                let scatterTrace = {
                    x: lpData.x,
                    y: lpData.y,
                    text: lpData.text,
                    name: consequence,
                    showlegend: false,
                    mode: 'markers',
                    type: 'scatter',
                    marker: {
                        color: color,
                        size: BAR_WIDTH * 5
                    }
                };
                data.push(scatterTrace);
                yMax = Math.max(yMax, ...lpData.y);
            }


            let geneThickness = yMax / 5;
            let domainThickness = geneThickness; // by default no extra thickness
            let geneYTop = 0;
            let geneYBottom = geneYTop - geneThickness;
            // make domain lie entirely below, as it was sometimes obscuring classifications on top
            let domainYTop = geneYTop;
            let domainYBottom = geneYBottom - (domainThickness - geneThickness);

            // Need to give domains consistent colors
            let colorIndex = 0;
            let domainColors = {};

            let shapes = [
                {
                    type: "rect",
                    x0: 0,
                    y0: geneYTop,
                    x1: numCodons + 1, // Ends "on" that AA so need to go to next one to cover it.
                    y1: geneYBottom,
                    fillcolor: "#b0b0b0",
                }
            ];

            /* Fonts are specified in font-size units so we have to scale them ourselves
               and abbreviate or not show if they overflow the protein domain shape */
            const MIN_CHARS = 2;
            let graphSize = $("#" + selector).width();
            let fontScale = graphSize / (numCodons + 1);
            let fontSize = 10;
            let fontWidth = fontSize / fontScale;

            let annotations = [];
            for (let i = 0; i < domains.length; ++i) {
                let d = domains[i];
                let domainName = d[0];
                let domainStart = d[2];
                let domainEnd = d[3] + 1; // Ends "on" that AA so need to go to next one to cover it.
                let domainColor = domainColors[domainName];
                if (typeof (domainColor) == 'undefined') {
                    domainColor = DEFAULT_COLORS[colorIndex];
                    colorIndex++;
                    domainColors[domainName] = domainColor;
                }
                let domainShape = {
                    type: "rect",
                    x0: domainStart,
                    y0: domainYTop,
                    x1: domainEnd,
                    y1: domainYBottom,
                    fillcolor: domainColor,
                };
                shapes.push(domainShape);
                let domainWidth = domainEnd - domainStart;
                let domainMaxChars = domainWidth / fontWidth;
                let domainText = "";
                if (domainMaxChars >= MIN_CHARS) {
                    domainText = domainName;
                    if (domainText.length > domainMaxChars) {
                        domainText = domainText.substring(0, domainMaxChars - 1) + ".";
                    }
                }
                let annotation = {
                    showarrow: false,
                    x: domainStart + (domainEnd - domainStart) / 2,
                    y: domainYBottom + domainThickness / 2,
                    text: "<b>" + domainText + "</b>",
                    font: {
                        color: "white",
                        size: fontSize
                    },
                    xanchor: "center",
                };
                annotations.push(annotation);
            }
            let layout = {
                barmode: 'stack',
                title: {
                    text: titleText
                },
                showlegend: true,
                xaxis: {
                    title: "Amino acid",
                    range: [0, numCodons + 1],
                    showgrid: false
                },
                yaxis: {
                    title: "{{ y_title }}",
                    range: [domainYBottom, yMax + 2],
                    nticks: 5,
                    showgrid: false
                },
                shapes: shapes,
                annotations: annotations,
            };

            Plotly.newPlot(selector, data, layout);

            let myPlot = document.getElementById(selector);
            myPlot.on('plotly_click', function (data) {
                const hotspot_graph_click_func = hotSpotDiv.attr("hotspot_graph_click_func");
                if (typeof (hotspot_graph_click_func) != 'undefined') {
                    let barClicked = data.points[0];
                    let i = barClicked.data.x.indexOf(String(barClicked.x));
                    let text = "Hotspot click " + barClicked.data.text[i];
                    eval(hotspot_graph_click_func + "('{{ transcript_version.pk }}', text, barClicked.x);");
                }
            });
        }

        function clickTranscript() {
            const accession = $(this).attr("accession");
            const url = transcriptUrls[accession];
            const container = hotSpotDiv.parent();
            const hotspot_graph_click_func = hotSpotDiv.attr("hotspot_graph_click_func");
            container.empty();
            container.load(url, function() {
                // put click handler back
                $(".hotspot-graph", this).attr("hotspot_graph_click_func", hotspot_graph_click_func);
            });
        }

        $(document).ready(function () {
            let sliderDiv = $("#hotspot-graph-af-slider-{{ uuid }}");
            function drawMyTranscriptModel() {
                let gnomADMaxPercent = 100;
                if ($(sliderDiv).length) {
                    let value = $(sliderDiv).slider("value");
                    gnomADMaxPercent = GNOMAD_PERCENT[value];
                }
                drawTranscriptModel(hotSpotId, gnomADMaxPercent);
            }

            let gnomadMax = GNOMAD_PERCENT.length - 1;
            sliderDiv.slider({
                min: 0, max: gnomadMax, step: 1, value: gnomadMax,
                change: drawMyTranscriptModel
            }).each(function() {
                // From: https://stackoverflow.com/a/16877678/295724
                const opt = $(this).data().uiSlider.options;
                const vals = opt.max - opt.min;
                for (let i = 0; i <= vals; i++) {
                    const label = GNOMAD_PERCENT[i] + "%";
                    const el = $('<label>' + label + '</label>').css('left', (i/vals*100) + '%');
                    sliderDiv.append(el);
                }
            });

            drawMyTranscriptModel();

            $("a.hotspot-load-transcript-link", "#hotspot-transcripts-{{ uuid }}").on('click', clickTranscript);
        });
    })();
    </script>
{% endif %}