# Generated by Django 3.1 on 2020-09-29 06:09
import logging

from django.conf import settings
from django.db import migrations
from django.utils import timezone

from library.django_utils import bulk_insert_class_data
from library.database_utils import run_sql
from library.utils import single_quote, invert_dict
from snpdb.genome import reference_contigs
from snpdb.genome.reference_contigs import get_assembly_report_df


def snpdb_initial_data(apps, schema_editor):
    SITES = [
        {'id': 2, 'domain': 'localhost', 'name': 'localhost'},
        {'id': 3, 'domain': 'variantgrid.com', 'name': 'variantgrid.com'},
        {'id': 4, 'domain': 'frgeneseq02.imvs.sa.gov.au', 'name': 'TAU'},
        {'id': 5, 'domain': 'shariant.org.au', 'name': 'shariant.org.au'},
        {'id': 6, 'domain': 'test.shariant.org.au', 'name': 'Shariant Test'},
        {'id': 7, 'domain': 'runx1db.runx1-fpd.org', 'name': 'RUNX1db'}
    ]

    bulk_insert_class_data(apps, "sites", [("Site", SITES)])

    GENOMIC_INTERVAL_CATEGORIES = [
        {'name': 'Exome Capture Regions',
         'description': "Manufacturer's capture region bed files"},
        {'name': 'Uploaded', 'description': 'BED files uploaded by users'}
    ]

    bulk_insert_class_data(apps, "snpdb", [("GenomicIntervalsCategory", GENOMIC_INTERVAL_CATEGORIES)])


def create_columns(apps, schema_editor):
    CustomColumnsCollection = apps.get_model("snpdb", "CustomColumnsCollection")
    CustomColumn = apps.get_model("snpdb", "CustomColumn")

    VARIANT_GRID_COLUMN = [
        {'grid_column_name': 'af_1kg',
         'variant_column': 'variantannotation__af_1kg',
         'annotation_level': 'V',
         'width': None,
         'label': '1kg AF',
         'description': '<a href="http://www.1000genomes.org/">The 1000 Genomes project.</a> ALT allele frequency. Ranges from 0 to 1, calculated from ~2500 (phase3_v5) global individuals.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'af_uk10k',
         'variant_column': 'variantannotation__af_uk10k',
         'annotation_level': 'V',
         'width': None,
         'label': 'UK10K AF',
         'description': "<a href='http://www.uk10k.org'>The UK10K project</a> UK10K ALT allele frequency <b>ranging from 0 to 1</b>), which was calculated from the WGS data of 3,781 individuals. Those 3,781 individuals can be used as controls and were a combination of two sub-projects, <a href='http://www.uk10k.org/studies/cohorts.html'>The Avon Longitudinal Study of Parents and Children (ALSPAC) and the Department of Twin Research and Genetic Epidemiology (DTR)</a>).  At the multi-ALT-allele variant site, a comma is used as seperator to show the frequencies for each of the listed ALT alleles in the same order. Blank or '.' indicates that the ALT allele is not in the UK10K dataset.",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'alt',
         'variant_column': 'alt__seq',
         'annotation_level': None,
         'width': 40,
         'label': 'Alt',
         'description': 'Alternate non-reference alleles. A Blank alt means a reference call',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'amino_acids',
         'variant_column': 'variantannotation__amino_acids',
         'annotation_level': 'T',
         'width': None,
         'label': 'Amino acids',
         'description': 'Reference and alternative amino acids as generated by variant',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'band',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__band',
         'annotation_level': 'G',
         'width': 60,
         'label': 'Band',
         'description': 'Cytogenetic band.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'cadd_phred',
         'variant_column': 'variantannotation__cadd_phred',
         'annotation_level': 'T',
         'width': None,
         'label': 'CADD phred',
         'description': 'CADD phred-like score. This is phred-like rank score based on whole genome CADD raw scores. Please refer to Kircher et al. (2014) Nature Genetics 46(3):310-5 for details. The larger the score the more likely the SNP has damaging effect. <b>Currently missense variants only</b>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'cadd_raw',
         'variant_column': 'variantannotation__cadd_raw',
         'annotation_level': 'T',
         'width': None,
         'label': 'CADD raw score',
         'description': 'CADD raw score for functional prediction of a SNP. Please refer to Kircher et al. (2014) Nature Genetics 46(3):310-5 for details. The larger the score the more likely the SNP has damaging effect. Scores range from -6.458163 to 18.301497. <b>Currently missense variants only</b>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'canonical',
         'variant_column': 'variantannotation__canonical',
         'annotation_level': 'T',
         'width': None,
         'label': 'Canonical',
         'description': 'Transcript is denoted as the canonical transcript for this gene',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'ccds_ids',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__ccds_ids',
         'annotation_level': 'G',
         'width': 60,
         'label': 'CCDS Ids',
         'description': 'Consensus Coding Sequence Project Ids',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'chrom',
         'variant_column': 'locus__contig__name',
         'annotation_level': None,
         'width': 20,
         'label': 'Chrom',
         'description': 'Chromosome',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clingen_allele',
         'variant_column': 'variantallele__allele__clingen_allele',
         'annotation_level': 'V',
         'width': None,
         'label': 'ClinGen Canonical Allele ID',
         'description': '<a href="http://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/landing">ClinGen Allele Registry</a> Globally unique ID for each sequence change (works across genome builds). WARNING: values are NOT always populated - empty values do not mean there is no record',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinical_significance',
         'variant_column': 'clinvar__clinical_significance',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Clin Sig',
         'description': 'Variant Clinical Significance, 0 - Uncertain significance, 1 - not provided, 2 - Benign, 3 - Likely benign, 4 - Likely pathogenic, 5 - Pathogenic, 6 - drug response, 7 - histocompatibility, 255 - other"',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_allele_id',
         'variant_column': 'clinvar__clinvar_allele_id',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Allele',
         'description': 'ClinVar Allele ID',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_clinical_sources',
         'variant_column': 'clinvar__clinvar_clinical_sources',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar source',
         'description': "Variant's clinical sources reported as tag-value pairs of database and variant identifier",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_disease_database_name',
         'variant_column': 'clinvar__clinvar_disease_database_name',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar DB',
         'description': 'Tag-value pairs of disease database name and identifier, e.g. OMIM:NNNNNN',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_highest_pathogenicity',
         'variant_column': 'clinvar__highest_pathogenicity',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Max ClinSig',
         'description': 'Highest clinical_significance entry (CLNSIG)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_origin',
         'variant_column': 'clinvar__clinvar_origin',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Origin',
         'description': 'Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_preferred_disease_name',
         'variant_column': 'clinvar__clinvar_preferred_disease_name',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Disease',
         'description': "ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_review_status',
         'variant_column': 'clinvar__clinvar_review_status',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Review',
         'description': 'Clinical significance for this single variant',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'clinvar_variation_id',
         'variant_column': 'clinvar__clinvar_variation_id',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Variation',
         'description': 'ClinVar Variation ID',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'codons',
         'variant_column': 'variantannotation__codons',
         'annotation_level': 'T',
         'width': None,
         'label': 'Codons',
         'description': 'Reference and alternative codons as generated by variant',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'consequence',
         'variant_column': 'variantannotation__consequence',
         'annotation_level': 'T',
         'width': None,
         'label': 'Consequence',
         'description': 'Sequence Ontology consequence type of input variant allele on transcript',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'cosmic_count',
         'variant_column': 'variantannotation__cosmic_count',
         'annotation_level': 'V',
         'width': None,
         'label': 'COSMIC count',
         'description': 'Number of COSMIC samples with mutation',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'cosmic_id',
         'variant_column': 'variantannotation__cosmic_id',
         'annotation_level': 'V',
         'width': None,
         'label': 'COSMIC',
         'description': 'COSMIC ID for the variation included in the COSMIC database. A genetic variation can have more than one IDs, as COSMIC assigns different IDs to the same variation affecting different transcripts. Multiple COSMIC-IDs are separated by comma.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'cosmic_legacy_id',
         'variant_column': 'variantannotation__cosmic_legacy_id',
         'annotation_level': 'V',
         'width': None,
         'label': 'COSMIC Legacy ID',
         'description': "Legacy <a href='https://cancer.sanger.ac.uk/cosmic'>COSMIC</a> ID (COSM)",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'dbscsnv_ada_score',
         'variant_column': 'variantannotation__dbscsnv_ada_score',
         'annotation_level': 'V',
         'width': None,
         'label': 'dbscSNV Ada',
         'description': 'Splicing consensus score (Ada) a 0-1 probability of altering splicing. From <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267638/">dbscSNV</a>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'dbscsnv_rf_score',
         'variant_column': 'variantannotation__dbscsnv_rf_score',
         'annotation_level': 'V',
         'width': None,
         'label': 'dbscSNV RF',
         'description': 'Splicing consensus score (RF) a 0-1 probability of altering splicing. From <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267638/">dbscSNV</a>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'dbsnp_rs_id',
         'variant_column': 'variantannotation__dbsnp_rs_id',
         'annotation_level': 'T',
         'width': 70,
         'label': 'dbSNP',
         'description': 'dbSNP id',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'distance',
         'variant_column': 'variantannotation__distance',
         'annotation_level': 'T',
         'width': None,
         'label': 'Distance',
         'description': 'Distance to transcript if variant falls outside transcript boundaries',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'domains',
         'variant_column': 'variantannotation__domains',
         'annotation_level': 'T',
         'width': None,
         'label': 'Domains',
         'description': 'Protein domains overlapping variant',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'drug_response',
         'variant_column': 'clinvar__drug_response',
         'annotation_level': 'C',
         'width': None,
         'label': 'ClinVar Drug Response',
         'description': 'ClinVar - set if CLNSIG contains 6 - drug response.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'end_position',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__end_position',
         'annotation_level': 'G',
         'width': 50,
         'label': 'End',
         'description': ' The Ensembl gene start and end coordinates correspond to the outermost transcript start and end coordinates (containing the UTRs). These can be part of the same transcript or belong to two different transcripts. An example of the latter is the human BRCA2 gene, where the start coordinate corresponds to the start coordinate of transcript ENST00000380152 (BRCA2-001) and the end coordinate to the end coordinate of transcript ENST00000544455 (BRCA2-201).',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'ensembl_protein',
         'variant_column': 'variantannotation__ensembl_protein',
         'annotation_level': 'T',
         'width': None,
         'label': 'Ensembl Protein',
         'description': 'Ensembl stable identifier for affected protein product',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'entrez_gene_id',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__entrez_gene_id',
         'annotation_level': 'G',
         'width': None,
         'label': 'Entrez Gene',
         'description': 'Entrez Gene (NCBI) generates unique integers (GeneID) as stable identifiers for genes and other loci ',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'enzyme_ids',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__enzyme_ids',
         'annotation_level': 'G',
         'width': None,
         'label': 'Enzyme IDs',
         'description': 'Enzyme Nomenclature from <a href="http://www.chem.qmul.ac.uk/iubmb/enzyme/">NC-IUBMB</a>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'exon',
         'variant_column': 'variantannotation__exon',
         'annotation_level': 'T',
         'width': None,
         'label': 'Exon',
         'description': 'Number(s) of affected exon(s)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'external_gene_name',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__external_gene_name',
         'annotation_level': 'G',
         'width': None,
         'label': 'External Gene Name',
         'description': 'The gene name from Ensembl.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'fathmm_pred',
         'variant_column': 'variantannotation__fathmm_pred',
         'annotation_level': 'T',
         'width': None,
         'label': 'FATHMM Predictions',
         'description': '<a href="http://fathmm.biocompute.org.uk/">FATHMM</a>: Functional Analysis through Hidden Markov Models - predicts functional consequences of variants - all transcripts',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'fathmm_pred_most_damaging',
         'variant_column': 'variantannotation__fathmm_pred_most_damaging',
         'annotation_level': 'T',
         'width': None,
         'label': 'FATHMM Prediction',
         'description': 'Most damaging of <b>fathmm_pred</b>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'flags',
         'variant_column': 'variantannotation__flags',
         'annotation_level': 'T',
         'width': None,
         'label': 'Flags',
         'description': 'Transcript quality flags - cds_start_NF and/or ds_end_NF, indicating CDS 5 or 3 is incomplete',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'function_from_uniprotkb',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__function_from_uniprotkb',
         'annotation_level': 'G',
         'width': None,
         'label': 'Function from UniProtKB',
         'description': 'Function from <a href="http://www.uniprot.org">Universal Protein Knowledgebase.</a> The Swiss-Prot section is used, and it is manually annotated and reviewed and is of high-quality.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_description',
         'variant_column': 'variantannotation__transcript_version__gene_version__description',
         'annotation_level': 'G',
         'width': None,
         'label': 'Gene description',
         'description': 'Gene description',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_family_description',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__gene_family_description',
         'annotation_level': 'G',
         'width': None,
         'label': 'Gene family description',
         'description': 'From HGNC database.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_family_tag',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__gene_family_tag',
         'annotation_level': 'G',
         'width': None,
         'label': 'Gene family tag',
         'description': 'From HGNC database.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_id',
         'variant_column': 'variantannotation__gene',
         'annotation_level': 'V',
         'width': None,
         'label': 'Gene',
         'description': 'Ensembl Gene ID',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_symbol',
         'variant_column': 'variantannotation__transcript_version__gene_version__gene_symbol',
         'annotation_level': 'G',
         'width': 80,
         'label': 'Gene symbol',
         'description': 'Ensembl Gene Symbol',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gerp_pp_rs',
         'variant_column': 'variantannotation__gerp_pp_rs',
         'annotation_level': 'T',
         'width': None,
         'label': 'GERP++ RS',
         'description': '<a href="http://mendel.stanford.edu/SidowLab/downloads/gerp/">GERP++</a> (from dbNSFP) GERP++ RS score, the larger the score, the more conserved the site. Scores range from -12.3 to 6.17.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_af',
         'variant_column': 'variantannotation__gnomad_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD AF',
         'description': '"Allele Frequency (0-1) among all gnomAD genotypes (exome+genome)",',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_afr_af',
         'variant_column': 'variantannotation__gnomad_afr_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD AFR AF',
         'description': 'Allele Frequency (0-1) among African/African American genotypes (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_amr_af',
         'variant_column': 'variantannotation__gnomad_amr_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD AMR AF',
         'description': 'Allele Frequency (0-1) among Latino/Admixed American genotypes (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_asj_af',
         'variant_column': 'variantannotation__gnomad_asj_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD ASJ AF',
         'description': 'Allele Frequency (0-1) among Ashkenazi Jewish genotypes (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_eas_af',
         'variant_column': 'variantannotation__gnomad_eas_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD EAS AF',
         'description': 'Allele Frequency (0-1) among East Asian genotypes (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_filtered',
         'variant_column': 'variantannotation__gnomad_filtered',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD filtered',
         'description': 'Filters were applied in gnomAD VCF file (possibly indicating QC issues)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_fin_af',
         'variant_column': 'variantannotation__gnomad_fin_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD FIN AF',
         'description': 'Allele Frequency (0-1) among Finnish genotypes (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_hom_alt',
         'variant_column': 'variantannotation__gnomad_hom_alt',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD Hom Alt',
         'description': 'Count of homozygous individuals',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_nfe_af',
         'variant_column': 'variantannotation__gnomad_nfe_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD NFE AF',
         'description': 'Allele Frequency (0-1) among Non-Finnish European genotypes (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_oth_af',
         'variant_column': 'variantannotation__gnomad_oth_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD OTH AF',
         'description': 'Allele Frequency (0-1) among genotypes w/o ancestry (other) (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_popmax',
         'variant_column': 'variantannotation__gnomad_popmax',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD PopMax',
         'description': 'Population with highest allele frequency (value in gnomad_popmax_af',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_popmax_af',
         'variant_column': 'variantannotation__gnomad_popmax_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD PopMax AF',
         'description': 'Allele Frequency (0-1) in pop with highest AF (excluding Ashkenazi, Finnish, and indeterminate ancestry) (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_sas_af',
         'variant_column': 'variantannotation__gnomad_sas_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'gnomAD SAS AF',
         'description': 'Allele Frequency (0-1) among South Asian genotypes (exome+genome)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'grantham',
         'variant_column': 'variantannotation__grantham',
         'annotation_level': 'T',
         'width': None,
         'label': 'Grantham',
         'description': 'Predicts the effect of substitutions between amino acids based on chemical properties, including polarity and molecular volume. Score rangees from 0 to 215',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hgnc_id',
         'variant_column': 'variantannotation__hgnc_id',
         'annotation_level': 'T',
         'width': None,
         'label': 'HGNC ID',
         'description': 'HUGO Gene Nomenclature Committee Identifier',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hgnc_name',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__hgnc_name',
         'annotation_level': 'G',
         'width': None,
         'label': 'HGNC name',
         'description': 'Gene Symbol/name from HGNC. The HUGO Gene Nomenclature Committee (HGNC) is the only worldwide authority that assigns standardised nomenclature to human genes.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hgnc_symbol',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__hgnc_symbol',
         'annotation_level': 'G',
         'width': None,
         'label': 'HGNC symbol',
         'description': 'Gene symbol/name from HGNC. The HUGO Gene Nomenclature Committee (HGNC) is the only worldwide authority that assigns standardised nomenclature to human genes.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hgnc_symbol_lower',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__hgnc_symbol_lower',
         'annotation_level': 'G',
         'width': None,
         'label': 'HGNC symbol (lower)',
         'description': 'Lower case hgnc_symbol',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hgvs_c',
         'variant_column': 'variantannotation__hgvs_c',
         'annotation_level': 'T',
         'width': 200,
         'label': 'c.HGVS',
         'description': 'HGVS coding change',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hgvs_p',
         'variant_column': 'variantannotation__hgvs_p',
         'annotation_level': 'T',
         'width': 200,
         'label': 'p.HGVS',
         'description': 'HGVS protein change',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'impact',
         'variant_column': 'variantannotation__impact',
         'annotation_level': 'T',
         'width': None,
         'label': 'Impact',
         'description': 'Impact summary (MODIFIER/LOW/MODERATE/HIGH)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'in_cancer_gene_census',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__in_cancer_gene_census',
         'annotation_level': 'G',
         'width': None,
         'label': 'Cancer Gene Census',
         'description': 'Cancer gene census is a list of genes for which mutations have been causally implicated in cancer. It is from the current most release of COSMIC. In SACGF here we only mark genes in the list, and if you want to see more details about those gene (e.g. what kind of tumor the gene causes), check <a href="http://cancer.sanger.ac.uk/cancergenome/projects/census/">here</a>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'internally_classified',
         'variant_column': 'internally_classified',
         'annotation_level': 'D',
         'width': None,
         'label': 'Internal classification',
         'description': "Internal classifications joined by '|'",
         'model_field': False,
         'queryset_field': False},
        {'grid_column_name': 'interpro_domain',
         'variant_column': 'variantannotation__interpro_domain',
         'annotation_level': 'T',
         'width': None,
         'label': 'Interpro domain',
         'description': '<a href="https://www.ebi.ac.uk/interpro/">InterPro</a> - (from dbNSFP) domain or conserved site on which the variant locates. Domain annotations come from Interpro database. The number in the brackets following a specific domain is the count of times Interpro assigns the variant position to that domain, typically coming from different predicting databases. Multiple entries separated by ";".',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'intron',
         'variant_column': 'variantannotation__intron',
         'annotation_level': 'T',
         'width': None,
         'label': 'Intron',
         'description': 'Number(s) of affected intron(s)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'loftool',
         'variant_column': 'variantannotation__loftool',
         'annotation_level': 'T',
         'width': None,
         'label': 'LOF Tool',
         'description': '<a href="https://www.ncbi.nlm.nih.gov/pubmed/27563026">LoFtool</a>: a gene intolerance score based on loss-of-function variants in 60,706 individuals.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mastermind_count_1_cdna',
         'variant_column': 'variantannotation__mastermind_count_1_cdna',
         'annotation_level': 'V',
         'width': None,
         'label': 'MM variant article count',
         'description': "Count of <a href='https://www.genomenon.com/cvr/'>Mastermind</a> articles with cDNA matches for this specific variant",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mastermind_count_2_cdna_prot',
         'variant_column': 'variantannotation__mastermind_count_2_cdna_prot',
         'annotation_level': 'V',
         'width': None,
         'label': 'MM variant/protein article count',
         'description': "Count of <a href='https://www.genomenon.com/cvr/'>Mastermind</a> articles with variants either explicitly matching at the cDNA level or given only at protein level",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mastermind_count_3_aa_change',
         'variant_column': 'variantannotation__mastermind_count_3_aa_change',
         'annotation_level': 'V',
         'width': None,
         'label': 'MM AA article count',
         'description': "Count of <a href='https://www.genomenon.com/cvr/'>Mastermind</a> articles including other DNA-level variants resulting in the same amino acid change.",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mastermind_mmid3',
         'variant_column': 'variantannotation__mastermind_mmid3',
         'annotation_level': 'V',
         'width': None,
         'label': 'MM AA ID',
         'description': 'Mastermind MMID3 variant identifier(s), as gene:key',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'maxentscan_alt',
         'variant_column': 'variantannotation__maxentscan_alt',
         'annotation_level': 'T',
         'width': None,
         'label': 'MaxEntScan alt',
         'description': '<a href="http://hollywood.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq.html"></a>MaxEntScan Scores Human 5 splice sites. Alternate sequence score',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'maxentscan_diff',
         'variant_column': 'variantannotation__maxentscan_diff',
         'annotation_level': 'T',
         'width': None,
         'label': 'MaxEntScan diff',
         'description': '<a href="http://hollywood.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq.html"></a>MaxEntScan Scores Human 5 splice sites. Score difference',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'maxentscan_percent_diff_ref',
         'variant_column': 'variantannotation__maxentscan_percent_diff_ref',
         'annotation_level': 'T',
         'width': None,
         'label': 'MaxEntScan % diff/ref',
         'description': '<a href="http://hollywood.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq.html">MaxEntScan Scores Human 5 splice sites.</a> 100 * maxentscan_diff / maxentscan_ref',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'maxentscan_ref',
         'variant_column': 'variantannotation__maxentscan_ref',
         'annotation_level': 'T',
         'width': None,
         'label': 'MaxEntScan ref',
         'description': '<a href="http://hollywood.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq.html"></a>MaxEntScan Scores Human 5 splice sites. Reference sequence score',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'max_internal_classification',
         'variant_column': 'max_internal_classification',
         'annotation_level': 'D',
         'width': None,
         'label': 'Internal classification (Max)',
         'description': 'Highest internal classification',
         'model_field': False,
         'queryset_field': False},
        {'grid_column_name': 'mgi_id',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__mgi_id',
         'annotation_level': 'G',
         'width': None,
         'label': 'MGI Id',
         'description': 'Mouse Genome Informatics ID',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mutation_assessor_pred',
         'variant_column': 'variantannotation__mutation_assessor_pred',
         'annotation_level': 'T',
         'width': None,
         'label': 'Mutation Assessor preds',
         'description': '<a href="http://mutationassessor.org/r3/">MutationAssessors</a> functional impact of a variant - predicted functional, i.e. high ("H") or medium ("M"), or predicted non-functional, i.e. low ("L") or neutral ("N"). The MAori score cutoffs between "H" and "M", "M" and "L", and "L" and "N", are 3.5, 1.935 and 0.8, respectively. The rankscore cutoffs between "H" and "M", "M" and "L", and "L" and "N", are 0.9307, 0.52043 and 0.19675, respectively. All Transcripts.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mutation_assessor_pred_most_damaging',
         'variant_column': 'variantannotation__mutation_assessor_pred_most_damaging',
         'annotation_level': 'T',
         'width': None,
         'label': 'Mutation Assessor',
         'description': 'Most damaging of <b>mutation_assessor_pred</b>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mutation_taster_pred',
         'variant_column': 'variantannotation__mutation_taster_pred',
         'annotation_level': 'T',
         'width': None,
         'label': 'Mutation Taster preds',
         'description': '<a href="http://www.mutationtaster.org/">MutationTaster</a> prediction, "A" ("disease_causing_automatic"), "D" ("disease_causing"), "N" ("polymorphism") or "P" ("polymorphism_automatic"). The score cutoff between "D" and "N" is 0.5 for MTnew and 0.31733 for the rankscore.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'mutation_taster_pred_most_damaging',
         'variant_column': 'variantannotation__mutation_taster_pred_most_damaging',
         'annotation_level': 'T',
         'width': None,
         'label': 'Mutation Taster',
         'description': 'Most damaging of <b>mutation_taster_pred</b>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'omim_id',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__omim_id',
         'annotation_level': 'G',
         'width': None,
         'label': 'OMIM',
         'description': 'Online Mendelian Inheritance in Man for Mendelian traits and disorders. Official website at http://www.omim.org/',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'omim_phenotypes',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__omim_phenotypes',
         'annotation_level': 'G',
         'width': None,
         'label': 'OMIM phenotypes',
         'description': 'Online Mendelian Inheritance in Man Phenotypes',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'overlapping_symbols',
         'variant_column': 'variantannotation__overlapping_symbols',
         'annotation_level': 'V',
         'width': None,
         'label': 'Overlapping symbols',
         'description': 'Symbols from all overlapping transcripts',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'pathway_from_uniprotkb',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__pathway_from_uniprotkb',
         'annotation_level': 'G',
         'width': None,
         'label': 'Pathway UniProtKB',
         'description': 'From Universal Protein Knowledgebase',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'percentage_gc_content',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__percentage_gc_content',
         'annotation_level': 'G',
         'width': 20,
         'label': '%GC',
         'description': 'GC Percentage',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phastcons_100_way_vertebrate',
         'variant_column': 'variantannotation__phastcons_100_way_vertebrate',
         'annotation_level': 'T',
         'width': None,
         'label': 'PhastCons 100way vert',
         'description': 'phastCons conservation score based on the multiple alignments of 100 vertebrate genomes (including human). The larger the score, the more conserved the site. Scores range from 0 to 1.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phastcons_30_way_mammalian',
         'variant_column': 'variantannotation__phastcons_30_way_mammalian',
         'annotation_level': 'T',
         'width': None,
         'label': 'PhastCons 30 way mammal',
         'description': 'phastCons conservation score based on the multiple alignments of 30 mammalian genomes (including human). The larger the score, the more conserved the site. Scores range from 0 to 1.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phastcons_46_way_mammalian',
         'variant_column': 'variantannotation__phastcons_46_way_mammalian',
         'annotation_level': 'T',
         'width': None,
         'label': 'PhastCons 46 way mammal',
         'description': 'phastCons conservation score based on the multiple alignments of 46 vertebrate genomes (including human). The larger the score, the more conserved the site. Scores range from 0 to 1.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phenotypes_from_ensembl',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__phenotypes_from_ensembl',
         'annotation_level': 'G',
         'width': None,
         'label': 'Ensembl Phenotypes',
         'description': "Phenotype annotation from Ensembl, which incorporates OMIM, Orphanet, and DDG2P. Multiple entries are separated by '::' (double colons).",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phylop_100_way_vertebrate',
         'variant_column': 'variantannotation__phylop_100_way_vertebrate',
         'annotation_level': 'T',
         'width': None,
         'label': 'PhyloP 100way vert',
         'description': 'phyloP100way_vertebrate scores were ranked among all phyloP100way_vertebrate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phyloP100way_vertebrate scores in dbNSFP.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phylop_30_way_mammalian',
         'variant_column': 'variantannotation__phylop_30_way_mammalian',
         'annotation_level': 'T',
         'width': None,
         'label': 'PhyloP 30way mammal',
         'description': 'phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 100 vertebrate genomes (including human). The larger the score, the more conserved the site. Scores range from -20.0 to 10.003 in dbNSFP.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phylop_46_way_mammalian',
         'variant_column': 'variantannotation__phylop_46_way_mammalian',
         'annotation_level': 'T',
         'width': None,
         'label': 'PhyloP 46way mammal',
         'description': "('phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 46 vertebrate genomes (including human). The larger the score, the more conserved the site.',)",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'polyphen2_hvar_pred',
         'variant_column': 'variantannotation__polyphen2_hvar_pred',
         'annotation_level': 'T',
         'width': None,
         'label': 'Polyphen2 HVAR preds',
         'description': 'Polyphen2 prediction based on HumVar, "D" ("probably damaging", HVAR score in [0.909,1] or rankscore in [0.65694,0.97581]), "P" ("possibly damaging", HVAR in [0.447,0.908] or rankscore in [0.47121,0.65622]) and "B" ("benign", HVAR score in [0,0.446] or rankscore in [0.01493,0.47076]). Score cutoff for binary classification is 0.5 for HVAR score or 0.48762 for rankscore, i.e. the prediction is "neutral" if the HVAR score is smaller than 0.5 (rankscore is smaller than 0.48762), and "deleterious" if the HVAR score is larger than 0.5 (rankscore is larger than 0.48762). Multiple entries are separated by ";", corresponding to Uniprot_acc.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'polyphen2_hvar_pred_most_damaging',
         'variant_column': 'variantannotation__polyphen2_hvar_pred_most_damaging',
         'annotation_level': 'T',
         'width': None,
         'label': 'Polyphen2 HVAR',
         'description': 'Most damaging of <b>polyphen2_hvar_pred</b>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'position',
         'variant_column': 'locus__position',
         'annotation_level': None,
         'width': 70,
         'label': 'Position',
         'description': 'Reference position',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'previous_symbols',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__previous_symbols',
         'annotation_level': 'G',
         'width': None,
         'label': 'Previous gene symbols',
         'description': 'Previous Gene Names',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'protein_position',
         'variant_column': 'variantannotation__protein_position',
         'annotation_level': 'T',
         'width': None,
         'label': 'Protein position',
         'description': 'Amino acid position',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'pubmed',
         'variant_column': 'variantannotation__pubmed',
         'annotation_level': 'T',
         'width': None,
         'label': 'PubMed',
         'description': 'NCBI PubMed IDs of publications citing co-located variant',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'ref',
         'variant_column': 'locus__ref__seq',
         'annotation_level': None,
         'width': 40,
         'label': 'Reference',
         'description': 'Base(s) on the reference genome. Note that this is not necessarily the most prevalent base in the general population.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'refseq_gene_summary',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__refseq_gene_summary',
         'annotation_level': 'G',
         'width': None,
         'label': 'RefSeq gene summary',
         'description': 'RefSeq Gene Summary',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'repeat_masker',
         'variant_column': 'variantannotation__repeat_masker',
         'annotation_level': 'V',
         'width': None,
         'label': 'Repeat Masker',
         'description': '<a href="http://www.repeatmasker.org/">RepeatMasker</a> - interspersed repeats and low complexity DNA sequences',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'revel_score',
         'variant_column': 'variantannotation__revel_score',
         'annotation_level': 'T',
         'width': None,
         'label': 'REVEL score',
         'description': 'REVEL is an ensemble score based on 13 individual scores for predicting the pathogenicity of missense variants. Scores range from 0 to 1. The larger the score the more likely the SNP has damaging effect.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'rgd_id',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__rgd_id',
         'annotation_level': 'G',
         'width': None,
         'label': 'RGD ID',
         'description': 'Rat Genome Database ID',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'rvis_percentile',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__rvis_percentile',
         'annotation_level': 'G',
         'width': None,
         'label': 'RVIS percentile',
         'description': 'Residual Variation Intolerance Score (in percentile). The smaller the number, the higher the intolerance. For example, gene ATP1A3 has a percentile of 3.37%, meaning it is among the 3.37% most intolerant of human genes. Authors of the RVIS system assess each gene to see whether it has relatively more or less functional genetic variation than expected, with the data of more than six thousand human individuals from the public Exome Sequencing Project. Less functional variation means higher intolerance. In SACGF here we use the results of ranking (in percentile) for annotation. According to the authors, genes causing neurodevelopmental disease are more likely to be intolerant, and genes causing immunological diseases are often tolerant. I think RVIS is useful when looking for genes causing lethal or very severe diseases in early life (e.g. fetal death in utero), and for genes causing infertility. It is not so useful for looking for oncogenic genes. Reference at http://www.ncbi.nlm.nih.gov/pubmed/23990802',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'Sample',
         'variant_column': '',
         'annotation_level': 'S',
         'width': None,
         'label': 'Sample',
         'description': 'Sample Zygosities and call information',
         'model_field': False,
         'queryset_field': False},
        {'grid_column_name': 'sift',
         'variant_column': 'variantannotation__sift',
         'annotation_level': 'T',
         'width': None,
         'label': 'SIFT',
         'description': ' SIFT prediction. Prediction is "D(amaging)" or "T(olerated)". <a href="http://sift.jcvi.org/">Official website</a>',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'somatic',
         'variant_column': 'variantannotation__somatic',
         'annotation_level': 'V',
         'width': None,
         'label': 'Somatic',
         'description': 'Somatic status of co-located variant',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_dp_ag',
         'variant_column': 'variantannotation__spliceai_pred_dp_ag',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI AG pos',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta position for Acceptor Gain",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_dp_al',
         'variant_column': 'variantannotation__spliceai_pred_dp_al',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI AL pos',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta position for Acceptor Loss",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_dp_dg',
         'variant_column': 'variantannotation__spliceai_pred_dp_dg',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI DG pos',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta position for Donor Gain",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_dp_dl',
         'variant_column': 'variantannotation__spliceai_pred_dp_dl',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI DL pos',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta position for Donor Loss",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_ds_ag',
         'variant_column': 'variantannotation__spliceai_pred_ds_ag',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI AG score',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta score for Acceptor Gain",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_ds_al',
         'variant_column': 'variantannotation__spliceai_pred_ds_al',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI AL score',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta score for Acceptor Loss",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_ds_dg',
         'variant_column': 'variantannotation__spliceai_pred_ds_dg',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI DG score',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta score for Donor Gain",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'spliceai_pred_ds_dl',
         'variant_column': 'variantannotation__spliceai_pred_ds_dl',
         'annotation_level': 'V',
         'width': None,
         'label': 'SpliceAI DL score',
         'description': "<a href='https://pubmed.ncbi.nlm.nih.gov/30661751/'>SpliceAI</a> delta score for Donor Loss",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'splice_region',
         'variant_column': 'variantannotation__splice_region',
         'annotation_level': 'T',
         'width': None,
         'label': 'Splice region',
         'description': 'SpliceRegion predictions',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'start_position',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__start_position',
         'annotation_level': 'G',
         'width': 50,
         'label': 'Start position',
         'description': ' The Ensembl gene start and end coordinates correspond to the outermost transcript start and end coordinates (containing the UTRs). These can be part of the same transcript or belong to two different transcripts. An example of the latter is the human BRCA2 gene, where the start coordinate corresponds to the start coordinate of transcript ENST00000380152 (BRCA2-001) and the end coordinate to the end coordinate of transcript ENST00000544455 (BRCA2-201).',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'status',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__status',
         'annotation_level': 'G',
         'width': 60,
         'label': 'status',
         'description': "From Ensembl. 'KNOWN', 'NOVEL', or 'PUTATIVE'. A known gene or transcript matches to a sequence in a public, scientific database such as UniProtKB and NCBI RefSeq. The match must be for the same species, otherwise the gene is considered to be novel for that species. Putative genes are sequences identical to human ESTs that splice into two or more exons but without an ORF.",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'strand',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__strand',
         'annotation_level': 'G',
         'width': 30,
         'label': 'strand',
         'description': 'Genomic strand of gene 1 (+) / -1(-)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'swissprot',
         'variant_column': 'variantannotation__swissprot',
         'annotation_level': 'T',
         'width': None,
         'label': 'SwissProt',
         'description': 'Manually curated protein identifier from UniProt',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'symbol',
         'variant_column': 'variantannotation__symbol',
         'annotation_level': 'T',
         'width': None,
         'label': 'Symbol',
         'description': 'Gene Symbol',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'symbol_source',
         'variant_column': 'variantannotation__symbol_source',
         'annotation_level': 'T',
         'width': None,
         'label': 'Symbol source',
         'description': 'Gene Symbol source (HGNC, Ensembl, Entrez etc)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'synonyms',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__synonyms',
         'annotation_level': 'G',
         'width': None,
         'label': 'Synonyms',
         'description': 'Synonyms (From the HGNC)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'tags',
         'variant_column': 'tags',
         'annotation_level': 'D',
         'width': None,
         'label': 'Tags',
         'description': "Tags in this analysis joined by '|'",
         'model_field': False,
         'queryset_field': False},
        {'grid_column_name': 'tissue_specificity_from_uniprotkb',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__tissue_specificity_from_uniprotkb',
         'annotation_level': 'G',
         'width': None,
         'label': 'Tissue specificity UniProtKB',
         'description': 'From Universal Protein Knowledgebase',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'topmed_af',
         'variant_column': 'variantannotation__topmed_af',
         'annotation_level': 'V',
         'width': None,
         'label': 'TOPMed AF',
         'description': 'Allele Frequency from <a href="https://www.nhlbiwgs.org/">TOPMed</a> ~144k participants from >80 different studies',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'total_db_het',
         'variant_column': 'global_variant_zygosity__het_count',
         'annotation_level': 'D',
         'width': 35,
         'label': 'Het DB count',
         'description': 'Database het count',
         'model_field': False,
         'queryset_field': True},
        {'grid_column_name': 'total_db_hom',
         'variant_column': 'global_variant_zygosity__hom_count',
         'annotation_level': 'D',
         'width': 35,
         'label': 'Hom DB count',
         'description': 'Database hom count',
         'model_field': False,
         'queryset_field': True},
        {'grid_column_name': 'transcript_biotype',
         'variant_column': 'variantannotation__transcript_version__biotype',
         'annotation_level': 'T',
         'width': None,
         'label': 'Transcript biotype',
         'description': 'Transcript Biotype eg miRNA, lncRNA, snoRNA etc',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'transcript_count',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__transcript_count',
         'annotation_level': 'G',
         'width': 20,
         'label': 'Transcript count',
         'description': 'The number of transcripts produced by that gene, from Ensembl.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'trembl',
         'variant_column': 'variantannotation__trembl',
         'annotation_level': 'T',
         'width': None,
         'label': 'Trembl',
         'description': 'Automatically generated identifier from UniProt',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'ucsc_id',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__ucsc_id',
         'annotation_level': 'G',
         'width': None,
         'label': 'UCSC ID',
         'description': 'UCSC ID',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'uniparc',
         'variant_column': 'variantannotation__uniparc',
         'annotation_level': 'T',
         'width': None,
         'label': 'Uniparc',
         'description': 'Combined protein identifier from UniProt',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'uniprot_id',
         'variant_column': 'variantannotation__gene__ensemblgeneannotation__uniprot_id',
         'annotation_level': 'G',
         'width': None,
         'label': 'UniProt',
         'description': 'Function from <a href="http://www.uniprot.org">Universal Protein Knowledgebase.</a> The Swiss-Prot section is used, and it is manually annotated and reviewed and is of high-quality.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'variant',
         'variant_column': 'id',
         'annotation_level': None,
         'width': None,
         'label': 'Variant',
         'description': 'Unique ID of a variant in VariantGrid',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'variant_class',
         'variant_column': 'variantannotation__variant_class',
         'annotation_level': 'V',
         'width': None,
         'label': 'Variant class',
         'description': '<a href="https://asia.ensembl.org/info/genome/variation/prediction/classification.html#classes">Variant Class</a> using Sequence Ontology terms, eg SNV/indel/deletion etc',
         'model_field': True,
         'queryset_field': True}
    ]

    bulk_insert_class_data(apps, "snpdb", [("VariantGridColumn", VARIANT_GRID_COLUMN)])

    CUSTOM_COLUMNS_DEFAULT = [
        'variant',
        'chrom',
        'position',
        'ref',
        'alt',
        'gene_symbol',
        'hgvs_c',
        'tags',
        'Sample',
        'total_db_hom',
        'total_db_het',
        'impact',
        'consequence',
        'variant_class',
        'hgvs_p',
        'overlapping_symbols',
        'gnomad_af',
        'gnomad_popmax',
        'gnomad_popmax_af',
        'gnomad_filtered',
        'af_1kg',
        'af_uk10k',
        'topmed_af',
        'fathmm_pred_most_damaging',
        'mutation_assessor_pred_most_damaging',
        'mutation_taster_pred_most_damaging',
        'polyphen2_hvar_pred_most_damaging',
        'revel_score',
        'sift',
        'splice_region',
        'dbscsnv_ada_score',
        'dbscsnv_rf_score',
        'maxentscan_alt',
        'maxentscan_diff',
        'maxentscan_ref',
        'maxentscan_percent_diff_ref',
        'spliceai_pred_ds_ag',
        'spliceai_pred_dp_ag',
        'spliceai_pred_ds_al',
        'spliceai_pred_dp_al',
        'spliceai_pred_ds_dg',
        'spliceai_pred_dp_dg',
        'spliceai_pred_ds_dl',
        'spliceai_pred_dp_dl',
        'grantham',
        'loftool',
        'cadd_phred',
        'gerp_pp_rs',
        'max_internal_classification',
        'internally_classified',
        'clinvar_highest_pathogenicity',
        'clinical_significance',
        'clinvar_review_status',
        'clinvar_clinical_sources',
        'clinvar_variation_id',
        'clinvar_preferred_disease_name',
        'flags',
        'repeat_masker',
        'pubmed',
        'mastermind_count_1_cdna',
        'mastermind_count_2_cdna_prot',
        'mastermind_count_3_aa_change',
        'mastermind_mmid3',
        'domains',
        'interpro_domain',
        'cosmic_id',
        'cosmic_legacy_id',
        'cosmic_count',
        'dbsnp_rs_id',
        'ensembl_protein',
        'omim_id',
        'function_from_uniprotkb',
        'transcript_biotype',
        'gene_family_description',
        'gene_family_tag',
        'hgnc_symbol',
        'hgnc_name',
        'omim_phenotypes',
        'pathway_from_uniprotkb',
        'phenotypes_from_ensembl',
        'refseq_gene_summary',
        'rvis_percentile',
        'tissue_specificity_from_uniprotkb',
        'synonyms',
        'previous_symbols',
     ]

    CUSTOM_COLUMNS_MINIMALISM = [
        'variant',
        'chrom',
        'position',
        'ref',
        'alt',
        'gene_symbol',
        'tags',
        'hgvs_c',
        'impact',
        'consequence',
        'variant_class',
        'Sample',
        'total_db_hom',
        'total_db_het',
        'hgvs_p',
        'gnomad_af',
        'gnomad_popmax',
        'gnomad_popmax_af',
        'max_internal_classification',
        'internally_classified',
        'clinvar_highest_pathogenicity',
        'clinical_significance',
     ]

    CUSTOM_COLUMNS_ALL = [
        'variant',
        'chrom',
        'position',
        'ref',
        'alt',
        'gene_symbol',
        'tags',
        'hgvs_c',
        'impact',
        'consequence',
        'variant_class',
        'Sample',
        'total_db_hom',
        'total_db_het',
        'hgvs_p',
        'overlapping_symbols',
        'gnomad_af',
        'gnomad_popmax',
        'gnomad_popmax_af',
        'gnomad_afr_af',
        'gnomad_amr_af',
        'gnomad_asj_af',
        'gnomad_eas_af',
        'gnomad_filtered',
        'gnomad_fin_af',
        'gnomad_hom_alt',
        'gnomad_nfe_af',
        'gnomad_oth_af',
        'gnomad_sas_af',
        'af_1kg',
        'af_uk10k',
        'topmed_af',
        'cadd_phred',
        'fathmm_pred_most_damaging',
        'grantham',
        'loftool',
        'mutation_assessor_pred_most_damaging',
        'mutation_taster_pred_most_damaging',
        'polyphen2_hvar_pred_most_damaging',
        'revel_score',
        'sift',
        'dbscsnv_ada_score',
        'dbscsnv_rf_score',
        'maxentscan_alt',
        'maxentscan_diff',
        'maxentscan_ref',
        'maxentscan_percent_diff_ref',
        'spliceai_pred_ds_ag',
        'spliceai_pred_dp_ag',
        'spliceai_pred_ds_al',
        'spliceai_pred_dp_al',
        'spliceai_pred_ds_dg',
        'spliceai_pred_dp_dg',
        'spliceai_pred_ds_dl',
        'spliceai_pred_dp_dl',
        'splice_region',
        'gerp_pp_rs',
        'phastcons_100_way_vertebrate',
        'phastcons_30_way_mammalian',
        'phastcons_46_way_mammalian',
        'phylop_100_way_vertebrate',
        'phylop_30_way_mammalian',
        'phylop_46_way_mammalian',
        'distance',
        'repeat_masker',
        'pubmed',
        'mastermind_count_1_cdna',
        'mastermind_count_2_cdna_prot',
        'mastermind_count_3_aa_change',
        'mastermind_mmid3',
        'max_internal_classification',
        'internally_classified',
        'clinvar_highest_pathogenicity',
        'clinical_significance',
        'clinvar_review_status',
        'clinvar_clinical_sources',
        'clinvar_origin',
        'drug_response',
        'clinvar_variation_id',
        'clinvar_allele_id',
        'clinvar_preferred_disease_name',
        'clinvar_disease_database_name',
        'amino_acids',
        'canonical',
        'codons',
        'exon',
        'flags',
        'intron',
        'protein_position',
        'symbol_source',
        'domains',
        'interpro_domain',
        'cosmic_id',
        'cosmic_legacy_id',
        'cosmic_count',
        'dbsnp_rs_id',
        'ensembl_protein',
        'somatic',
        'swissprot',
        'trembl',
        'uniparc',
        'gene_description',
        'gene_id',
        'external_gene_name',
        'omim_id',
        'entrez_gene_id',
        'enzyme_ids',
        'function_from_uniprotkb',
        'transcript_biotype',
        'gene_family_description',
        'gene_family_tag',
        'hgnc_id',
        'hgnc_symbol',
        'hgnc_name',
        'mgi_id',
        'percentage_gc_content',
        'omim_phenotypes',
        'pathway_from_uniprotkb',
        'phenotypes_from_ensembl',
        'previous_symbols',
        'refseq_gene_summary',
        'rgd_id',
        'strand',
        'end_position',
        'rvis_percentile',
        'synonyms',
        'tissue_specificity_from_uniprotkb',
        'ucsc_id',
        'uniprot_id',
        'band',
        'ccds_ids',
        'start_position',
        'transcript_count',
        'status',
        'in_cancer_gene_census',
    ]

    CUSTOM_COLUMNS_COLLECTION = {
        'Default columns': CUSTOM_COLUMNS_DEFAULT,
        'Minimalism': CUSTOM_COLUMNS_MINIMALISM,
        'All columns': CUSTOM_COLUMNS_ALL,
    }

    for name, custom_columns_list in CUSTOM_COLUMNS_COLLECTION.items():
        ccc = CustomColumnsCollection.objects.create(name=name)
        records = []
        for i, column in enumerate(custom_columns_list):
            records.append(CustomColumn(custom_columns_collection=ccc, column_id=column, sort_order=i))
        CustomColumn.objects.bulk_create(records)


def add_requires_classification_tag(apps, schema_editor):
    Tag = apps.get_model("snpdb", "Tag")
    Tag.objects.get_or_create(id=settings.TAG_REQUIRES_CLASSIFICATION)


def create_bot_user(apps, schema_editor):
    User = apps.get_model("auth", "User")
    User.objects.get_or_create(
        username='admin_bot',
        defaults={
            'is_superuser': True,
            'first_name': 'Admin',
            'last_name': 'Bot',
            'email': 'admin@variantgrid.com',
            'is_staff': True,
            'is_active': True,
            'date_joined': timezone.now()
        }
    )


def import_contigs(apps, schema_editor):
    GenomeBuild = apps.get_model("snpdb", "GenomeBuild")
    GenomeBuildContig = apps.get_model("snpdb", "GenomeBuildContig")
    Contig = apps.get_model("snpdb", "Contig")

    # Maybe just pull genome build out of my machine?

    class SequenceRole:
        ASSEMBLED_MOLECULE = 'AM'
        UNLOCALIZED_SCAFFOLD = 'ULS'
        UNPLACED_SCAFFOLD = 'UPS'
        ALT_SCAFFOLD = 'ALT'
        FIX_PATCH = 'FP'
        NOVEL_PATCH = 'NP'

        CHOICES = (
            (ASSEMBLED_MOLECULE, "assembled-molecule"),
            (UNLOCALIZED_SCAFFOLD, "unlocalized-scaffold"),
            (UNPLACED_SCAFFOLD, "unplaced-scaffold"),
            (ALT_SCAFFOLD, "alt-scaffold"),
            (FIX_PATCH, "fix-patch"),
            (NOVEL_PATCH, "novel-patch"),
        )

    class AssemblyMoleculeType:
        CHROMOSOME = 'C'
        MITOCHONDRION = 'M'

        CHOICES = (
            (CHROMOSOME, "Chromosome"),
            (MITOCHONDRION, "Mitochondrion"),
        )

    BUILDS = {"hg19": None,
              "GRCh37": "hg19",
              "GRCh38": "hg38"}

    for build_name, alias in BUILDS.items():
        df = get_assembly_report_df(build_name)

        existing_build = GenomeBuild.objects.filter(name=build_name).first()
        if existing_build:
            raise ValueError(
                f"GenomeBuild {build_name} already exists. Use --clear if you want to delete (will delete all variants!)")

        kwargs = {"name": build_name}
        if alias:
            kwargs["alias"] = alias
        genome_build = GenomeBuild.objects.create(**kwargs)
        logging.info(f"Created build {genome_build}")
        role_lookup = invert_dict(dict(SequenceRole.CHOICES))
        molecule_type_lookup = invert_dict(dict(AssemblyMoleculeType.CHOICES))
        molecule_type_lookup["na"] = None

        i = 0
        for _, row in df.iterrows():
            role = row["Sequence-Role"]
            molecule_type = row["Assigned-Molecule-Location/Type"]
            defaults = {
                "name": row["Sequence-Name"],
                "role": role_lookup[role],
                "assigned_molecule": row["Assigned-Molecule"],
                "molecule_type": molecule_type_lookup[molecule_type],
                "genbank_accession": row["GenBank-Accn"],
                "ucsc_name": row["UCSC-style-name"],
                "length": row["Sequence-Length"]
            }
            contig, _ = Contig.objects.get_or_create(refseq_accession=row["RefSeq-Accn"],
                                                     defaults=defaults)
            GenomeBuildContig.objects.get_or_create(genome_build=genome_build,
                                                    contig=contig,
                                                    defaults={"order": i})
            i += 1

    GenomeBuild.objects.filter(name='hg19').update(igv_genome='hg19')
    GenomeBuild.objects.filter(name='GRCh37').update(igv_genome='b37')
    GenomeBuild.objects.filter(name='GRCh38').update(igv_genome='hg38')


def genome_build_accession(apps, schema_editor):
    # Move this pu into main bit
    GenomeBuild = apps.get_model("snpdb", "GenomeBuild")

    for build_name, assembly_report in reference_contigs.ASSEMBLY_REPORTS.items():
        report_parts = assembly_report.split("_")
        if not (report_parts[0] == 'GCF' and len(report_parts) == 5):
            raise ValueError(f"Assembly report {assembly_report} does not match expected format!")

        accession = "_".join(report_parts[:2])
        genome_build = GenomeBuild.objects.get(pk=build_name)
        genome_build.accession = accession
        genome_build.save()


def global_settings(apps, schema_editor):
    GenomeBuild = apps.get_model("snpdb", "GenomeBuild")
    GlobalSettings = apps.get_model("snpdb", "GlobalSettings")
    CustomColumnsCollection = apps.get_model("snpdb", "CustomColumnsCollection")

    # Set GlobalSettings with defaults
    DEFAULT_BUILD = "GRCh37"
    genome_build = GenomeBuild.objects.get(name=DEFAULT_BUILD)
    default_columns, _ = CustomColumnsCollection.objects.get_or_create(user=None, name=settings.DEFAULT_COLUMNS_NAME)
    GlobalSettings.objects.create(email_weekly_updates=True,
                                  email_discordance_updates=True,
                                  columns=default_columns,
                                  default_sort_by_column=None,
                                  variant_link_in_analysis_opens_new_tab=False,
                                  tool_tips=True,
                                  node_sql_tab=False,
                                  import_messages=True,
                                  igv_port=60151,
                                  default_genome_build=genome_build)


def add_global_initial_group_permissions(apps, schema_editor):
    Group = apps.get_model("auth", "Group")
    GlobalSettings = apps.get_model("snpdb", "GlobalSettings")
    SettingsInitialGroupPermission = apps.get_model("snpdb", "SettingsInitialGroupPermission")
    global_settings = GlobalSettings.objects.get()

    # Make public groups so they can be set default non share (not created in test yet)
    Group.objects.get_or_create(name=settings.PUBLIC_GROUP_NAME)
    Group.objects.get_or_create(name=settings.LOGGED_IN_USERS_GROUP_NAME)

    for group in Group.objects.all():
        public_group = group.name in [settings.PUBLIC_GROUP_NAME, settings.LOGGED_IN_USERS_GROUP_NAME]
        SettingsInitialGroupPermission.objects.create(settings=global_settings,
                                                      group=group, read=not public_group, write=False)


def create_vzc_partition(pk):
    base_table_name = "snpdb_variantzygositycount"
    sql_template = """
CREATE TABLE "%(table_name)s" (
    LIKE %(base_table_name)s including indexes,
    CHECK (%(records_fk_field)s = %(pk)s)
) INHERITS (%(base_table_name)s);
"""

    table_name = f"{base_table_name}_collection_{pk}"
    print(f"Creating Partition '{table_name}'")
    if isinstance(pk, str):
        pk = single_quote(pk)
    sql = sql_template % {"base_table_name": base_table_name,
                          "table_name": table_name,
                          "records_fk_field": "collection_id",
                          "pk": pk}
    run_sql(sql)


def variant_zygosity_count_defaults(apps, schema_editor):
    VariantZygosityCountCollection = apps.get_model("snpdb", "VariantZygosityCountCollection")

    vzcc = VariantZygosityCountCollection.objects.create(name="global", description="All uploaded samples")
    create_vzc_partition(vzcc.pk)


class Migration(migrations.Migration):
    dependencies = [
        ('snpdb', '0001_initial'),
        ('sites', '0002_alter_domain_unique')
    ]

    operations = [
        migrations.RunPython(snpdb_initial_data),
        migrations.RunPython(create_columns),
        migrations.RunPython(add_requires_classification_tag),
        migrations.RunPython(create_bot_user),
        migrations.RunPython(import_contigs),
        migrations.RunPython(genome_build_accession),
        migrations.RunPython(global_settings),
        migrations.RunPython(add_global_initial_group_permissions),
        migrations.RunPython(variant_zygosity_count_defaults),
    ]
