# Generated by Django 4.0.6 on 2022-08-09 07:12

from django.db import migrations
from django.db.models import Max

from library.django_utils import bulk_insert_class_data

# 'loftool' already existed, we just changed path
_NEW_COLUMNS = ['pathway_biocarta_full', 'pathway_consensus_pathdb', 'pathway_kegg_id', 'pathway_kegg_full',
                'gwas_trait_association', 'go_biological_process', 'go_cellular_component', 'go_molecular_function',
                'interactions_biogrid', 'interactions_consensus_pathdb', 'gnomad_pli', 'gnomad_prec', 'gnomad_pnull',
                'essential_gene_crispr', 'essential_gene_crispr2', 'essential_gene_gene_trap',
                'gene_damage_index_score', 'gene_damage_index_phred', 'phi', 'ghis', 'prec', 'hipred_score',
                'gene_indispensability_score', 'hipred_prediction', 'gene_indispensability_pred',
                'expression_egenetics', 'expression_gnf_atlas']


def _new_dbnsfp_gene_columns(apps, _schema_editor):
    VARIANT_GRID_COLUMN = [
        {'grid_column_name': 'gene_damage_index_score',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gene_damage_index_score',
         'annotation_level': 'G',
         'width': None,
         'label': 'GDI (Gene Damage Index Score)',
         'description': 'A genome-wide, gene-level metric of the mutational damage that has accumulated in the general population. From doi: 10.1073/pnas.1518646112. The higher the score the less likely the gene is to be responsible for monogenic diseases.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_damage_index_phred',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gene_damage_index_phred',
         'annotation_level': 'G',
         'width': None,
         'label': 'GDI (Gene Damage Index Phred)',
         'description': 'A genome-wide, gene-level metric of the mutational damage that has accumulated in the general population. From doi: 10.1073/pnas.1518646112. The higher the score the less likely the gene is to be responsible for monogenic diseases. Phred scaled',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'phi',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__phi',
         'annotation_level': 'G',
         'width': None,
         'label': 'Prob Haplo Ins',
         'description': 'P(HI): Estimated probability of haploinsufficiency of the gene (from doi:10.1371/journal.pgen.1001154)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'ghis',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__ghis',
         'annotation_level': 'G',
         'width': None,
         'label': 'Gene Haplo Ins',
         'description': 'GHIS: A score predicting the gene haploinsufficiency. The higher the score the more likely the gene is haploinsufficient. (from doi: 10.1093/nar/gkv474)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'prec',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__prec',
         'annotation_level': 'G',
         'width': None,
         'label': 'Prob Recessive disease',
         'description': 'P(rec): Estimated probability that gene is a recessive disease gene (from DOI:10.1126/science.1215040)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hipred_score',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__hipred_score',
         'annotation_level': 'G',
         'width': None,
         'label': 'HiPred prob Haplo Ins',
         'description': 'HIPred_score: Estimated probability of haploinsufficiency of the gene (from doi:10.1093/bioinformatics/btx028)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_indispensability_score',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gene_indispensability_score',
         'annotation_level': 'G',
         'width': None,
         'label': 'Gene Indispensability',
         'description': 'Gene_indispensability_score: A probability prediction of the gene being essential. From doi:10.1371/journal.pcbi.1002886',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'hipred_prediction',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__hipred_prediction',
         'annotation_level': 'G',
         'width': None,
         'label': 'HiPred predicted Haplo Ins',
         'description': 'HIPred: HIPred prediction of haploinsufficiency of the gene. Y(es) or N(o). (from doi:10.1093/bioinformatics/btx028)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gene_indispensability_pred',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gene_indispensability_pred',
         'annotation_level': 'G',
         'width': None,
         'label': 'Gene Indispensability prediction',
         'description': 'Gene_indispensability_pred: True = Essential or False = loss-of-function tolerant based on Gene_indispensability_score.',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'pathway_biocarta_full',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__pathway_biocarta_full',
         'annotation_level': 'G',
         'width': None,
         'label': 'BioCarta Pathway',
         'description': 'Full name(s) of the Pathway(s) the gene belongs to (from BioCarta)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'pathway_consensus_pathdb',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__pathway_consensus_pathdb',
         'annotation_level': 'G',
         'width': None,
         'label': 'ConsensusPathDB Pathway',
         'description': 'Pathway(s) the gene belongs to (from ConsensusPathDB)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'pathway_kegg_id',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__pathway_kegg_id',
         'annotation_level': 'G',
         'width': None,
         'label': 'KEGG IDs',
         'description': 'ID(s) of the Pathway(s) the gene belongs to (from KEGG)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'pathway_kegg_full',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__pathway_kegg_full',
         'annotation_level': 'G',
         'width': None,
         'label': 'KEGG Pathway',
         'description': 'Full name(s) of the Pathway(s) the gene belongs to (from KEGG)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gwas_trait_association',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gwas_trait_association',
         'annotation_level': 'G',
         'width': None,
         'label': 'GWAS traits',
         'description': 'Trait(s) the gene associated with (from GWAS catalog)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'go_biological_process',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__go_biological_process',
         'annotation_level': 'G',
         'width': None,
         'label': 'GO process',
         'description': 'GO terms for biological process',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'go_cellular_component',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__go_cellular_component',
         'annotation_level': 'G',
         'width': None,
         'label': 'GO cellular',
         'description': 'GO terms for cellular component',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'go_molecular_function',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__go_molecular_function',
         'annotation_level': 'G',
         'width': None,
         'label': 'GO mol function',
         'description': 'GO terms for molecular function',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'interactions_biogrid',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__interactions_biogrid',
         'annotation_level': 'G',
         'width': None,
         'label': 'BioGRID interactions',
         'description': 'The number of other genes this gene interacting with (from BioGRID)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'interactions_consensus_pathdb',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__interactions_consensus_pathdb',
         'annotation_level': 'G',
         'width': None,
         'label': 'ConsensusPathDB interactions',
         'description': 'The number of other genes this gene interacting with (from ConsensusPathDB).',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_pli',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gnomad_pli',
         'annotation_level': 'G',
         'width': None,
         'label': 'gnomAD prob LOF intolerant',
         'description': '"The probability of being loss-of-function intolerant (intolerant of both heterozygous and homozygous lof variants)" based on gnomAD 2.1 data',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_prec',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gnomad_prec',
         'annotation_level': 'G',
         'width': None,
         'label': 'gnomAD prob LOF HOM',
         'description': '"the probability of being intolerant of homozygous, but not heterozygous lof variants" based on gnomAD 2.1 data',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_pnull',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__gnomad_pnull',
         'annotation_level': 'G',
         'width': None,
         'label': 'gnomAD prob LOF tolerant',
         'description': '"the probability of being tolerant of both heterozygous and homozygous lof variants" based on gnomAD 2.1 data',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'expression_egenetics',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__expression_egenetics',
         'annotation_level': 'G',
         'width': None,
         'label': 'Expression(egenetics)',
         'description': 'Tissues/organs the gene expressed in (egenetics data from BioMart)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'expression_gnf_atlas',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__expression_gnf_atlas',
         'annotation_level': 'G',
         'width': None,
         'label': 'Expression(GNF/Atlas)',
         'description': 'Tissues/organs the gene expressed in (GNF/Atlas data from BioMart)',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'essential_gene_crispr',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__essential_gene_crispr',
         'annotation_level': 'G',
         'width': None,
         'label': 'Essential Gene (CRISPR)',
         'description': 'Essential ("E") or Non-essential phenotype-changing ("N") based on large scale CRISPR experiments. from doi: 10.1126/science.aac7041',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'essential_gene_crispr2',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__essential_gene_crispr2',
         'annotation_level': 'G',
         'width': None,
         'label': 'Essential Gene (CRISPR2)',
         'description': 'Essential ("E"), context-Specific essential ("S"), or Non-essential phenotype-changing ("N") based on large scale CRISPR experiments. from http://dx.doi.org/10.1016/j.cell.2015.11.015',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'essential_gene_gene_trap',
         'variant_column': 'variantannotation__gene__geneannotation__dbnsfp_gene__essential_gene_gene_trap',
         'annotation_level': 'G',
         'width': None,
         'label': 'Essential Gene (Gene Trap)',
         'description': 'Essential ("E"), HAP1-Specific essential ("H"), KBM7-Specific essential ("K"), or Non-essential phenotype-changing ("N"), based on large scale mutagenesis experiments. from doi: 10.1126/science.aac7557',
         'model_field': True,
         'queryset_field': True},
    ]
    bulk_insert_class_data(apps, "snpdb", [("VariantGridColumn", VARIANT_GRID_COLUMN)])

    VariantGridColumn = apps.get_model("snpdb", "VariantGridColumn")
    # LofTool already exists - but is on variant annotation
    loftool_qs = VariantGridColumn.objects.filter(grid_column_name='loftool')
    loftool_qs.update(variant_column='variantannotation__gene__geneannotation__dbnsfp_gene__loftool',
                      annotation_level='G')


def _reverse_new_dbnsfp_gene_columns(apps, _schema_editor):
    VariantGridColumn = apps.get_model("snpdb", "VariantGridColumn")

    VariantGridColumn.objects.filter(grid_column_name__in=_NEW_COLUMNS).delete()


def _custom_columns_for_new_dbnsfp_gene_columns(apps, _schema_editor):
    CustomColumnsCollection = apps.get_model("snpdb", "CustomColumnsCollection")
    CustomColumn = apps.get_model("snpdb", "CustomColumn")

    all_columns = CustomColumnsCollection.objects.get(name='All columns')
    data = all_columns.customcolumn_set.aggregate(Max("sort_order"))
    sort_order_max = data["sort_order__max"] or 0
    custom_columns = []
    for i, column_name in enumerate(_NEW_COLUMNS):
        custom_columns.append(CustomColumn(custom_columns_collection=all_columns,
                                           sort_order=sort_order_max + i,
                                           column_id=column_name))
    if custom_columns:
        CustomColumn.objects.bulk_create(custom_columns)


def _reverse_custom_columns_for_new_dbnsfp_gene_columns(apps, _schema_editor):
    CustomColumn = apps.get_model("snpdb", "CustomColumn")
    CustomColumn.objects.filter(column__grid_column_name__in=_NEW_COLUMNS).delete()


class Migration(migrations.Migration):
    dependencies = [
        ('annotation', '0061_dbnsfpgeneannotationversion_and_more'),
        ('snpdb', '0080_clinvarkey_last_full_run'),
    ]

    operations = [
        migrations.RunPython(_new_dbnsfp_gene_columns, reverse_code=_reverse_new_dbnsfp_gene_columns),
        migrations.RunPython(_custom_columns_for_new_dbnsfp_gene_columns,
                             reverse_code=_reverse_custom_columns_for_new_dbnsfp_gene_columns),
    ]
