# Generated by Django 4.2.2 on 2023-11-29 06:39

from django.db import migrations

from library.django_utils import bulk_insert_class_data


def _new_vep_110_annotation(apps, _schema_editor):
    TRANSCRIPT_LEVEL = 'T'
    VARIANT_LEVEL = 'V'

    NEW_VARIANT_GRID_COLUMNS = [
        {'grid_column_name': 'alphamissense_class',
         'variant_column': 'variantannotation__alphamissense_class',
         'annotation_level': TRANSCRIPT_LEVEL,
         'width': None,
         'label': 'AlphaMissense Class',
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'alphamissense_pathogenicity',
         'variant_column': 'variantannotation__alphamissense_pathogenicity',
         'annotation_level': TRANSCRIPT_LEVEL,
         'width': None,
         'label': 'AlphaMissense Pathogenicity',
         'model_field': True,
         'queryset_field': True},

        {'grid_column_name': 'gnomad_faf95',
         'variant_column': 'variantannotation__gnomad_faf95',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD FAF95',
         'description': "Filtering allele frequency (using Poisson 95% CI) (max of exomes/genomes) (GRCh38/gnomad4 only)",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_faf99',
         'variant_column': 'variantannotation__gnomad_faf99',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD FAF99',
         'description': "Filtering allele frequency (using Poisson 99% CI) (max of exomes/genomes) (GRCh38/gnomad4 only)",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_fafmax_faf95_max',
         'variant_column': 'variantannotation__gnomad_fafmax_faf95_max',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD FAF95 Max',
         'description': "Maximum filtering allele frequency (using Poisson 95% CI) across genetic_ancestry groups (max of exomes/genomes) (GRCh38/gnomad4 only)",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_fafmax_faf99_max',
         'variant_column': 'variantannotation__gnomad_fafmax_faf99_max',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD FAF99 Max',
         'description': "Maximum filtering allele frequency (using Poisson 99% CI) across genetic_ancestry groups (max of exomes/genomes) (GRCh38/gnomad4 only)",
         'model_field': True,
         'queryset_field': True},

        {'grid_column_name': 'gnomad_mid_af',
         'variant_column': 'variantannotation__gnomad_mid_af',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD MID AF',
         'description': "Allele Frequency (0-1) among Middle Eastern genotypes (exome+genome) (GRCh38/gnomad4 only)",
         'model_field': True,
         'queryset_field': True},

        {'grid_column_name': 'gnomad_non_par',
         'variant_column': 'variantannotation__gnomad_non_par',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD non-PAR',
         'description': "non_par in genomes or exomes",
         'model_field': True,
         'queryset_field': True},

        {'grid_column_name': 'gnomad_xy_ac',
         'variant_column': 'variantannotation__gnomad_xy_ac',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD XY AC',
         'description': "Allele Count in XY",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_xy_af',
         'variant_column': 'variantannotation__gnomad_xy_af',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD XY AF',
         'description': "Allele Frequency in XY",
         'model_field': True,
         'queryset_field': True},
        {'grid_column_name': 'gnomad_xy_an',
         'variant_column': 'variantannotation__gnomad_xy_an',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD XY AN',
         'description': "Allele Number in XY",
         'model_field': True,
         'queryset_field': True},

        {'grid_column_name': 'gnomad_hemi_count',
         'variant_column': 'variantannotation__gnomad_hemi_count',
         'annotation_level': VARIANT_LEVEL,
         'width': None,
         'label': 'gnomAD2 Hemizygous count',
         'description': "XY count (in non-PAR regions) - only on chrX",
         'model_field': True,
         'queryset_field': True},

        {'grid_column_name': 'mavedb_score',
         'variant_column': 'variantannotation__mavedb_score',
         'annotation_level': TRANSCRIPT_LEVEL,
         'width': None,
         'label': 'MAVEdb score',
         'model_field': True,
         'queryset_field': True},

        {'grid_column_name': 'mavedb_urn',
         'variant_column': 'variantannotation__mavedb_urn',
         'annotation_level': TRANSCRIPT_LEVEL,
         'width': None,
         'label': 'MAVEdb urn',
         'model_field': True,
         'queryset_field': True},
    ]

    NEW_COLUMN_VCF_INFO = [
        {'info_id': 'ALPHAMISSENSE_class',
         'column_id': 'alphamissense_class',
         'number': 1,
         'type': 'S',
         'description': 'AlphaMissense pathogenicity prediction'},
        {'info_id': 'ALPHAMISSENSE_pathogenicity',
         'column_id': 'alphamissense_pathogenicity',
         'number': 1,
         'type': 'F',
         'description': 'AlphaMissense pathogenicity prediction score'},
        {'info_id': 'GNOMAD_faf95',
         'column_id': 'gnomad_faf95',
         'number': 1,
         'type': 'F',
         'description': 'Filtering allele frequency (using Poisson 95% CI) (max of exomes/genomes) (GRCh38/gnomad4 only)'},
        {'info_id': 'GNOMAD_faf99',
         'column_id': 'gnomad_faf99',
         'number': 1,
         'type': 'F',
         'description': 'Filtering allele frequency (using Poisson 99% CI) (max of exomes/genomes) (GRCh38/gnomad4 only)'},
        {'info_id': 'GNOMAD_fafmax_faf95_max',
         'column_id': 'gnomad_fafmax_faf95_max',
         'number': 1,
         'type': 'F',
         'description': 'Maximum filtering allele frequency (using Poisson 95% CI) across genetic_ancestry groups (max of exomes/genomes) (GRCh38/gnomad4 only)'},
        {'info_id': 'GNOMAD_fafmax_faf99_max',
         'column_id': 'gnomad_fafmax_faf99_max',
         'number': 1,
         'type': 'F',
         'description': 'Maximum filtering allele frequency (using Poisson 99% CI) across genetic_ancestry groups (max of exomes/genomes) (GRCh38/gnomad4 only)'},
        {'info_id': 'GNOMAD_AF_mid',
         'column_id': 'gnomad_mid_af',
         'number': 1,
         'type': 'F',
         'description': 'Allele Frequency (0-1) among Middle Eastern genotypes (exome+genome) (GRCh38/gnomad4 only)'},
        {'info_id': 'GNOMAD_non_par',
         'column_id': 'gnomad_non_par',
         'number': 1,
         'type': 'F',
         'description': 'non_par in genomes or exomes'},
        {'info_id': 'GNOMAD_AC_XY',
         'column_id': 'gnomad_xy_ac',
         'number': 1,
         'type': 'I',
         'description': 'Allele Count in XY'},
        {'info_id': 'GNOMAD_AF_XY',
         'column_id': 'gnomad_xy_af',
         'number': 1,
         'type': 'F',
         'description': 'Allele Frequency in XY'},
        {'info_id': 'GNOMAD_AN_XY',
         'column_id': 'gnomad_xy_an',
         'number': 1,
         'type': 'I',
         'description': 'Allele Number in XY'},
        {'info_id': 'GNOMAD_HEMI_COUNT',
         'column_id': 'gnomad_hemi_count',
         'number': 1,
         'type': 'I',
         'description': 'XY count (in non-PAR regions) - only on chrX'},
        {'info_id': 'MaveDB_score',
         'column_id': 'mavedb_score',
         'number': 1,
         'type': 'F',
         'description': 'MaveDB score - see MaveDB for interpretation of scores'},
        {'info_id': 'MaveDB_urn',
         'column_id': 'mavedb_urn',
         'number': 1,
         'type': 'S',
         'description': 'MaveDB database identifier'}
    ]

    bulk_insert_class_data(apps, "snpdb", [("VariantGridColumn", NEW_VARIANT_GRID_COLUMNS)])
    bulk_insert_class_data(apps, "snpdb", [("ColumnVCFInfo", NEW_COLUMN_VCF_INFO)])


class Migration(migrations.Migration):
    dependencies = [
        ('snpdb', '0106_alter_allelemergelog_allele_linking_tool_and_more'),
    ]

    operations = [
        migrations.RunPython(_new_vep_110_annotation)
    ]
