{% load static %}
{% load js_tags %}
{% load ui_menu_bars %}
{% load ui_utils %}
{% load humanize %}
{% load tz %}
{% if not annotations_all_imported %}
<div class='card'>
    <div class="card-header"><h5>Installation Instructions</h5></div>
        <div class="card-body">
        <p>
        See instructions at <a href="https://github.com/sacgf/variantgrid/wiki/Annotation%20Setup">Annotation Setup (VariantGrid Wiki)</a>

        <p>Command line examples should be executed in the variantgrid dir <code>(${VG_INSTALL_DIR})</code> running as the <code>(${VARIANTGRID_USER})</code>

        <p>The variables should be set as per <a href="https://github.com/sacgf/variantgrid/wiki/Annotation%20Setup">wiki link (same as above)</a>:

        <ul>
         <li><code>${ANNOTATION_BASEDIR}</code> where you downloaded reference annotation
        </ul>

        <p>
        <b>Follow the instruction steps below in order</b> (as some have dependencies)
        </p>

        <p>
        Details instructions contain the way to download the latest copy of the annotations, or you can just use the out of date data from <code>${VARIANTGRID_SETUP_DATA}</code> dir (which came from downloading annotations)
        </p>
    </div>
</div>
{% endif %}

{% for build_name, details in genome_build_annotations.items %}
    <div id="build-{{ build_name }}-details" class="card mt-4">
        <div class="card-header">{{ build_name }}</div>
        <div class="card-body">
            {% labelled label="Latest" %}<a class="hover-link" href="{% url 'annotation_versions' %}">{{ details.latest }}</a>{% endlabelled %}
            {% labelled label="Genome Fasta" %}
                {% if details.reference_fasta %}
                    {{ details.reference_fasta }}
                {% else %}
                    <div class="not-installed">Not found</div>
                {% endif %}
                {% install-instructions "Reference Fasta" installed=details.reference_fasta %}
                    <div class="mono">{{ details.reference_fasta_error }}</div>
                    You need to download the annotations, or change <code>settings.GENOME_BUILD_REFERENCE_FASTA</code>
                {% endinstall-instructions %}
            {% endlabelled %}
            {% labelled label="Contigs" %}
                <div class="{{ build_name }}">
                    <b>Contigs:</b> {{ details.contigs.num_contigs }} <a href="#{{ build_name }}-contigs" data-toggle="collapse" class="toggle-link">Toggle</a>
                    <div id="{{ build_name }}-contigs" class='contig-details collapse'>{{ details.contigs.contigs }}</div>
                </div>
            {% endlabelled %}
            {% labelled label="Gene/Transcript Information" %}
                {% if details.genes_and_transcripts %}
                    <div>
                        <b>Annotation:</b> {{ details.annotation_consortium }}
                        <b>Genes:</b> {{ details.genes_and_transcripts.genes | intcomma }}
                        <b>Transcripts:</b> {{ details.genes_and_transcripts.transcripts | intcomma }}
                        <a href="#{{ build_name }}-import-sources" data-toggle="collapse" class="toggle-link">Toggle import sources</a>
                    </div>
                    <div id="{{ build_name }}-import-sources" class='gene-transcript-instructions collapse'>
                        <b>Import Sources</b>
                        {% for import_source, import_stats in details.genes_and_transcripts.import_sources.items %}
                            <div>
                            {{ import_source }}: <b>{{ import_stats.transcripts }}</b> transcripts ({{ import_stats.created|localtime }})
                            </div>
                        {% endfor %}
                    </div>
                {% endif %}

                {% for cons_name, other_consortia_data in details.other_consortia.items %}
                    <div>
                        <b>Annotation:</b> {{ cons_name }}
                        <b>Genes:</b> {{ other_consortia_data.genes | intcomma }}
                        <b>Transcripts:</b> {{ other_consortia_data.transcripts | intcomma }}
                        <a href="#{{ build_name }}-other-import-sources" data-toggle="collapse" class="toggle-link">Toggle import sources</a>
                    </div>
                    <div id="{{ build_name }}-other-import-sources" class='gene-transcript-instructions collapse'>
                        <b>Import Sources</b>
                        {% for file_name, import_stats in other_consortia_data.import_sources.items %}
                            <div>
                            {{ file_name }}: <b>{{ import_stats.transcripts }}</b> transcripts ({{ import_stats.created|localtime }})
                            </div>
                        {% endfor %}
                    </div>
                {% endfor %}

                {% install-instructions build_name installed=details.genes_and_transcripts.transcripts %}
                    <p>
                    You can install both Ensembl and RefSeq transcripts (for HGVS) but you NEED the ones used by your variant annotation pipeline which is: '<b>{{ details.annotation_consortium }}</b>'
                    </p>

                    <p>
                    Download data from <a href="https://github.com/SACGF/cdot/wiki/Download-JSON.gz-files">cdot</a>. See <a href="https://github.com/SACGF/variantgrid/wiki/Gene-and-Transcript-version-data">Gene and Transcript version data</a> for details and how to generate this from scratch
                    </p>

<code>
export gene_annotation_dir=${VARIANTGRID_SETUP_DATA}/gene_annotation
mkdir -p ${gene_annotation_dir}
cd ${gene_annotation_dir}
</code>
                    {% if build_name == 'GRCh37' %}
                    <h5>Ensembl GRCh37</h5>

<code>
wget https://cdot.cc/download/cdot-{{ cdot_version }}.ensembl.grch37.json.gz
python3 manage.py import_gene_annotation --annotation-consortium=Ensembl --genome-build=GRCh37 --json-file ${gene_annotation_dir}/cdot-{{ cdot_version }}.ensembl.grch37.json.gz
</code>
                    {% elif build_name == 'GRCh38' %}
                    <h5>Ensembl GRCh38</h5>
<code>
wget https://cdot.cc/download/cdot-{{ cdot_version }}.ensembl.grch38.json.gz
python3 manage.py import_gene_annotation --annotation-consortium=Ensembl --genome-build=GRCh38 --json-file ${gene_annotation_dir}/cdot-{{ cdot_version }}.ensembl.grch38.json.gz
</code>
                    {% endif %}

                    {% if build_name == 'GRCh37' %}
                    <h5>RefSeq GRCh37</h5>
<code>
wget https://cdot.cc/download/cdot-{{ cdot_version }}.refseq.grch37.json.gz
python3 manage.py import_gene_annotation --annotation-consortium=RefSeq --genome-build=GRCh37 --json-file ${gene_annotation_dir}/cdot-{{ cdot_version }}.refseq.grch37.json.gz
</code>
                    {% elif build_name == 'GRCh38' %}
                    <h5>RefSeq GRCh38</h5>
<code>
wget https://cdot.cc/download/cdot-{{ cdot_version }}.refseq.grch38.json.gz
python3 manage.py import_gene_annotation --annotation-consortium=RefSeq --genome-build=GRCh38 --json-file ${gene_annotation_dir}/cdot-{{ cdot_version }}.refseq.grch38.json.gz
</code>
                    {% endif %}
                {% endinstall-instructions %}
            {% endlabelled %}

            {% labelled label="Gene Annotation Release" %}
                {% if details.gene_annotation_release %}
                <span title="{{ details.gene_annotation_release.gene_annotation_import }}">{{ details.gene_annotation_release.name }}</span>
                {% else %}
                    <div class="no-value">Not Installed</div>
                {% endif %}
                {% install-instructions "Gene Annotation Release" installed=details.gene_annotation_release %}
                    You need a Gene Annotation Release that matches the exact symbol/gene/transcript versions used by VEP.
                    <a href="https://github.com/SACGF/variantgrid/wiki/Gene-and-Transcript-Version-Data#gene-annotation-release---install">See Wiki for details</a>

                    {% with vav=details.latest.variant_annotation_version %}
                        {% if vav.cdot_gene_release_filename %}
<code>
wget http://cdot.cc/download/{{ vav.cdot_gene_release_filename }}
python3 manage.py import_gene_annotation --annotation-consortium={{ vav.get_annotation_consortium_display }} --genome-build={{ build_name }} --json-file {{ vav.cdot_gene_release_filename }} --release={{ vav.suggested_gene_annotation_release_name }}
</code>
                            If it doesn't auto-link, go to <a href="{% url 'admin:annotation_variantannotationversion_change' vav.pk %}">Admin</a> then set the gene annotation release.
                        {% endif %}
                    {% endwith %}

                {% endinstall-instructions %}
            {% endlabelled %}
            {% labelled label="Gene Annotation" %}
               {% if details.gene_level_annotation %}
                    {{ details.gene_level_annotation }}
               {% else %}
                    <div class="no-value">Not Installed</div>
               {% endif %}
                {% install-instructions "Gene Annotation" installed=details.gene_level_annotation %}
                    <code>{{ python_command }} manage.py gene_annotation --gene-annotation-release={{ details.latest.variant_annotation_version.gene_annotation_release_id|default_if_none:'(Gene Annotation Release ID)' }}</code>
                {% endinstall-instructions %}
            {% endlabelled %}
            {% labelled label="ClinVar" %}
                {% if details.clinvar %}
                    {{ details.clinvar }}
                {% else %}
                    <div class="no-value">Not installed</div>
                {% endif %}
                {% install-instructions "Clinvar" installed=details.clinvar %}
                    Download VCF (with date in file name) from <a href="https://www.ncbi.nlm.nih.gov/clinvar/">NCBI ClinVar</a> site, then run:

                    <code>{{ python_command }} manage.py clinvar_import ${VARIANTGRID_SETUP_DATA}/clinvar_20180401.vcf.gz</code>
               {% endinstall-instructions %}

            {% endlabelled %}
            {% if somalier_enabled %}
                {% labelled label="Somalier" %}
                    {% if details.somalier %}
                        {{ details.somalier }}
                    {% else %}
                       Not installed
                    {% endif %}
                    {% install-instructions "Somalier" installed=details.somalier %}
                        Upload the Somalier sites VCF for {{ build_name }}.
                    {% endinstall-instructions %}
                {% endlabelled %}
            {% endif %}
        </div>
    </div>
{% endfor %}

<div class="card mt-4">
    <div class="card-header"><h5>Information</h5></div>
    <div class="card-body">
        {% labelled label="Human Protein Atlas" %}
            {% if hpa_counts %}
                Version: {{ hpa_version.hpa_version }} ({{ hpa_version.unit}}) Records: {{ hpa_counts | intcomma }}
            {% else %}
                <div class="no-value">Not installed</div>
            {% endif %}
            {% install-instructions "Human Protein Atlas" installed=hpa_counts %}
                <div>Go to <a href='https://www.proteinatlas.org/about/download'>Protein Atlas Download Page</a> and download <b>rna_tissue_consensus.tsv.zip</b> note the version (ie <b>21</b>):</div>
<code>
wget https://v21.proteinatlas.org/download/rna_tissue_consensus.tsv.zip
{{ python_command }} manage.py human_protein_atlas_import --hpa-version 21 ${VARIANTGRID_SETUP_DATA}/rna_tissue_consensus.tsv.zip
</code>
            {% endinstall-instructions %}
        {% endlabelled %}

        {% labelled label="Gene Symbol Aliases" %}
            {% if gene_symbol_alias_counts %}
                {% for source, count in gene_symbol_alias_counts.items %}
                    <b>{{ source }}:</b> {{ count | intcomma }}
                {% endfor %}
            {% else %}
                <div class="no-value">Not installed</div>
            {% endif %}
            {% install-instructions "Gene Symbol Aliases" installed=gene_symbol_alias_counts %}
<code>
wget ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz
{{ python_command }} manage.py import_ncbi_gene_info ${VARIANTGRID_SETUP_DATA}/Homo_sapiens.gene_info.gz
</code>
             {% endinstall-instructions %}
        {% endlabelled %}

        {% labelled label="dbNSFP Gene Annotation" %}
            {% if dbnsfp_gene_annotation %}
                <div>{{ dbnsfp_gene_annotation }}</div>
            {% else %}
                <div class="no-value">Not installed</div>
            {% endif %}
            {% install-instructions "dbNSFP gene annotation" installed=dbnsfp_gene_annotation %}
                {% comment %}
                    I asked dbNSFP guy to host just the genes gz on their site, so check maybe it's up and we can remove our own hosting of it
                {% endcomment %}
                Download <a href="http://variantgrid.com/static/annotation_download/dbNSFP4.3_gene.complete.gz">dbNSFP4.3_gene.complete.gz</a> (43Mb) - hosted on variantgrid.com
                This is just the gene data extracted from the <a href="https://sites.google.com/site/jpopgen/dbNSFP">dbNSFP</a> release.

                <code>{{ python_command }} manage.py import_dbnsfp_gene_annotation --dbnsfp-version=4.3 dbNSFP4.3_gene.complete.gz</code>
             {% endinstall-instructions %}
        {% endlabelled %}

        {% labelled label="TranscriptVersion Sequence Info" %}
            {% if transcript_version_sequence_info %}
                {{ transcript_version_sequence_info }}
                {% for fasta_import in transcript_fasta_imports %}
                    <div title="{{ fasta_import.md5_hash }}">
                        {{ fasta_import }} ({{ fasta_import.created|localtime }})
                    </div>
                {% endfor %}
            {% else %}
                <div class="no-value">None</div>
            {% endif %}
            {% install-instructions "TranscriptVersion Sequence Info" installed=transcript_fasta_imports %}
            This is genome build independent, so just download most recently updated version.
<code>wget https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_rna.fna.gz
{{ python_command }} manage.py import_transcript_fasta --annotation-consortium=RefSeq ${VARIANTGRID_SETUP_DATA}/GRCh38_latest_rna.fna.gz</code>
             {% endinstall-instructions %}
        {% endlabelled %}

        {% for ontology_count in ontology_counts %}
            {% labelled label=ontology_count.service %}
                {{ ontology_count.count | intcomma }}
            {% endlabelled %}
        {% endfor %}
        {% labelled label="Ontology Relationships" %}
            <table class="table">
                <thead>
                    <tr>
                        <th></th>{% for service in ontology_services %}<th class="num">{{ service }}</th>{% endfor %}
                    </tr>
                </thead>
                <tbody>
                    {% for ontology_row in ontology_services %}
                        <tr>
                        <th>{{ ontology_row }}</th>
                        {% for ontology_col in ontology_services %}
                            {% with key=ontology_row|add:ontology_col %}

                            <td class="num {% if forloop.counter > forloop.parentloop.counter %}no-value{% endif %}">{{ ontology_relationship_counts | get_item:key | intcomma}}</td>
                            {% endwith %}
                        {% endfor %}
                        </tr>
                    {% endfor %}
                </tbody>
            </table>
        {% endlabelled %}
        {% labelled label="Ontology Imports" %}
            {% for ontology_import in ontology_imports %}
                <div><b>{{ ontology_import.context }}</b> : {% timestamp ontology_import.last_import.processed_date time_ago=True %}</div>
            {% endfor %}
            {% install-instructions "Ontology Imports" installed=ontology_imports %}
                <div>
                    <b>HGNC</b><br/>
                    OntologyTerms have a redundant copy of HGNC to help map relationships between terms and genes. To load HGNC OntologyTerms
                    using HGNC Gene Names as a basis, run
<code>{{ python_command }} manage.py ontology_import --hgnc_sync</code>
                </div>
                <div class="mt-4">
                    <b>MONDO</b><br/>mondo.json can be downloaded from Download mondo.owl from <a href="http://www.obofoundry.org/ontology/mondo.html">Monarch Disease Ontology</a> on OBO foundry.<br/>
<code>wget http://purl.obolibrary.org/obo/mondo.json
{{ python_command }} manage.py ontology_import --mondo (location of mondo.json)</code>
                </div>
                <div class="mt-4">
                    <b>HPO</b><br/>hp.owl can be downloaded from
<code>wget https://raw.githubusercontent.com/obophenotype/human-phenotype-ontology/master/hp.owl
{{ python_command }} manage.py ontology_import --hpo (location of hp.owl) </code>
                </div>
                <div class="mt-4">
                    <b>OMIM</b><br/>Direct OMIM import requires a license, but will be populated via MONDO, HPO Diseases File and biomart<br/>
                    Export: "MIM Morbid Accession", "MIM Morbid Description" columns from <a href="https://www.ensembl.org/biomart/martview/">Ensembl Biomart</a><br/> as a TSV.
                    Use (GRC38 version or later as it has aliases) and make sure you tick "unique"
<code>{{ python_command }} manage.py ontology_import --biomart (location of mart_export)</code><br/>
                    If you are licensed to have access to mimTitles.txt it can be imported via
<code>{{ python_command }} manage.py ontology_import --omim (location of mimTitles.txt)</code>
                </div>
                <div class="mt-4">
                    <b>HPO/OMIM/HGNC phenotype_to_genes</b><br/>
                    Download phenotype_to_genes.txt from <a href="https://hpo.jax.org/app/download/annotation">Human Phenotype Ontology</a><br/>
<code>wget http://purl.obolibrary.org/obo/hp/hpoa/phenotype_to_genes.txt
{{ python_command }} manage.py ontology_import --phenotype_to_genes (location of phenotype_to_genes.txt) </code>
                </div>
                <div class="mt-4">
                    <b>MONDO/HGNC GenCC</b><br/>
                    Download gencc-submissions.csv from <a href="https://search.thegencc.org/download">GenCC</a><br/>
<code>wget https://search.thegencc.org/download/action/submissions-export-csv --output-document=gencc-submissions.csv
{{ python_command }} manage.py ontology_import --gencc (location of gencc-submissions.csv) </code>
                </div>
                <div class="mt-4">
                    <b>Bulk</b><br/>
If all the files are in place, the uploads can be triggered by a single command rather than waiting for each step to finish, e.g.
<code>{{ python_command }} manage.py ontology_import --hgnc_sync --mondo (location of mondo.json) --hpo (location of hp.owl) --biomart (location of mart_export) --phenotype_to_genes (location of phenotype_to_genes.txt) --gencc (location of gencc-submissions.csv)</code>
                </div>

            {% endinstall-instructions %}
        {% endlabelled %}

        {% if somalier_enabled %}
             <hr/>
            {% labelled label="Somalier" %}
                {% if somalier %}
                    {{ somalier }}
                {% else %}
                    <div class="no-value">Not Installed</div>
                {% endif %}
                {% install-instructions "Somalier" installed=somalier %}
                    <p>
                        Change install dir in settings.SOMALIER
                    <code>
mkdir /data/annotation/somalier
cd /data/annotation/somalier
${VARIANTGRID_DIR}/scripts/install/get_somalier_release.sh
                    </code>
                    </p>
                {% endinstall-instructions %}
            {% endlabelled %}
        {% endif %}
        {% for cwr in cached_web_resources %}
            {% labelled label=cwr.name %}
                <div class='cached-web-resource-container'>
                {% if cwr.import_status == 'I' %}
                    <img src="{% static 'images/spinner.gif' %}" /> <b>Import started: </b>{{ cwr.import_date }}
                {% else %}
                    {% if user.is_superuser %}
                        {% if cwr.import_status != 'S' %}
                        <b>Status:</b> {{ cwr.get_import_status_display }}
                        <div class='install-instructions'>Click the button below to retrieve annotations</div>
                        {% endif %}
                        <button class='load-cached-web-resource btn btn-primary mr-2' cached-web-resource='{{ cwr.name }}' >Update from web</button>
                    {% endif %}
                    <b>Imported:</b>&nbsp;{% timestamp cwr.modified time_ago=True %}.
                    {% if cwr.import_status == 'S' or user.is_superuser %}
                        <b>Details:</b>&nbsp;{{ cwr.description }}
                    {% else %}
                        <b>Status:</b> {{ cwr.get_import_status_display }}
                    {% endif %}
                {% endif %}
                </div>
            {% endlabelled %}
        {% endfor %}
    </div>
</div>

<script>
    $(".load-cached-web-resource").click(function() {
        const cwr_name = $(this).attr("cached-web-resource");
        const loadButton = $(this);

        $.ajax({
            type: "POST",
            url: Urls.load_cached_web_resource(cwr_name),
            success: function(data) {
                const img = $("<img />").attr("src", "{% static 'images/spinner.gif' %}");
                const container = loadButton.parent(".cached-web-resource-container");
                container.empty().append(img);
            }
        });
    });
</script>