{% extends "uicore/page/base.html" %}
{% load static %}
{% load js_tags %}
{% load ui_help %}
{% load ui_menu_bars %}
{% load ui_utils %}
{% load genome_build_tags %}
{% block submenu %}{% menu_bar_annotations %}{% endblock %}
{% block title %}VariantGrid Annotations{% endblock title %}

{% block head %}
    <style>
        .color-box {
            float: left;
            margin-right: 10px;
        }

         .annotation-level-header {
            margin: 10px 0px 10px 0px;
         }

        .annotation-level-title {
            float: left;
        }

        .annotations-table .annotation-row th {
            text-align: right;
            padding-right: 5px;
            max-width: 250px;
            min-width: 250px;
        }

        .vep-category {
            color: #fff;
            border-radius: 5px;
            white-space: nowrap;
            text-align: center;
            padding: 2px;
        }

        .vep-category-C {
            background-color: #02599C;
        }

        .vep-category-E {
            background-color: #333333;
        }

        .vep-category-F {
            background-color: #FF7F50;
        }

        .vep-category-G {
            background-color: #97f988;
            color: #680677;
        }

        .vep-category-H {
            background-color: #9400D3;
        }

        .vep-category-L {
            background-color: #6A5ACD;
        }

        .vep-category-N {
            background-color: #E75480;
        }

        .vep-category-P {
            background-color: #1E90FF;
        }

        .vep-category-Y {
            background-color: ;
        }

        .vep-category-D {
            background-color: #ff9900;
        }

        .vep-category-Q {
            background-color: #5F81A9;
        }

        .vep-category-S {
            background-color: red;
        }

        .vep-category-V {
        }
    </style>
{% endblock head %}
{% block content %}
    <div class="container">
    {% genome_build_url_arg genome_build 'view_annotation_descriptions_genome_build' %}

	{% page_help page_id='annotation/view_annotation_descriptions_help' title="VariantGrid Annotations"%}

    <ul>
    <li><a class="hover-link" href="{% url 'about_new_vep_columns' %}">Column changes switching to VEP</a>
    <li>To customise how these annotations appear in analysis grids, see <a class="hover-link" href="{% url 'custom_columns' %}">VariantGrid Columns</a><br>
    <li>To see an example of annotations in use, <a class="hover-link" href="{% url 'search' %}">search</a> for a variant in the top right box.
    </ul>

    <div>
        <div class="card">
            <div class="card-header d-flex align-center">
                <div class="color-box genotype-column">&nbsp;</div>
                <div class="annotation-level-title"><b>Sample Level</b> Imported VCF sample/genotype fields</div>
            </div>
            <!-- Hardcode Sample level special fields -->
            <div class="card-body">
                <table class='annotations-table table'>
                    <tr>
                        <th>Sample Zygosity</th>
                        <td>Zygosity call (HET/HOM_ALT/REF or Unknown)</td>
                    </tr>
                    <tr>
                        <th>AD (Allele Depth)</th>
                        <td>
                            Unfiltered allele depth, the number of reads that support each of the reported alleles. All reads at the position
                            (including reads that did not pass the variant caller’s filters) are included in this number, except reads that were
                            considered uninformative. Reads are considered uninformative when they do not provide enough statistical evidence
                            to support one allele over another.
                        </td>
                    </tr>

                    <tr>
                        <th>AF (Allele Frequency)</th>
                        <td>Unit value (0-1) of Allele Frequency. From "AF" FORMAT field if present in VCF otherwise calculated for each allele as <code>AD / (sum of all AD for that locus)</code>. See "VCF Info" tab on VCF page to see which was used for a particular file.</td>
                    </tr>

                    <tr>
                        <th>DP (Depth)</th>
                        <td>
                            Filtered depth, at the sample level. This gives you the number of filtered reads that support each of the reported alleles.
                            You can check the variant caller’s documentation to see which filters are applied by default. Only reads that passed the variant
                            caller’s filters are included in this number. However, unlike the AD calculation, uninformative reads are included in DP.
                        </td>
                    </tr>

                    <tr>
                        <th>GQ (Genotype Quality)</th>
                        <td>
                            Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype.
                        </td>
                    </tr>

                    <tr>
                        <th>PL (Phred Likelihood)</th>
                        <td>Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification</td>
                    </tr>
                </table>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header d-flex align-center">
                <div class="color-box database-column">&nbsp;</div>
                <div class="annotation-level-title"><b>Database</b> Internal database stats</div>
            </div>
            <div class="card-body">
                <table class='annotations-table table'>
                {% for v in variantgrid_columns_by_annotation_level.D %}
                    <tr>
                        <th>{{ v.pk }}</th>
                        <td>{{ v.description | safe }}</td>
                    </tr>
                {% endfor %}
                </table>
            </div>
        </div>

        {% for annotation_level, columns_and_vep in columns_and_vep_by_annotation_level.items %}
            <div class='card mt-4'>
                <div class="card-header align-center">
                    <div class="color-box {{ annotation_level|lower }}-column">&nbsp;</div>
                    <div class="annotation-level-title"><b>{{ annotation_level }} Level</b> Annotated by an <a class="hover-link" href="{% url 'variant_annotation_runs' %}">Variant Annotation Run</a>
                        {% if annotation_level == 'Transcript' %}
                          - can have different values per transcript
                        {% endif %}
                    </div>
                </div>
                <div class="card-body">
                    <table class='annotations-table table'>
                    <tr>
                        <th>Name</th>
                        <th>Category</th>
                        <th>VEP field</th>
                        <th>VEP custom/plugin</th>
                        <th>Details</th>
                    </tr>
                    {% for v, columnvepfield in columns_and_vep.items %}
                        <tr class='annotation-row'>
                            <th>{{ v.pk }}
                            <td>
                            {% if columnvepfield %}
                                <div class='vep-category vep-category-{{ columnvepfield.category }}'>
                                    {{ columnvepfield.get_category_display }}
                                </div>
                            {% endif %}
                            </td>
                            <td>
                            {% if columnvepfield %}
                                <div class='vep-source-field'>
                                    {{ columnvepfield.source_field }}
                                    {% if columnvepfield.source_field_processing_description %}
                                        ({{ columnvepfield.source_field_processing_description }})
                                    {% endif %}
                                </div>
                            {% endif %}
                            </td>
                            <td>
                            {% if columnvepfield %}
                                <div class='vep-source-field'>
                                    {% if columnvepfield.vep_plugin %}
                                        Plugin: <b>{{ columnvepfield.get_vep_plugin_display }}</b>
                                    {% endif %}

                                    {% if columnvepfield.vep_custom %}
                                        (custom)
                                    {% endif %}
                                </div>
                            {% endif %}
                            </td>
                            <td>{{ v.description | safe }}
                            {% if columnvepfield.min_vep_columns_version or columnvepfield.max_vep_columns_version %}
                                <div>
                                    Note:
                                    <b>{{ columnvepfield.columns_version_description }}</b> only
                                </div>
                            {% endif %}
                            {% if columnvepfield.vep_plugin == 'd' %}
                                <div>
                                Note: Values from dbNSFP are for non-synonymous SNVs only (no indels or intergenic)
                                </div>
                            {% endif %}
                            </td>
                         </tr>
                    {% endfor %}
                    </table>
                </div>
            </div>
        {% endfor %}

        <div class="card mt-4">
            <div class="card-header align-center">
                <div class="color-box hgnc-column">&nbsp;</div>
                <div class="annotation-level-title"><b>HGNC</b> From <a class="hover-link" target="_blank" href="https://www.genenames.org/">HUGO Gene Nomenclature Committee</a></div>
            </div>
            <div class="card-body">
                <table class='annotations-table table'>
                {% for v in variantgrid_columns_by_annotation_level.H %}
                    <tr>
                        <th>{{ v.pk }}</th>
                        <td>{{ v.description | safe }}</td>
                    </tr>
                {% endfor %}
                </table>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header align-center">
                <div class="color-box uniprot-column">&nbsp;</div>
                <div class="annotation-level-title"><b>UniProt</b> Protein function from <a class="hover-link" target="_blank" href="https://www.uniprot.org/">UniProt</a></div>
            </div>
            <div class="card-body">
                <table class='annotations-table table'>
                {% for v in variantgrid_columns_by_annotation_level.U %}
                    <tr>
                        <th>{{ v.pk }}</th>
                        <td>{{ v.description | safe }}</td>
                    </tr>
                {% endfor %}
                </table>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header align-center">
                <div class="color-box clinvar-column">&nbsp;</div>
                <div class="annotation-level-title"><b>ClinVar</b> Clinical Variants from <a class="hover-link" target="_blank" href="https://www.ncbi.nlm.nih.gov/clinvar/intro/">NCBI</a></div>
            </div>
            <div class="card-body">
                <table class='annotations-table table'>
                {% for v in variantgrid_columns_by_annotation_level.C %}
                    <tr>
                        <th>{{ v.pk }}</th>
                        <td>{{ v.description | safe }}</td>
                    </tr>
                {% endfor %}
                </table>
            </div>
        </div>

        <div class="card mt-4">
            <div class="card-header align-center">
                <div class="color-box gene-column">&nbsp;</div>
                <div class="annotation-level-title"><b>Gene Level</b> Per-gene information, matched via gene associated with variant's assigned ensembl_transcript_id</div>
            </div>
            <div class="card-body">
                <table class='annotations-table table'>
                {% for v in variantgrid_columns_by_annotation_level.G %}
                    <tr>
                        <th>{{ v.pk }}</th>
                        <td>{{ v.description | safe }}</td>
                    </tr>
                {% endfor %}
            </div>
        </div>
    </div>
{% endblock content %}